// @sentio/sdk/src/sui/builtin/0x3.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

/* Generated types for 0x3, original address 0x3 */

import { TypeDescriptor, ANY_TYPE } from "@typemove/move";
import { MoveCoder, TypedEventInstance } from "@typemove/sui";

import { defaultMoveCoder } from "@sentio/sdk/sui";

import { CallFilter, MoveFetchConfig } from "@sentio/sdk/move";
import {
  SuiBindOptions,
  SuiBaseProcessor,
  SuiNetwork,
  TypedFunctionPayload,
  SuiContext,
} from "@sentio/sdk/sui";

import * as _0x1 from "./0x1.js";
import * as _0x2 from "./0x2.js";

export class genesis extends SuiBaseProcessor {
  constructor(options: SuiBindOptions) {
    super("genesis", options);
  }
  static DEFAULT_OPTIONS: SuiBindOptions = {
    address: "0x3",
    network: SuiNetwork.MAIN_NET,
  };

  static bind(options: Partial<SuiBindOptions> = {}): genesis {
    return new genesis({ ...genesis.DEFAULT_OPTIONS, ...options });
  }

  onEventGenesisChainParameters(
    func: (
      event: genesis.GenesisChainParametersInstance,
      ctx: SuiContext,
    ) => void,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): genesis {
    this.onMoveEvent(
      func,
      { type: "genesis::GenesisChainParameters" },
      fetchConfig,
    );
    return this;
  }

  onEventGenesisValidatorMetadata(
    func: (
      event: genesis.GenesisValidatorMetadataInstance,
      ctx: SuiContext,
    ) => void,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): genesis {
    this.onMoveEvent(
      func,
      { type: "genesis::GenesisValidatorMetadata" },
      fetchConfig,
    );
    return this;
  }
}

export namespace genesis {
  export interface GenesisChainParameters {
    protocol_version: bigint;
    chain_start_timestamp_ms: bigint;
    epoch_duration_ms: bigint;
    stake_subsidy_start_epoch: bigint;
    stake_subsidy_initial_distribution_amount: bigint;
    stake_subsidy_period_length: bigint;
    stake_subsidy_decrease_rate: number;
    max_validator_count: bigint;
    min_validator_joining_stake: bigint;
    validator_low_stake_threshold: bigint;
    validator_very_low_stake_threshold: bigint;
    validator_low_stake_grace_period: bigint;
  }

  export namespace GenesisChainParameters {
    export const TYPE_QNAME = "0x3::genesis::GenesisChainParameters";

    const TYPE = new TypeDescriptor<GenesisChainParameters>(
      GenesisChainParameters.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<GenesisChainParameters> {
      return TYPE.apply();
    }
  }

  export interface GenesisChainParametersInstance
    extends TypedEventInstance<GenesisChainParameters> {
    data_decoded: GenesisChainParameters;
    type_arguments: [];
  }

  export interface GenesisValidatorMetadata {
    name: number[];
    description: number[];
    image_url: number[];
    project_url: number[];
    sui_address: string;
    gas_price: bigint;
    commission_rate: bigint;
    protocol_public_key: number[];
    proof_of_possession: number[];
    network_public_key: number[];
    worker_public_key: number[];
    network_address: number[];
    p2p_address: number[];
    primary_address: number[];
    worker_address: number[];
  }

  export namespace GenesisValidatorMetadata {
    export const TYPE_QNAME = "0x3::genesis::GenesisValidatorMetadata";

    const TYPE = new TypeDescriptor<GenesisValidatorMetadata>(
      GenesisValidatorMetadata.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<GenesisValidatorMetadata> {
      return TYPE.apply();
    }
  }

  export interface GenesisValidatorMetadataInstance
    extends TypedEventInstance<GenesisValidatorMetadata> {
    data_decoded: GenesisValidatorMetadata;
    type_arguments: [];
  }

  export interface TokenAllocation {
    recipient_address: string;
    amount_mist: bigint;
    staked_with_validator: _0x1.option.Option<string>;
  }

  export namespace TokenAllocation {
    export const TYPE_QNAME = "0x3::genesis::TokenAllocation";

    const TYPE = new TypeDescriptor<TokenAllocation>(
      TokenAllocation.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<TokenAllocation> {
      return TYPE.apply();
    }
  }

  export interface TokenDistributionSchedule {
    stake_subsidy_fund_mist: bigint;
    allocations: genesis.TokenAllocation[];
  }

  export namespace TokenDistributionSchedule {
    export const TYPE_QNAME = "0x3::genesis::TokenDistributionSchedule";

    const TYPE = new TypeDescriptor<TokenDistributionSchedule>(
      TokenDistributionSchedule.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<TokenDistributionSchedule> {
      return TYPE.apply();
    }
  }
}

export namespace stake_subsidy {
  export interface StakeSubsidy {
    balance: _0x2.balance.Balance<_0x2.sui.SUI>;
    distribution_counter: bigint;
    current_distribution_amount: bigint;
    stake_subsidy_period_length: bigint;
    stake_subsidy_decrease_rate: number;
    extra_fields: _0x2.bag.Bag;
  }

  export namespace StakeSubsidy {
    export const TYPE_QNAME = "0x3::stake_subsidy::StakeSubsidy";

    const TYPE = new TypeDescriptor<StakeSubsidy>(StakeSubsidy.TYPE_QNAME);

    export function type(): TypeDescriptor<StakeSubsidy> {
      return TYPE.apply();
    }
  }
}

export class staking_pool extends SuiBaseProcessor {
  constructor(options: SuiBindOptions) {
    super("staking_pool", options);
  }
  static DEFAULT_OPTIONS: SuiBindOptions = {
    address: "0x3",
    network: SuiNetwork.MAIN_NET,
  };

  static bind(options: Partial<SuiBindOptions> = {}): staking_pool {
    return new staking_pool({ ...staking_pool.DEFAULT_OPTIONS, ...options });
  }

  onEntryJoinStakedSui(
    func: (call: staking_pool.JoinStakedSuiPayload, ctx: SuiContext) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): staking_pool {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "staking_pool::join_staked_sui",
      },
      fetchConfig,
    );
    return this;
  }

  onEntrySplitStakedSui(
    func: (call: staking_pool.SplitStakedSuiPayload, ctx: SuiContext) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): staking_pool {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "staking_pool::split_staked_sui",
      },
      fetchConfig,
    );
    return this;
  }

  onEventPoolTokenExchangeRate(
    func: (
      event: staking_pool.PoolTokenExchangeRateInstance,
      ctx: SuiContext,
    ) => void,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): staking_pool {
    this.onMoveEvent(
      func,
      { type: "staking_pool::PoolTokenExchangeRate" },
      fetchConfig,
    );
    return this;
  }
}

export namespace staking_pool {
  export interface PoolTokenExchangeRate {
    sui_amount: bigint;
    pool_token_amount: bigint;
  }

  export namespace PoolTokenExchangeRate {
    export const TYPE_QNAME = "0x3::staking_pool::PoolTokenExchangeRate";

    const TYPE = new TypeDescriptor<PoolTokenExchangeRate>(
      PoolTokenExchangeRate.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<PoolTokenExchangeRate> {
      return TYPE.apply();
    }
  }

  export interface PoolTokenExchangeRateInstance
    extends TypedEventInstance<PoolTokenExchangeRate> {
    data_decoded: PoolTokenExchangeRate;
    type_arguments: [];
  }

  export interface StakedSui {
    id: _0x2.object_.UID;
    pool_id: _0x2.object_.ID;
    stake_activation_epoch: bigint;
    principal: _0x2.balance.Balance<_0x2.sui.SUI>;
  }

  export namespace StakedSui {
    export const TYPE_QNAME = "0x3::staking_pool::StakedSui";

    const TYPE = new TypeDescriptor<StakedSui>(StakedSui.TYPE_QNAME);

    export function type(): TypeDescriptor<StakedSui> {
      return TYPE.apply();
    }
  }

  export interface StakingPool {
    id: _0x2.object_.UID;
    activation_epoch: _0x1.option.Option<bigint>;
    deactivation_epoch: _0x1.option.Option<bigint>;
    sui_balance: bigint;
    rewards_pool: _0x2.balance.Balance<_0x2.sui.SUI>;
    pool_token_balance: bigint;
    exchange_rates: _0x2.table.Table<
      bigint,
      staking_pool.PoolTokenExchangeRate
    >;
    pending_stake: bigint;
    pending_total_sui_withdraw: bigint;
    pending_pool_token_withdraw: bigint;
    extra_fields: _0x2.bag.Bag;
  }

  export namespace StakingPool {
    export const TYPE_QNAME = "0x3::staking_pool::StakingPool";

    const TYPE = new TypeDescriptor<StakingPool>(StakingPool.TYPE_QNAME);

    export function type(): TypeDescriptor<StakingPool> {
      return TYPE.apply();
    }
  }

  export interface JoinStakedSuiPayload
    extends TypedFunctionPayload<[string, staking_pool.StakedSui]> {
    arguments_decoded: [string, staking_pool.StakedSui];
    type_arguments: [];
  }

  export interface SplitStakedSuiPayload
    extends TypedFunctionPayload<[string, bigint, string]> {
    arguments_decoded: [string, bigint, string];
    type_arguments: [];
  }
}

export namespace storage_fund {
  export interface StorageFund {
    total_object_storage_rebates: _0x2.balance.Balance<_0x2.sui.SUI>;
    non_refundable_balance: _0x2.balance.Balance<_0x2.sui.SUI>;
  }

  export namespace StorageFund {
    export const TYPE_QNAME = "0x3::storage_fund::StorageFund";

    const TYPE = new TypeDescriptor<StorageFund>(StorageFund.TYPE_QNAME);

    export function type(): TypeDescriptor<StorageFund> {
      return TYPE.apply();
    }
  }
}

export class sui_system extends SuiBaseProcessor {
  constructor(options: SuiBindOptions) {
    super("sui_system", options);
  }
  static DEFAULT_OPTIONS: SuiBindOptions = {
    address: "0x3",
    network: SuiNetwork.MAIN_NET,
  };

  static bind(options: Partial<SuiBindOptions> = {}): sui_system {
    return new sui_system({ ...sui_system.DEFAULT_OPTIONS, ...options });
  }

  onEntryReportValidator(
    func: (call: sui_system.ReportValidatorPayload, ctx: SuiContext) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::report_validator",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryRequestAddStake(
    func: (call: sui_system.RequestAddStakePayload, ctx: SuiContext) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::request_add_stake",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryRequestAddStakeMulCoin(
    func: (
      call: sui_system.RequestAddStakeMulCoinPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::request_add_stake_mul_coin",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryRequestAddValidator(
    func: (
      call: sui_system.RequestAddValidatorPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::request_add_validator",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryRequestAddValidatorCandidate(
    func: (
      call: sui_system.RequestAddValidatorCandidatePayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::request_add_validator_candidate",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryRequestRemoveValidator(
    func: (
      call: sui_system.RequestRemoveValidatorPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::request_remove_validator",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryRequestRemoveValidatorCandidate(
    func: (
      call: sui_system.RequestRemoveValidatorCandidatePayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::request_remove_validator_candidate",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryRequestSetCommissionRate(
    func: (
      call: sui_system.RequestSetCommissionRatePayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::request_set_commission_rate",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryRequestSetGasPrice(
    func: (call: sui_system.RequestSetGasPricePayload, ctx: SuiContext) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::request_set_gas_price",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryRequestWithdrawStake(
    func: (
      call: sui_system.RequestWithdrawStakePayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::request_withdraw_stake",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryRotateOperationCap(
    func: (call: sui_system.RotateOperationCapPayload, ctx: SuiContext) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::rotate_operation_cap",
      },
      fetchConfig,
    );
    return this;
  }

  onEntrySetCandidateValidatorCommissionRate(
    func: (
      call: sui_system.SetCandidateValidatorCommissionRatePayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::set_candidate_validator_commission_rate",
      },
      fetchConfig,
    );
    return this;
  }

  onEntrySetCandidateValidatorGasPrice(
    func: (
      call: sui_system.SetCandidateValidatorGasPricePayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::set_candidate_validator_gas_price",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUndoReportValidator(
    func: (
      call: sui_system.UndoReportValidatorPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::undo_report_validator",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateCandidateValidatorNetworkAddress(
    func: (
      call: sui_system.UpdateCandidateValidatorNetworkAddressPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_candidate_validator_network_address",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateCandidateValidatorNetworkPubkey(
    func: (
      call: sui_system.UpdateCandidateValidatorNetworkPubkeyPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_candidate_validator_network_pubkey",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateCandidateValidatorP2pAddress(
    func: (
      call: sui_system.UpdateCandidateValidatorP2pAddressPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_candidate_validator_p2p_address",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateCandidateValidatorPrimaryAddress(
    func: (
      call: sui_system.UpdateCandidateValidatorPrimaryAddressPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_candidate_validator_primary_address",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateCandidateValidatorProtocolPubkey(
    func: (
      call: sui_system.UpdateCandidateValidatorProtocolPubkeyPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_candidate_validator_protocol_pubkey",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateCandidateValidatorWorkerAddress(
    func: (
      call: sui_system.UpdateCandidateValidatorWorkerAddressPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_candidate_validator_worker_address",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateCandidateValidatorWorkerPubkey(
    func: (
      call: sui_system.UpdateCandidateValidatorWorkerPubkeyPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_candidate_validator_worker_pubkey",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateValidatorDescription(
    func: (
      call: sui_system.UpdateValidatorDescriptionPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_validator_description",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateValidatorImageUrl(
    func: (
      call: sui_system.UpdateValidatorImageUrlPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_validator_image_url",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateValidatorName(
    func: (
      call: sui_system.UpdateValidatorNamePayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_validator_name",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateValidatorNextEpochNetworkAddress(
    func: (
      call: sui_system.UpdateValidatorNextEpochNetworkAddressPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_validator_next_epoch_network_address",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateValidatorNextEpochNetworkPubkey(
    func: (
      call: sui_system.UpdateValidatorNextEpochNetworkPubkeyPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_validator_next_epoch_network_pubkey",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateValidatorNextEpochP2pAddress(
    func: (
      call: sui_system.UpdateValidatorNextEpochP2pAddressPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_validator_next_epoch_p2p_address",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateValidatorNextEpochPrimaryAddress(
    func: (
      call: sui_system.UpdateValidatorNextEpochPrimaryAddressPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_validator_next_epoch_primary_address",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateValidatorNextEpochProtocolPubkey(
    func: (
      call: sui_system.UpdateValidatorNextEpochProtocolPubkeyPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_validator_next_epoch_protocol_pubkey",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateValidatorNextEpochWorkerAddress(
    func: (
      call: sui_system.UpdateValidatorNextEpochWorkerAddressPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_validator_next_epoch_worker_address",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateValidatorNextEpochWorkerPubkey(
    func: (
      call: sui_system.UpdateValidatorNextEpochWorkerPubkeyPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_validator_next_epoch_worker_pubkey",
      },
      fetchConfig,
    );
    return this;
  }

  onEntryUpdateValidatorProjectUrl(
    func: (
      call: sui_system.UpdateValidatorProjectUrlPayload,
      ctx: SuiContext,
    ) => void,
    filter?: CallFilter,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system {
    this.onEntryFunctionCall(
      func,
      {
        ...filter,
        function: "sui_system::update_validator_project_url",
      },
      fetchConfig,
    );
    return this;
  }
}

export namespace sui_system {
  export interface SuiSystemState {
    id: _0x2.object_.UID;
    version: bigint;
  }

  export namespace SuiSystemState {
    export const TYPE_QNAME = "0x3::sui_system::SuiSystemState";

    const TYPE = new TypeDescriptor<SuiSystemState>(SuiSystemState.TYPE_QNAME);

    export function type(): TypeDescriptor<SuiSystemState> {
      return TYPE.apply();
    }
  }

  export interface ReportValidatorPayload
    extends TypedFunctionPayload<[string, string, string]> {
    arguments_decoded: [string, string, string];
    type_arguments: [];
  }

  export interface RequestAddStakePayload
    extends TypedFunctionPayload<
      [string, _0x2.coin.Coin<_0x2.sui.SUI>, string, string]
    > {
    arguments_decoded: [string, _0x2.coin.Coin<_0x2.sui.SUI>, string, string];
    type_arguments: [];
  }

  export interface RequestAddStakeMulCoinPayload
    extends TypedFunctionPayload<
      [
        string,
        _0x2.coin.Coin<_0x2.sui.SUI>[],
        _0x1.option.Option<bigint>,
        string,
        string,
      ]
    > {
    arguments_decoded: [
      string,
      _0x2.coin.Coin<_0x2.sui.SUI>[],
      _0x1.option.Option<bigint>,
      string,
      string,
    ];
    type_arguments: [];
  }

  export interface RequestAddValidatorPayload
    extends TypedFunctionPayload<[string, string]> {
    arguments_decoded: [string, string];
    type_arguments: [];
  }

  export interface RequestAddValidatorCandidatePayload
    extends TypedFunctionPayload<
      [
        string,
        number[],
        number[],
        number[],
        number[],
        number[],
        number[],
        number[],
        number[],
        number[],
        number[],
        number[],
        number[],
        bigint,
        bigint,
        string,
      ]
    > {
    arguments_decoded: [
      string,
      number[],
      number[],
      number[],
      number[],
      number[],
      number[],
      number[],
      number[],
      number[],
      number[],
      number[],
      number[],
      bigint,
      bigint,
      string,
    ];
    type_arguments: [];
  }

  export interface RequestRemoveValidatorPayload
    extends TypedFunctionPayload<[string, string]> {
    arguments_decoded: [string, string];
    type_arguments: [];
  }

  export interface RequestRemoveValidatorCandidatePayload
    extends TypedFunctionPayload<[string, string]> {
    arguments_decoded: [string, string];
    type_arguments: [];
  }

  export interface RequestSetCommissionRatePayload
    extends TypedFunctionPayload<[string, bigint, string]> {
    arguments_decoded: [string, bigint, string];
    type_arguments: [];
  }

  export interface RequestSetGasPricePayload
    extends TypedFunctionPayload<[string, string, bigint]> {
    arguments_decoded: [string, string, bigint];
    type_arguments: [];
  }

  export interface RequestWithdrawStakePayload
    extends TypedFunctionPayload<[string, staking_pool.StakedSui, string]> {
    arguments_decoded: [string, staking_pool.StakedSui, string];
    type_arguments: [];
  }

  export interface RotateOperationCapPayload
    extends TypedFunctionPayload<[string, string]> {
    arguments_decoded: [string, string];
    type_arguments: [];
  }

  export interface SetCandidateValidatorCommissionRatePayload
    extends TypedFunctionPayload<[string, bigint, string]> {
    arguments_decoded: [string, bigint, string];
    type_arguments: [];
  }

  export interface SetCandidateValidatorGasPricePayload
    extends TypedFunctionPayload<[string, string, bigint]> {
    arguments_decoded: [string, string, bigint];
    type_arguments: [];
  }

  export interface UndoReportValidatorPayload
    extends TypedFunctionPayload<[string, string, string]> {
    arguments_decoded: [string, string, string];
    type_arguments: [];
  }

  export interface UpdateCandidateValidatorNetworkAddressPayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }

  export interface UpdateCandidateValidatorNetworkPubkeyPayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }

  export interface UpdateCandidateValidatorP2pAddressPayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }

  export interface UpdateCandidateValidatorPrimaryAddressPayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }

  export interface UpdateCandidateValidatorProtocolPubkeyPayload
    extends TypedFunctionPayload<[string, number[], number[], string]> {
    arguments_decoded: [string, number[], number[], string];
    type_arguments: [];
  }

  export interface UpdateCandidateValidatorWorkerAddressPayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }

  export interface UpdateCandidateValidatorWorkerPubkeyPayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }

  export interface UpdateValidatorDescriptionPayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }

  export interface UpdateValidatorImageUrlPayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }

  export interface UpdateValidatorNamePayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }

  export interface UpdateValidatorNextEpochNetworkAddressPayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }

  export interface UpdateValidatorNextEpochNetworkPubkeyPayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }

  export interface UpdateValidatorNextEpochP2pAddressPayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }

  export interface UpdateValidatorNextEpochPrimaryAddressPayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }

  export interface UpdateValidatorNextEpochProtocolPubkeyPayload
    extends TypedFunctionPayload<[string, number[], number[], string]> {
    arguments_decoded: [string, number[], number[], string];
    type_arguments: [];
  }

  export interface UpdateValidatorNextEpochWorkerAddressPayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }

  export interface UpdateValidatorNextEpochWorkerPubkeyPayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }

  export interface UpdateValidatorProjectUrlPayload
    extends TypedFunctionPayload<[string, number[], string]> {
    arguments_decoded: [string, number[], string];
    type_arguments: [];
  }
}

export class sui_system_state_inner extends SuiBaseProcessor {
  constructor(options: SuiBindOptions) {
    super("sui_system_state_inner", options);
  }
  static DEFAULT_OPTIONS: SuiBindOptions = {
    address: "0x3",
    network: SuiNetwork.MAIN_NET,
  };

  static bind(options: Partial<SuiBindOptions> = {}): sui_system_state_inner {
    return new sui_system_state_inner({
      ...sui_system_state_inner.DEFAULT_OPTIONS,
      ...options,
    });
  }

  onEventSystemEpochInfoEvent(
    func: (
      event: sui_system_state_inner.SystemEpochInfoEventInstance,
      ctx: SuiContext,
    ) => void,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): sui_system_state_inner {
    this.onMoveEvent(
      func,
      { type: "sui_system_state_inner::SystemEpochInfoEvent" },
      fetchConfig,
    );
    return this;
  }
}

export namespace sui_system_state_inner {
  export interface SuiSystemStateInner {
    epoch: bigint;
    protocol_version: bigint;
    system_state_version: bigint;
    validators: validator_set.ValidatorSet;
    storage_fund: storage_fund.StorageFund;
    parameters: sui_system_state_inner.SystemParameters;
    reference_gas_price: bigint;
    validator_report_records: _0x2.vec_map.VecMap<
      string,
      _0x2.vec_set.VecSet<string>
    >;
    stake_subsidy: stake_subsidy.StakeSubsidy;
    safe_mode: Boolean;
    safe_mode_storage_rewards: _0x2.balance.Balance<_0x2.sui.SUI>;
    safe_mode_computation_rewards: _0x2.balance.Balance<_0x2.sui.SUI>;
    safe_mode_storage_rebates: bigint;
    safe_mode_non_refundable_storage_fee: bigint;
    epoch_start_timestamp_ms: bigint;
    extra_fields: _0x2.bag.Bag;
  }

  export namespace SuiSystemStateInner {
    export const TYPE_QNAME =
      "0x3::sui_system_state_inner::SuiSystemStateInner";

    const TYPE = new TypeDescriptor<SuiSystemStateInner>(
      SuiSystemStateInner.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<SuiSystemStateInner> {
      return TYPE.apply();
    }
  }

  export interface SuiSystemStateInnerV2 {
    epoch: bigint;
    protocol_version: bigint;
    system_state_version: bigint;
    validators: validator_set.ValidatorSet;
    storage_fund: storage_fund.StorageFund;
    parameters: sui_system_state_inner.SystemParametersV2;
    reference_gas_price: bigint;
    validator_report_records: _0x2.vec_map.VecMap<
      string,
      _0x2.vec_set.VecSet<string>
    >;
    stake_subsidy: stake_subsidy.StakeSubsidy;
    safe_mode: Boolean;
    safe_mode_storage_rewards: _0x2.balance.Balance<_0x2.sui.SUI>;
    safe_mode_computation_rewards: _0x2.balance.Balance<_0x2.sui.SUI>;
    safe_mode_storage_rebates: bigint;
    safe_mode_non_refundable_storage_fee: bigint;
    epoch_start_timestamp_ms: bigint;
    extra_fields: _0x2.bag.Bag;
  }

  export namespace SuiSystemStateInnerV2 {
    export const TYPE_QNAME =
      "0x3::sui_system_state_inner::SuiSystemStateInnerV2";

    const TYPE = new TypeDescriptor<SuiSystemStateInnerV2>(
      SuiSystemStateInnerV2.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<SuiSystemStateInnerV2> {
      return TYPE.apply();
    }
  }

  export interface SystemEpochInfoEvent {
    epoch: bigint;
    protocol_version: bigint;
    reference_gas_price: bigint;
    total_stake: bigint;
    storage_fund_reinvestment: bigint;
    storage_charge: bigint;
    storage_rebate: bigint;
    storage_fund_balance: bigint;
    stake_subsidy_amount: bigint;
    total_gas_fees: bigint;
    total_stake_rewards_distributed: bigint;
    leftover_storage_fund_inflow: bigint;
  }

  export namespace SystemEpochInfoEvent {
    export const TYPE_QNAME =
      "0x3::sui_system_state_inner::SystemEpochInfoEvent";

    const TYPE = new TypeDescriptor<SystemEpochInfoEvent>(
      SystemEpochInfoEvent.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<SystemEpochInfoEvent> {
      return TYPE.apply();
    }
  }

  export interface SystemEpochInfoEventInstance
    extends TypedEventInstance<SystemEpochInfoEvent> {
    data_decoded: SystemEpochInfoEvent;
    type_arguments: [];
  }

  export interface SystemParameters {
    epoch_duration_ms: bigint;
    stake_subsidy_start_epoch: bigint;
    max_validator_count: bigint;
    min_validator_joining_stake: bigint;
    validator_low_stake_threshold: bigint;
    validator_very_low_stake_threshold: bigint;
    validator_low_stake_grace_period: bigint;
    extra_fields: _0x2.bag.Bag;
  }

  export namespace SystemParameters {
    export const TYPE_QNAME = "0x3::sui_system_state_inner::SystemParameters";

    const TYPE = new TypeDescriptor<SystemParameters>(
      SystemParameters.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<SystemParameters> {
      return TYPE.apply();
    }
  }

  export interface SystemParametersV2 {
    epoch_duration_ms: bigint;
    stake_subsidy_start_epoch: bigint;
    min_validator_count: bigint;
    max_validator_count: bigint;
    min_validator_joining_stake: bigint;
    validator_low_stake_threshold: bigint;
    validator_very_low_stake_threshold: bigint;
    validator_low_stake_grace_period: bigint;
    extra_fields: _0x2.bag.Bag;
  }

  export namespace SystemParametersV2 {
    export const TYPE_QNAME = "0x3::sui_system_state_inner::SystemParametersV2";

    const TYPE = new TypeDescriptor<SystemParametersV2>(
      SystemParametersV2.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<SystemParametersV2> {
      return TYPE.apply();
    }
  }
}

export class validator extends SuiBaseProcessor {
  constructor(options: SuiBindOptions) {
    super("validator", options);
  }
  static DEFAULT_OPTIONS: SuiBindOptions = {
    address: "0x3",
    network: SuiNetwork.MAIN_NET,
  };

  static bind(options: Partial<SuiBindOptions> = {}): validator {
    return new validator({ ...validator.DEFAULT_OPTIONS, ...options });
  }

  onEventStakingRequestEvent(
    func: (
      event: validator.StakingRequestEventInstance,
      ctx: SuiContext,
    ) => void,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): validator {
    this.onMoveEvent(
      func,
      { type: "validator::StakingRequestEvent" },
      fetchConfig,
    );
    return this;
  }

  onEventUnstakingRequestEvent(
    func: (
      event: validator.UnstakingRequestEventInstance,
      ctx: SuiContext,
    ) => void,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): validator {
    this.onMoveEvent(
      func,
      { type: "validator::UnstakingRequestEvent" },
      fetchConfig,
    );
    return this;
  }
}

export namespace validator {
  export interface StakingRequestEvent {
    pool_id: _0x2.object_.ID;
    validator_address: string;
    staker_address: string;
    epoch: bigint;
    amount: bigint;
  }

  export namespace StakingRequestEvent {
    export const TYPE_QNAME = "0x3::validator::StakingRequestEvent";

    const TYPE = new TypeDescriptor<StakingRequestEvent>(
      StakingRequestEvent.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<StakingRequestEvent> {
      return TYPE.apply();
    }
  }

  export interface StakingRequestEventInstance
    extends TypedEventInstance<StakingRequestEvent> {
    data_decoded: StakingRequestEvent;
    type_arguments: [];
  }

  export interface UnstakingRequestEvent {
    pool_id: _0x2.object_.ID;
    validator_address: string;
    staker_address: string;
    stake_activation_epoch: bigint;
    unstaking_epoch: bigint;
    principal_amount: bigint;
    reward_amount: bigint;
  }

  export namespace UnstakingRequestEvent {
    export const TYPE_QNAME = "0x3::validator::UnstakingRequestEvent";

    const TYPE = new TypeDescriptor<UnstakingRequestEvent>(
      UnstakingRequestEvent.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<UnstakingRequestEvent> {
      return TYPE.apply();
    }
  }

  export interface UnstakingRequestEventInstance
    extends TypedEventInstance<UnstakingRequestEvent> {
    data_decoded: UnstakingRequestEvent;
    type_arguments: [];
  }

  export interface Validator {
    metadata: validator.ValidatorMetadata;
    voting_power: bigint;
    operation_cap_id: _0x2.object_.ID;
    gas_price: bigint;
    staking_pool: staking_pool.StakingPool;
    commission_rate: bigint;
    next_epoch_stake: bigint;
    next_epoch_gas_price: bigint;
    next_epoch_commission_rate: bigint;
    extra_fields: _0x2.bag.Bag;
  }

  export namespace Validator {
    export const TYPE_QNAME = "0x3::validator::Validator";

    const TYPE = new TypeDescriptor<Validator>(Validator.TYPE_QNAME);

    export function type(): TypeDescriptor<Validator> {
      return TYPE.apply();
    }
  }

  export interface ValidatorMetadata {
    sui_address: string;
    protocol_pubkey_bytes: number[];
    network_pubkey_bytes: number[];
    worker_pubkey_bytes: number[];
    proof_of_possession: number[];
    name: string;
    description: string;
    image_url: _0x2.url.Url;
    project_url: _0x2.url.Url;
    net_address: string;
    p2p_address: string;
    primary_address: string;
    worker_address: string;
    next_epoch_protocol_pubkey_bytes: _0x1.option.Option<number[]>;
    next_epoch_proof_of_possession: _0x1.option.Option<number[]>;
    next_epoch_network_pubkey_bytes: _0x1.option.Option<number[]>;
    next_epoch_worker_pubkey_bytes: _0x1.option.Option<number[]>;
    next_epoch_net_address: _0x1.option.Option<string>;
    next_epoch_p2p_address: _0x1.option.Option<string>;
    next_epoch_primary_address: _0x1.option.Option<string>;
    next_epoch_worker_address: _0x1.option.Option<string>;
    extra_fields: _0x2.bag.Bag;
  }

  export namespace ValidatorMetadata {
    export const TYPE_QNAME = "0x3::validator::ValidatorMetadata";

    const TYPE = new TypeDescriptor<ValidatorMetadata>(
      ValidatorMetadata.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<ValidatorMetadata> {
      return TYPE.apply();
    }
  }
}

export namespace validator_cap {
  export interface UnverifiedValidatorOperationCap {
    id: _0x2.object_.UID;
    authorizer_validator_address: string;
  }

  export namespace UnverifiedValidatorOperationCap {
    export const TYPE_QNAME =
      "0x3::validator_cap::UnverifiedValidatorOperationCap";

    const TYPE = new TypeDescriptor<UnverifiedValidatorOperationCap>(
      UnverifiedValidatorOperationCap.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<UnverifiedValidatorOperationCap> {
      return TYPE.apply();
    }
  }

  export interface ValidatorOperationCap {
    authorizer_validator_address: string;
  }

  export namespace ValidatorOperationCap {
    export const TYPE_QNAME = "0x3::validator_cap::ValidatorOperationCap";

    const TYPE = new TypeDescriptor<ValidatorOperationCap>(
      ValidatorOperationCap.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<ValidatorOperationCap> {
      return TYPE.apply();
    }
  }
}

export class validator_set extends SuiBaseProcessor {
  constructor(options: SuiBindOptions) {
    super("validator_set", options);
  }
  static DEFAULT_OPTIONS: SuiBindOptions = {
    address: "0x3",
    network: SuiNetwork.MAIN_NET,
  };

  static bind(options: Partial<SuiBindOptions> = {}): validator_set {
    return new validator_set({ ...validator_set.DEFAULT_OPTIONS, ...options });
  }

  onEventValidatorEpochInfoEvent(
    func: (
      event: validator_set.ValidatorEpochInfoEventInstance,
      ctx: SuiContext,
    ) => void,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): validator_set {
    this.onMoveEvent(
      func,
      { type: "validator_set::ValidatorEpochInfoEvent" },
      fetchConfig,
    );
    return this;
  }

  onEventValidatorEpochInfoEventV2(
    func: (
      event: validator_set.ValidatorEpochInfoEventV2Instance,
      ctx: SuiContext,
    ) => void,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): validator_set {
    this.onMoveEvent(
      func,
      { type: "validator_set::ValidatorEpochInfoEventV2" },
      fetchConfig,
    );
    return this;
  }

  onEventValidatorJoinEvent(
    func: (
      event: validator_set.ValidatorJoinEventInstance,
      ctx: SuiContext,
    ) => void,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): validator_set {
    this.onMoveEvent(
      func,
      { type: "validator_set::ValidatorJoinEvent" },
      fetchConfig,
    );
    return this;
  }

  onEventValidatorLeaveEvent(
    func: (
      event: validator_set.ValidatorLeaveEventInstance,
      ctx: SuiContext,
    ) => void,
    fetchConfig?: Partial<MoveFetchConfig>,
  ): validator_set {
    this.onMoveEvent(
      func,
      { type: "validator_set::ValidatorLeaveEvent" },
      fetchConfig,
    );
    return this;
  }
}

export namespace validator_set {
  export interface ValidatorEpochInfoEvent {
    epoch: bigint;
    validator_address: string;
    reference_gas_survey_quote: bigint;
    stake: bigint;
    commission_rate: bigint;
    pool_staking_reward: bigint;
    storage_fund_staking_reward: bigint;
    pool_token_exchange_rate: staking_pool.PoolTokenExchangeRate;
    tallying_rule_reporters: string[];
    tallying_rule_global_score: bigint;
  }

  export namespace ValidatorEpochInfoEvent {
    export const TYPE_QNAME = "0x3::validator_set::ValidatorEpochInfoEvent";

    const TYPE = new TypeDescriptor<ValidatorEpochInfoEvent>(
      ValidatorEpochInfoEvent.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<ValidatorEpochInfoEvent> {
      return TYPE.apply();
    }
  }

  export interface ValidatorEpochInfoEventInstance
    extends TypedEventInstance<ValidatorEpochInfoEvent> {
    data_decoded: ValidatorEpochInfoEvent;
    type_arguments: [];
  }

  export interface ValidatorEpochInfoEventV2 {
    epoch: bigint;
    validator_address: string;
    reference_gas_survey_quote: bigint;
    stake: bigint;
    voting_power: bigint;
    commission_rate: bigint;
    pool_staking_reward: bigint;
    storage_fund_staking_reward: bigint;
    pool_token_exchange_rate: staking_pool.PoolTokenExchangeRate;
    tallying_rule_reporters: string[];
    tallying_rule_global_score: bigint;
  }

  export namespace ValidatorEpochInfoEventV2 {
    export const TYPE_QNAME = "0x3::validator_set::ValidatorEpochInfoEventV2";

    const TYPE = new TypeDescriptor<ValidatorEpochInfoEventV2>(
      ValidatorEpochInfoEventV2.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<ValidatorEpochInfoEventV2> {
      return TYPE.apply();
    }
  }

  export interface ValidatorEpochInfoEventV2Instance
    extends TypedEventInstance<ValidatorEpochInfoEventV2> {
    data_decoded: ValidatorEpochInfoEventV2;
    type_arguments: [];
  }

  export interface ValidatorJoinEvent {
    epoch: bigint;
    validator_address: string;
    staking_pool_id: _0x2.object_.ID;
  }

  export namespace ValidatorJoinEvent {
    export const TYPE_QNAME = "0x3::validator_set::ValidatorJoinEvent";

    const TYPE = new TypeDescriptor<ValidatorJoinEvent>(
      ValidatorJoinEvent.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<ValidatorJoinEvent> {
      return TYPE.apply();
    }
  }

  export interface ValidatorJoinEventInstance
    extends TypedEventInstance<ValidatorJoinEvent> {
    data_decoded: ValidatorJoinEvent;
    type_arguments: [];
  }

  export interface ValidatorLeaveEvent {
    epoch: bigint;
    validator_address: string;
    staking_pool_id: _0x2.object_.ID;
    is_voluntary: Boolean;
  }

  export namespace ValidatorLeaveEvent {
    export const TYPE_QNAME = "0x3::validator_set::ValidatorLeaveEvent";

    const TYPE = new TypeDescriptor<ValidatorLeaveEvent>(
      ValidatorLeaveEvent.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<ValidatorLeaveEvent> {
      return TYPE.apply();
    }
  }

  export interface ValidatorLeaveEventInstance
    extends TypedEventInstance<ValidatorLeaveEvent> {
    data_decoded: ValidatorLeaveEvent;
    type_arguments: [];
  }

  export interface ValidatorSet {
    total_stake: bigint;
    active_validators: validator.Validator[];
    pending_active_validators: _0x2.table_vec.TableVec<validator.Validator>;
    pending_removals: bigint[];
    staking_pool_mappings: _0x2.table.Table<_0x2.object_.ID, string>;
    inactive_validators: _0x2.table.Table<
      _0x2.object_.ID,
      validator_wrapper.ValidatorWrapper
    >;
    validator_candidates: _0x2.table.Table<
      string,
      validator_wrapper.ValidatorWrapper
    >;
    at_risk_validators: _0x2.vec_map.VecMap<string, bigint>;
    extra_fields: _0x2.bag.Bag;
  }

  export namespace ValidatorSet {
    export const TYPE_QNAME = "0x3::validator_set::ValidatorSet";

    const TYPE = new TypeDescriptor<ValidatorSet>(ValidatorSet.TYPE_QNAME);

    export function type(): TypeDescriptor<ValidatorSet> {
      return TYPE.apply();
    }
  }
}

export namespace validator_wrapper {
  export interface ValidatorWrapper {
    inner: _0x2.versioned.Versioned;
  }

  export namespace ValidatorWrapper {
    export const TYPE_QNAME = "0x3::validator_wrapper::ValidatorWrapper";

    const TYPE = new TypeDescriptor<ValidatorWrapper>(
      ValidatorWrapper.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<ValidatorWrapper> {
      return TYPE.apply();
    }
  }
}

export namespace voting_power {
  export interface VotingPowerInfo {
    validator_index: bigint;
    voting_power: bigint;
  }

  export namespace VotingPowerInfo {
    export const TYPE_QNAME = "0x3::voting_power::VotingPowerInfo";

    const TYPE = new TypeDescriptor<VotingPowerInfo>(
      VotingPowerInfo.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<VotingPowerInfo> {
      return TYPE.apply();
    }
  }

  export interface VotingPowerInfoV2 {
    validator_index: bigint;
    voting_power: bigint;
    stake: bigint;
  }

  export namespace VotingPowerInfoV2 {
    export const TYPE_QNAME = "0x3::voting_power::VotingPowerInfoV2";

    const TYPE = new TypeDescriptor<VotingPowerInfoV2>(
      VotingPowerInfoV2.TYPE_QNAME,
    );

    export function type(): TypeDescriptor<VotingPowerInfoV2> {
      return TYPE.apply();
    }
  }
}

const MODULES = JSON.parse(
  '{"genesis":{"fileFormatVersion":6,"address":"0x3","name":"genesis","friends":[],"structs":{"GenesisChainParameters":{"abilities":{"abilities":["Copy","Drop"]},"typeParameters":[],"fields":[{"name":"protocol_version","type":"U64"},{"name":"chain_start_timestamp_ms","type":"U64"},{"name":"epoch_duration_ms","type":"U64"},{"name":"stake_subsidy_start_epoch","type":"U64"},{"name":"stake_subsidy_initial_distribution_amount","type":"U64"},{"name":"stake_subsidy_period_length","type":"U64"},{"name":"stake_subsidy_decrease_rate","type":"U16"},{"name":"max_validator_count","type":"U64"},{"name":"min_validator_joining_stake","type":"U64"},{"name":"validator_low_stake_threshold","type":"U64"},{"name":"validator_very_low_stake_threshold","type":"U64"},{"name":"validator_low_stake_grace_period","type":"U64"}]},"GenesisValidatorMetadata":{"abilities":{"abilities":["Copy","Drop"]},"typeParameters":[],"fields":[{"name":"name","type":{"Vector":"U8"}},{"name":"description","type":{"Vector":"U8"}},{"name":"image_url","type":{"Vector":"U8"}},{"name":"project_url","type":{"Vector":"U8"}},{"name":"sui_address","type":"Address"},{"name":"gas_price","type":"U64"},{"name":"commission_rate","type":"U64"},{"name":"protocol_public_key","type":{"Vector":"U8"}},{"name":"proof_of_possession","type":{"Vector":"U8"}},{"name":"network_public_key","type":{"Vector":"U8"}},{"name":"worker_public_key","type":{"Vector":"U8"}},{"name":"network_address","type":{"Vector":"U8"}},{"name":"p2p_address","type":{"Vector":"U8"}},{"name":"primary_address","type":{"Vector":"U8"}},{"name":"worker_address","type":{"Vector":"U8"}}]},"TokenAllocation":{"abilities":{"abilities":[]},"typeParameters":[],"fields":[{"name":"recipient_address","type":"Address"},{"name":"amount_mist","type":"U64"},{"name":"staked_with_validator","type":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":["Address"]}}}]},"TokenDistributionSchedule":{"abilities":{"abilities":[]},"typeParameters":[],"fields":[{"name":"stake_subsidy_fund_mist","type":"U64"},{"name":"allocations","type":{"Vector":{"Struct":{"address":"0x3","module":"genesis","name":"TokenAllocation","typeArguments":[]}}}}]}},"exposedFunctions":{}},"stake_subsidy":{"fileFormatVersion":6,"address":"0x3","name":"stake_subsidy","friends":[{"address":"0x3","name":"genesis"},{"address":"0x3","name":"sui_system_state_inner"}],"structs":{"StakeSubsidy":{"abilities":{"abilities":["Store"]},"typeParameters":[],"fields":[{"name":"balance","type":{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}},{"name":"distribution_counter","type":"U64"},{"name":"current_distribution_amount","type":"U64"},{"name":"stake_subsidy_period_length","type":"U64"},{"name":"stake_subsidy_decrease_rate","type":"U16"},{"name":"extra_fields","type":{"Struct":{"address":"0x2","module":"bag","name":"Bag","typeArguments":[]}}}]}},"exposedFunctions":{"advance_epoch":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"stake_subsidy","name":"StakeSubsidy","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}]},"create":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}},"U64","U64","U16",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x3","module":"stake_subsidy","name":"StakeSubsidy","typeArguments":[]}}]},"current_epoch_subsidy_amount":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"stake_subsidy","name":"StakeSubsidy","typeArguments":[]}}}],"return":["U64"]}}},"staking_pool":{"fileFormatVersion":6,"address":"0x3","name":"staking_pool","friends":[{"address":"0x3","name":"validator"},{"address":"0x3","name":"validator_set"}],"structs":{"PoolTokenExchangeRate":{"abilities":{"abilities":["Copy","Drop","Store"]},"typeParameters":[],"fields":[{"name":"sui_amount","type":"U64"},{"name":"pool_token_amount","type":"U64"}]},"StakedSui":{"abilities":{"abilities":["Store","Key"]},"typeParameters":[],"fields":[{"name":"id","type":{"Struct":{"address":"0x2","module":"object","name":"UID","typeArguments":[]}}},{"name":"pool_id","type":{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}},{"name":"stake_activation_epoch","type":"U64"},{"name":"principal","type":{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}}]},"StakingPool":{"abilities":{"abilities":["Store","Key"]},"typeParameters":[],"fields":[{"name":"id","type":{"Struct":{"address":"0x2","module":"object","name":"UID","typeArguments":[]}}},{"name":"activation_epoch","type":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":["U64"]}}},{"name":"deactivation_epoch","type":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":["U64"]}}},{"name":"sui_balance","type":"U64"},{"name":"rewards_pool","type":{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}},{"name":"pool_token_balance","type":"U64"},{"name":"exchange_rates","type":{"Struct":{"address":"0x2","module":"table","name":"Table","typeArguments":["U64",{"Struct":{"address":"0x3","module":"staking_pool","name":"PoolTokenExchangeRate","typeArguments":[]}}]}}},{"name":"pending_stake","type":"U64"},{"name":"pending_total_sui_withdraw","type":"U64"},{"name":"pending_pool_token_withdraw","type":"U64"},{"name":"extra_fields","type":{"Struct":{"address":"0x2","module":"bag","name":"Bag","typeArguments":[]}}}]}},"exposedFunctions":{"activate_staking_pool":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}},"U64"],"return":[]},"deactivate_staking_pool":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}},"U64"],"return":[]},"deposit_rewards":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}},{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}],"return":[]},"exchange_rates":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x2","module":"table","name":"Table","typeArguments":["U64",{"Struct":{"address":"0x3","module":"staking_pool","name":"PoolTokenExchangeRate","typeArguments":[]}}]}}}]},"is_equal_staking_metadata":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}}],"return":["Bool"]},"is_inactive":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}}],"return":["Bool"]},"is_preactive":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}}],"return":["Bool"]},"join_staked_sui":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}},{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}],"return":[]},"new":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}]},"pending_stake_amount":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}}],"return":["U64"]},"pending_stake_withdraw_amount":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}}],"return":["U64"]},"pool_id":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}]},"pool_token_amount":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"PoolTokenExchangeRate","typeArguments":[]}}}],"return":["U64"]},"pool_token_exchange_rate_at_epoch":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}},"U64"],"return":[{"Struct":{"address":"0x3","module":"staking_pool","name":"PoolTokenExchangeRate","typeArguments":[]}}]},"process_pending_stake":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}}],"return":[]},"process_pending_stakes_and_withdraws":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_add_stake":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}},{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}},"U64",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}]},"request_withdraw_stake":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}},{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}]},"split":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}},"U64",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}]},"split_staked_sui":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}},"U64",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"stake_activation_epoch":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}}],"return":["U64"]},"staked_sui_amount":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}}],"return":["U64"]},"sui_amount":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"PoolTokenExchangeRate","typeArguments":[]}}}],"return":["U64"]},"sui_balance":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}}],"return":["U64"]},"withdraw_from_principal":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}},{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}],"return":["U64",{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}]}}},"storage_fund":{"fileFormatVersion":6,"address":"0x3","name":"storage_fund","friends":[{"address":"0x3","name":"sui_system_state_inner"}],"structs":{"StorageFund":{"abilities":{"abilities":["Store"]},"typeParameters":[],"fields":[{"name":"total_object_storage_rebates","type":{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}},{"name":"non_refundable_balance","type":{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}}]}},"exposedFunctions":{"advance_epoch":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"storage_fund","name":"StorageFund","typeArguments":[]}}},{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}},{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}},{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}},"U64","U64"],"return":[{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}]},"new":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}],"return":[{"Struct":{"address":"0x3","module":"storage_fund","name":"StorageFund","typeArguments":[]}}]},"total_balance":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"storage_fund","name":"StorageFund","typeArguments":[]}}}],"return":["U64"]},"total_object_storage_rebates":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"storage_fund","name":"StorageFund","typeArguments":[]}}}],"return":["U64"]}}},"sui_system":{"fileFormatVersion":6,"address":"0x3","name":"sui_system","friends":[{"address":"0x3","name":"genesis"}],"structs":{"SuiSystemState":{"abilities":{"abilities":["Key"]},"typeParameters":[],"fields":[{"name":"id","type":{"Struct":{"address":"0x2","module":"object","name":"UID","typeArguments":[]}}},{"name":"version","type":"U64"}]}},"exposedFunctions":{"active_validator_addresses":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}}],"return":[{"Vector":"Address"}]},"create":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Struct":{"address":"0x2","module":"object","name":"UID","typeArguments":[]}},{"Vector":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}},"U64","U64",{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SystemParameters","typeArguments":[]}},{"Struct":{"address":"0x3","module":"stake_subsidy","name":"StakeSubsidy","typeArguments":[]}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"pool_exchange_rates":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x2","module":"table","name":"Table","typeArguments":["U64",{"Struct":{"address":"0x3","module":"staking_pool","name":"PoolTokenExchangeRate","typeArguments":[]}}]}}}]},"report_validator":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x3","module":"validator_cap","name":"UnverifiedValidatorOperationCap","typeArguments":[]}}},"Address"],"return":[]},"request_add_stake":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Struct":{"address":"0x2","module":"coin","name":"Coin","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}},"Address",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_add_stake_mul_coin":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":{"Struct":{"address":"0x2","module":"coin","name":"Coin","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}},{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":["U64"]}},"Address",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_add_stake_non_entry":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Struct":{"address":"0x2","module":"coin","name":"Coin","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}},"Address",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}]},"request_add_validator":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_add_validator_candidate":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},"U64","U64",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_remove_validator":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_remove_validator_candidate":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_set_commission_rate":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},"U64",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_set_gas_price":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x3","module":"validator_cap","name":"UnverifiedValidatorOperationCap","typeArguments":[]}}},"U64"],"return":[]},"request_withdraw_stake":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_withdraw_stake_non_entry":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}]},"rotate_operation_cap":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"set_candidate_validator_commission_rate":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},"U64",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"set_candidate_validator_gas_price":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x3","module":"validator_cap","name":"UnverifiedValidatorOperationCap","typeArguments":[]}}},"U64"],"return":[]},"undo_report_validator":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x3","module":"validator_cap","name":"UnverifiedValidatorOperationCap","typeArguments":[]}}},"Address"],"return":[]},"update_candidate_validator_network_address":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_candidate_validator_network_pubkey":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_candidate_validator_p2p_address":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_candidate_validator_primary_address":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_candidate_validator_protocol_pubkey":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_candidate_validator_worker_address":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_candidate_validator_worker_pubkey":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_description":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_image_url":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_name":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_next_epoch_network_address":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_next_epoch_network_pubkey":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_next_epoch_p2p_address":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_next_epoch_primary_address":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_next_epoch_protocol_pubkey":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_next_epoch_worker_address":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_next_epoch_worker_pubkey":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_project_url":{"visibility":"Public","isEntry":true,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system","name":"SuiSystemState","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]}}},"sui_system_state_inner":{"fileFormatVersion":6,"address":"0x3","name":"sui_system_state_inner","friends":[{"address":"0x3","name":"genesis"},{"address":"0x3","name":"sui_system"}],"structs":{"SuiSystemStateInner":{"abilities":{"abilities":["Store"]},"typeParameters":[],"fields":[{"name":"epoch","type":"U64"},{"name":"protocol_version","type":"U64"},{"name":"system_state_version","type":"U64"},{"name":"validators","type":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},{"name":"storage_fund","type":{"Struct":{"address":"0x3","module":"storage_fund","name":"StorageFund","typeArguments":[]}}},{"name":"parameters","type":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SystemParameters","typeArguments":[]}}},{"name":"reference_gas_price","type":"U64"},{"name":"validator_report_records","type":{"Struct":{"address":"0x2","module":"vec_map","name":"VecMap","typeArguments":["Address",{"Struct":{"address":"0x2","module":"vec_set","name":"VecSet","typeArguments":["Address"]}}]}}},{"name":"stake_subsidy","type":{"Struct":{"address":"0x3","module":"stake_subsidy","name":"StakeSubsidy","typeArguments":[]}}},{"name":"safe_mode","type":"Bool"},{"name":"safe_mode_storage_rewards","type":{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}},{"name":"safe_mode_computation_rewards","type":{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}},{"name":"safe_mode_storage_rebates","type":"U64"},{"name":"safe_mode_non_refundable_storage_fee","type":"U64"},{"name":"epoch_start_timestamp_ms","type":"U64"},{"name":"extra_fields","type":{"Struct":{"address":"0x2","module":"bag","name":"Bag","typeArguments":[]}}}]},"SuiSystemStateInnerV2":{"abilities":{"abilities":["Store"]},"typeParameters":[],"fields":[{"name":"epoch","type":"U64"},{"name":"protocol_version","type":"U64"},{"name":"system_state_version","type":"U64"},{"name":"validators","type":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},{"name":"storage_fund","type":{"Struct":{"address":"0x3","module":"storage_fund","name":"StorageFund","typeArguments":[]}}},{"name":"parameters","type":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SystemParametersV2","typeArguments":[]}}},{"name":"reference_gas_price","type":"U64"},{"name":"validator_report_records","type":{"Struct":{"address":"0x2","module":"vec_map","name":"VecMap","typeArguments":["Address",{"Struct":{"address":"0x2","module":"vec_set","name":"VecSet","typeArguments":["Address"]}}]}}},{"name":"stake_subsidy","type":{"Struct":{"address":"0x3","module":"stake_subsidy","name":"StakeSubsidy","typeArguments":[]}}},{"name":"safe_mode","type":"Bool"},{"name":"safe_mode_storage_rewards","type":{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}},{"name":"safe_mode_computation_rewards","type":{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}},{"name":"safe_mode_storage_rebates","type":"U64"},{"name":"safe_mode_non_refundable_storage_fee","type":"U64"},{"name":"epoch_start_timestamp_ms","type":"U64"},{"name":"extra_fields","type":{"Struct":{"address":"0x2","module":"bag","name":"Bag","typeArguments":[]}}}]},"SystemEpochInfoEvent":{"abilities":{"abilities":["Copy","Drop"]},"typeParameters":[],"fields":[{"name":"epoch","type":"U64"},{"name":"protocol_version","type":"U64"},{"name":"reference_gas_price","type":"U64"},{"name":"total_stake","type":"U64"},{"name":"storage_fund_reinvestment","type":"U64"},{"name":"storage_charge","type":"U64"},{"name":"storage_rebate","type":"U64"},{"name":"storage_fund_balance","type":"U64"},{"name":"stake_subsidy_amount","type":"U64"},{"name":"total_gas_fees","type":"U64"},{"name":"total_stake_rewards_distributed","type":"U64"},{"name":"leftover_storage_fund_inflow","type":"U64"}]},"SystemParameters":{"abilities":{"abilities":["Store"]},"typeParameters":[],"fields":[{"name":"epoch_duration_ms","type":"U64"},{"name":"stake_subsidy_start_epoch","type":"U64"},{"name":"max_validator_count","type":"U64"},{"name":"min_validator_joining_stake","type":"U64"},{"name":"validator_low_stake_threshold","type":"U64"},{"name":"validator_very_low_stake_threshold","type":"U64"},{"name":"validator_low_stake_grace_period","type":"U64"},{"name":"extra_fields","type":{"Struct":{"address":"0x2","module":"bag","name":"Bag","typeArguments":[]}}}]},"SystemParametersV2":{"abilities":{"abilities":["Store"]},"typeParameters":[],"fields":[{"name":"epoch_duration_ms","type":"U64"},{"name":"stake_subsidy_start_epoch","type":"U64"},{"name":"min_validator_count","type":"U64"},{"name":"max_validator_count","type":"U64"},{"name":"min_validator_joining_stake","type":"U64"},{"name":"validator_low_stake_threshold","type":"U64"},{"name":"validator_very_low_stake_threshold","type":"U64"},{"name":"validator_low_stake_grace_period","type":"U64"},{"name":"extra_fields","type":{"Struct":{"address":"0x2","module":"bag","name":"Bag","typeArguments":[]}}}]}},"exposedFunctions":{"active_validator_addresses":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}}],"return":[{"Vector":"Address"}]},"active_validator_voting_powers":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x2","module":"vec_map","name":"VecMap","typeArguments":["Address","U64"]}}]},"advance_epoch":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},"U64","U64",{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}},{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}},"U64","U64","U64","U64","U64",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}]},"create":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Vector":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}},"U64","U64",{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SystemParameters","typeArguments":[]}},{"Struct":{"address":"0x3","module":"stake_subsidy","name":"StakeSubsidy","typeArguments":[]}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInner","typeArguments":[]}}]},"create_system_parameters":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":["U64","U64","U64","U64","U64","U64","U64",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SystemParameters","typeArguments":[]}}]},"epoch":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}}],"return":["U64"]},"epoch_start_timestamp_ms":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}}],"return":["U64"]},"genesis_system_state_version":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[],"return":["U64"]},"get_reporters_of":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},"Address"],"return":[{"Struct":{"address":"0x2","module":"vec_set","name":"VecSet","typeArguments":["Address"]}}]},"get_storage_fund_object_rebates":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}}],"return":["U64"]},"get_storage_fund_total_balance":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}}],"return":["U64"]},"pool_exchange_rates":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x2","module":"table","name":"Table","typeArguments":["U64",{"Struct":{"address":"0x3","module":"staking_pool","name":"PoolTokenExchangeRate","typeArguments":[]}}]}}}]},"protocol_version":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}}],"return":["U64"]},"report_validator":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x3","module":"validator_cap","name":"UnverifiedValidatorOperationCap","typeArguments":[]}}},"Address"],"return":[]},"request_add_stake":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Struct":{"address":"0x2","module":"coin","name":"Coin","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}},"Address",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}]},"request_add_stake_mul_coin":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":{"Struct":{"address":"0x2","module":"coin","name":"Coin","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}},{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":["U64"]}},"Address",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}]},"request_add_validator":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_add_validator_candidate":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},"U64","U64",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_remove_validator":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_remove_validator_candidate":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_set_commission_rate":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},"U64",{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_set_gas_price":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x3","module":"validator_cap","name":"UnverifiedValidatorOperationCap","typeArguments":[]}}},"U64"],"return":[]},"request_withdraw_stake":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}]},"rotate_operation_cap":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"set_candidate_validator_commission_rate":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},"U64",{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"set_candidate_validator_gas_price":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x3","module":"validator_cap","name":"UnverifiedValidatorOperationCap","typeArguments":[]}}},"U64"],"return":[]},"system_state_version":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}}],"return":["U64"]},"undo_report_validator":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x3","module":"validator_cap","name":"UnverifiedValidatorOperationCap","typeArguments":[]}}},"Address"],"return":[]},"update_candidate_validator_network_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_candidate_validator_network_pubkey":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_candidate_validator_p2p_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_candidate_validator_primary_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_candidate_validator_protocol_pubkey":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_candidate_validator_worker_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_candidate_validator_worker_pubkey":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_description":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_image_url":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_name":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_next_epoch_network_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_next_epoch_network_pubkey":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_next_epoch_p2p_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_next_epoch_primary_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_next_epoch_protocol_pubkey":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_next_epoch_worker_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_next_epoch_worker_pubkey":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"update_validator_project_url":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},{"Vector":"U8"},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"v1_to_v2":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInner","typeArguments":[]}}],"return":[{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}]},"validator_stake_amount":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},"Address"],"return":["U64"]},"validator_staking_pool_id":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}},"Address"],"return":[{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}]},"validator_staking_pool_mappings":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"sui_system_state_inner","name":"SuiSystemStateInnerV2","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x2","module":"table","name":"Table","typeArguments":[{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}},"Address"]}}}]}}},"validator":{"fileFormatVersion":6,"address":"0x3","name":"validator","friends":[{"address":"0x3","name":"genesis"},{"address":"0x3","name":"sui_system_state_inner"},{"address":"0x3","name":"validator_set"},{"address":"0x3","name":"voting_power"}],"structs":{"StakingRequestEvent":{"abilities":{"abilities":["Copy","Drop"]},"typeParameters":[],"fields":[{"name":"pool_id","type":{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}},{"name":"validator_address","type":"Address"},{"name":"staker_address","type":"Address"},{"name":"epoch","type":"U64"},{"name":"amount","type":"U64"}]},"UnstakingRequestEvent":{"abilities":{"abilities":["Copy","Drop"]},"typeParameters":[],"fields":[{"name":"pool_id","type":{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}},{"name":"validator_address","type":"Address"},{"name":"staker_address","type":"Address"},{"name":"stake_activation_epoch","type":"U64"},{"name":"unstaking_epoch","type":"U64"},{"name":"principal_amount","type":"U64"},{"name":"reward_amount","type":"U64"}]},"Validator":{"abilities":{"abilities":["Store"]},"typeParameters":[],"fields":[{"name":"metadata","type":{"Struct":{"address":"0x3","module":"validator","name":"ValidatorMetadata","typeArguments":[]}}},{"name":"voting_power","type":"U64"},{"name":"operation_cap_id","type":{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}},{"name":"gas_price","type":"U64"},{"name":"staking_pool","type":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}},{"name":"commission_rate","type":"U64"},{"name":"next_epoch_stake","type":"U64"},{"name":"next_epoch_gas_price","type":"U64"},{"name":"next_epoch_commission_rate","type":"U64"},{"name":"extra_fields","type":{"Struct":{"address":"0x2","module":"bag","name":"Bag","typeArguments":[]}}}]},"ValidatorMetadata":{"abilities":{"abilities":["Store"]},"typeParameters":[],"fields":[{"name":"sui_address","type":"Address"},{"name":"protocol_pubkey_bytes","type":{"Vector":"U8"}},{"name":"network_pubkey_bytes","type":{"Vector":"U8"}},{"name":"worker_pubkey_bytes","type":{"Vector":"U8"}},{"name":"proof_of_possession","type":{"Vector":"U8"}},{"name":"name","type":{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}},{"name":"description","type":{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}},{"name":"image_url","type":{"Struct":{"address":"0x2","module":"url","name":"Url","typeArguments":[]}}},{"name":"project_url","type":{"Struct":{"address":"0x2","module":"url","name":"Url","typeArguments":[]}}},{"name":"net_address","type":{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}},{"name":"p2p_address","type":{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}},{"name":"primary_address","type":{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}},{"name":"worker_address","type":{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}},{"name":"next_epoch_protocol_pubkey_bytes","type":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Vector":"U8"}]}}},{"name":"next_epoch_proof_of_possession","type":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Vector":"U8"}]}}},{"name":"next_epoch_network_pubkey_bytes","type":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Vector":"U8"}]}}},{"name":"next_epoch_worker_pubkey_bytes","type":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Vector":"U8"}]}}},{"name":"next_epoch_net_address","type":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}]}}},{"name":"next_epoch_p2p_address","type":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}]}}},{"name":"next_epoch_primary_address","type":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}]}}},{"name":"next_epoch_worker_address","type":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}]}}},{"name":"extra_fields","type":{"Struct":{"address":"0x2","module":"bag","name":"Bag","typeArguments":[]}}}]}},"exposedFunctions":{"activate":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},"U64"],"return":[]},"adjust_stake_and_gas_price":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[]},"commission_rate":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":["U64"]},"deactivate":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},"U64"],"return":[]},"deposit_stake_rewards":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}],"return":[]},"description":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}}]},"effectuate_staged_metadata":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[]},"gas_price":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":["U64"]},"get_staking_pool_ref":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x3","module":"staking_pool","name":"StakingPool","typeArguments":[]}}}]},"image_url":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x2","module":"url","name":"Url","typeArguments":[]}}}]},"is_duplicate":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":["Bool"]},"is_preactive":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":["Bool"]},"metadata":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"ValidatorMetadata","typeArguments":[]}}}]},"name":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}}]},"network_address":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}}]},"network_pubkey_bytes":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Vector":"U8"}}]},"new":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":["Address",{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},"U64","U64",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}]},"new_metadata":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":["Address",{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Vector":"U8"},{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}},{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}},{"Struct":{"address":"0x2","module":"url","name":"Url","typeArguments":[]}},{"Struct":{"address":"0x2","module":"url","name":"Url","typeArguments":[]}},{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}},{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}},{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}},{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}},{"Struct":{"address":"0x2","module":"bag","name":"Bag","typeArguments":[]}}],"return":[{"Struct":{"address":"0x3","module":"validator","name":"ValidatorMetadata","typeArguments":[]}}]},"new_unverified_validator_operation_cap_and_transfer":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"next_epoch_gas_price":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":["U64"]},"next_epoch_network_address":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}]}}}]},"next_epoch_network_pubkey_bytes":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Vector":"U8"}]}}}]},"next_epoch_p2p_address":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}]}}}]},"next_epoch_primary_address":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}]}}}]},"next_epoch_proof_of_possession":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Vector":"U8"}]}}}]},"next_epoch_protocol_pubkey_bytes":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Vector":"U8"}]}}}]},"next_epoch_worker_address":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}]}}}]},"next_epoch_worker_pubkey_bytes":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x1","module":"option","name":"Option","typeArguments":[{"Vector":"U8"}]}}}]},"operation_cap_id":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}}]},"p2p_address":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}}]},"pending_stake_amount":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":["U64"]},"pending_stake_withdraw_amount":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":["U64"]},"pool_token_exchange_rate_at_epoch":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},"U64"],"return":[{"Struct":{"address":"0x3","module":"staking_pool","name":"PoolTokenExchangeRate","typeArguments":[]}}]},"primary_address":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}}]},"process_pending_stakes_and_withdraws":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"project_url":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x2","module":"url","name":"Url","typeArguments":[]}}}]},"proof_of_possession":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Vector":"U8"}}]},"protocol_pubkey_bytes":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Vector":"U8"}}]},"request_add_stake":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}},"Address",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}]},"request_add_stake_at_genesis":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}},"Address",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_set_commission_rate":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},"U64"],"return":[]},"request_set_gas_price":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Struct":{"address":"0x3","module":"validator_cap","name":"ValidatorOperationCap","typeArguments":[]}},"U64"],"return":[]},"request_withdraw_stake":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}]},"set_candidate_commission_rate":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},"U64"],"return":[]},"set_candidate_gas_price":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Struct":{"address":"0x3","module":"validator_cap","name":"ValidatorOperationCap","typeArguments":[]}},"U64"],"return":[]},"set_voting_power":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},"U64"],"return":[]},"stake_amount":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":["U64"]},"staking_pool_id":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}]},"sui_address":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":["Address"]},"total_stake":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":["U64"]},"total_stake_amount":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":["U64"]},"update_candidate_network_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"update_candidate_network_pubkey":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"update_candidate_p2p_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"update_candidate_primary_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"update_candidate_protocol_pubkey":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"},{"Vector":"U8"}],"return":[]},"update_candidate_worker_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"update_candidate_worker_pubkey":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"update_description":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"update_image_url":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"update_name":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"update_next_epoch_network_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"update_next_epoch_network_pubkey":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"update_next_epoch_p2p_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"update_next_epoch_primary_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"update_next_epoch_protocol_pubkey":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"},{"Vector":"U8"}],"return":[]},"update_next_epoch_worker_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"update_next_epoch_worker_pubkey":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"update_project_url":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"Vector":"U8"}],"return":[]},"validate_metadata":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"ValidatorMetadata","typeArguments":[]}}}],"return":[]},"validate_metadata_bcs":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Vector":"U8"}],"return":[]},"voting_power":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":["U64"]},"worker_address":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x1","module":"string","name":"String","typeArguments":[]}}}]},"worker_pubkey_bytes":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[{"Reference":{"Vector":"U8"}}]}}},"validator_cap":{"fileFormatVersion":6,"address":"0x3","name":"validator_cap","friends":[{"address":"0x3","name":"sui_system_state_inner"},{"address":"0x3","name":"validator"},{"address":"0x3","name":"validator_set"}],"structs":{"UnverifiedValidatorOperationCap":{"abilities":{"abilities":["Store","Key"]},"typeParameters":[],"fields":[{"name":"id","type":{"Struct":{"address":"0x2","module":"object","name":"UID","typeArguments":[]}}},{"name":"authorizer_validator_address","type":"Address"}]},"ValidatorOperationCap":{"abilities":{"abilities":["Drop"]},"typeParameters":[],"fields":[{"name":"authorizer_validator_address","type":"Address"}]}},"exposedFunctions":{"new_from_unverified":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_cap","name":"UnverifiedValidatorOperationCap","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x3","module":"validator_cap","name":"ValidatorOperationCap","typeArguments":[]}}]},"new_unverified_validator_operation_cap_and_transfer":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":["Address",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}]},"unverified_operation_cap_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_cap","name":"UnverifiedValidatorOperationCap","typeArguments":[]}}}],"return":[{"Reference":"Address"}]},"verified_operation_cap_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_cap","name":"ValidatorOperationCap","typeArguments":[]}}}],"return":[{"Reference":"Address"}]}}},"validator_set":{"fileFormatVersion":6,"address":"0x3","name":"validator_set","friends":[{"address":"0x3","name":"genesis"},{"address":"0x3","name":"sui_system_state_inner"}],"structs":{"ValidatorEpochInfoEvent":{"abilities":{"abilities":["Copy","Drop"]},"typeParameters":[],"fields":[{"name":"epoch","type":"U64"},{"name":"validator_address","type":"Address"},{"name":"reference_gas_survey_quote","type":"U64"},{"name":"stake","type":"U64"},{"name":"commission_rate","type":"U64"},{"name":"pool_staking_reward","type":"U64"},{"name":"storage_fund_staking_reward","type":"U64"},{"name":"pool_token_exchange_rate","type":{"Struct":{"address":"0x3","module":"staking_pool","name":"PoolTokenExchangeRate","typeArguments":[]}}},{"name":"tallying_rule_reporters","type":{"Vector":"Address"}},{"name":"tallying_rule_global_score","type":"U64"}]},"ValidatorEpochInfoEventV2":{"abilities":{"abilities":["Copy","Drop"]},"typeParameters":[],"fields":[{"name":"epoch","type":"U64"},{"name":"validator_address","type":"Address"},{"name":"reference_gas_survey_quote","type":"U64"},{"name":"stake","type":"U64"},{"name":"voting_power","type":"U64"},{"name":"commission_rate","type":"U64"},{"name":"pool_staking_reward","type":"U64"},{"name":"storage_fund_staking_reward","type":"U64"},{"name":"pool_token_exchange_rate","type":{"Struct":{"address":"0x3","module":"staking_pool","name":"PoolTokenExchangeRate","typeArguments":[]}}},{"name":"tallying_rule_reporters","type":{"Vector":"Address"}},{"name":"tallying_rule_global_score","type":"U64"}]},"ValidatorJoinEvent":{"abilities":{"abilities":["Copy","Drop"]},"typeParameters":[],"fields":[{"name":"epoch","type":"U64"},{"name":"validator_address","type":"Address"},{"name":"staking_pool_id","type":{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}}]},"ValidatorLeaveEvent":{"abilities":{"abilities":["Copy","Drop"]},"typeParameters":[],"fields":[{"name":"epoch","type":"U64"},{"name":"validator_address","type":"Address"},{"name":"staking_pool_id","type":{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}},{"name":"is_voluntary","type":"Bool"}]},"ValidatorSet":{"abilities":{"abilities":["Store"]},"typeParameters":[],"fields":[{"name":"total_stake","type":"U64"},{"name":"active_validators","type":{"Vector":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}},{"name":"pending_active_validators","type":{"Struct":{"address":"0x2","module":"table_vec","name":"TableVec","typeArguments":[{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}]}}},{"name":"pending_removals","type":{"Vector":"U64"}},{"name":"staking_pool_mappings","type":{"Struct":{"address":"0x2","module":"table","name":"Table","typeArguments":[{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}},"Address"]}}},{"name":"inactive_validators","type":{"Struct":{"address":"0x2","module":"table","name":"Table","typeArguments":[{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}},{"Struct":{"address":"0x3","module":"validator_wrapper","name":"ValidatorWrapper","typeArguments":[]}}]}}},{"name":"validator_candidates","type":{"Struct":{"address":"0x2","module":"table","name":"Table","typeArguments":["Address",{"Struct":{"address":"0x3","module":"validator_wrapper","name":"ValidatorWrapper","typeArguments":[]}}]}}},{"name":"at_risk_validators","type":{"Struct":{"address":"0x2","module":"vec_map","name":"VecMap","typeArguments":["Address","U64"]}}},{"name":"extra_fields","type":{"Struct":{"address":"0x2","module":"bag","name":"Bag","typeArguments":[]}}}]}},"exposedFunctions":{"active_validator_addresses":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}}],"return":[{"Vector":"Address"}]},"active_validators":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}}],"return":[{"Reference":{"Vector":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}}]},"advance_epoch":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},{"MutableReference":{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}},{"MutableReference":{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}},{"MutableReference":{"Struct":{"address":"0x2","module":"vec_map","name":"VecMap","typeArguments":["Address",{"Struct":{"address":"0x2","module":"vec_set","name":"VecSet","typeArguments":["Address"]}}]}}},"U64","U64","U64","U64",{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"assert_no_pending_or_active_duplicates":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":[]},"derive_reference_gas_price":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}}],"return":["U64"]},"get_active_or_pending_or_candidate_validator_ref":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},"Address","U8"],"return":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}]},"get_active_validator_ref":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},"Address"],"return":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}]},"get_pending_validator_ref":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},"Address"],"return":[{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}]},"get_validator_mut":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Vector":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}},"Address"],"return":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}]},"get_validator_mut_with_ctx":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}]},"get_validator_mut_with_ctx_including_candidates":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}]},"get_validator_mut_with_verified_cap":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x3","module":"validator_cap","name":"ValidatorOperationCap","typeArguments":[]}}},"Bool"],"return":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}]},"is_active_validator_by_sui_address":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},"Address"],"return":["Bool"]},"is_duplicate_validator":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Vector":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}},{"Reference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}],"return":["Bool"]},"is_inactive_validator":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}],"return":["Bool"]},"is_validator_candidate":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},"Address"],"return":["Bool"]},"new":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Vector":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}]},"next_epoch_validator_count":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}}],"return":["U64"]},"pool_exchange_rates":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x2","module":"table","name":"Table","typeArguments":["U64",{"Struct":{"address":"0x3","module":"staking_pool","name":"PoolTokenExchangeRate","typeArguments":[]}}]}}}]},"request_add_stake":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},"Address",{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}}]},"request_add_validator":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},"U64",{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_add_validator_candidate":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_remove_validator":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_remove_validator_candidate":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_set_commission_rate":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},"U64",{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[]},"request_withdraw_stake":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},{"Struct":{"address":"0x3","module":"staking_pool","name":"StakedSui","typeArguments":[]}},{"Reference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x2","module":"balance","name":"Balance","typeArguments":[{"Struct":{"address":"0x2","module":"sui","name":"SUI","typeArguments":[]}}]}}]},"staking_pool_mappings":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}}],"return":[{"Reference":{"Struct":{"address":"0x2","module":"table","name":"Table","typeArguments":[{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}},"Address"]}}}]},"sum_voting_power_by_addresses":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Vector":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}},{"Reference":{"Vector":"Address"}}],"return":["U64"]},"total_stake":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}}],"return":["U64"]},"validator_stake_amount":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},"Address"],"return":["U64"]},"validator_staking_pool_id":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},"Address"],"return":[{"Struct":{"address":"0x2","module":"object","name":"ID","typeArguments":[]}}]},"validator_total_stake_amount":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},"Address"],"return":["U64"]},"validator_voting_power":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[{"Reference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},"Address"],"return":["U64"]},"verify_cap":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator_set","name":"ValidatorSet","typeArguments":[]}}},{"Reference":{"Struct":{"address":"0x3","module":"validator_cap","name":"UnverifiedValidatorOperationCap","typeArguments":[]}}},"U8"],"return":[{"Struct":{"address":"0x3","module":"validator_cap","name":"ValidatorOperationCap","typeArguments":[]}}]}}},"validator_wrapper":{"fileFormatVersion":6,"address":"0x3","name":"validator_wrapper","friends":[{"address":"0x3","name":"validator_set"}],"structs":{"ValidatorWrapper":{"abilities":{"abilities":["Store"]},"typeParameters":[],"fields":[{"name":"inner","type":{"Struct":{"address":"0x2","module":"versioned","name":"Versioned","typeArguments":[]}}}]}},"exposedFunctions":{"create_v1":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}},{"MutableReference":{"Struct":{"address":"0x2","module":"tx_context","name":"TxContext","typeArguments":[]}}}],"return":[{"Struct":{"address":"0x3","module":"validator_wrapper","name":"ValidatorWrapper","typeArguments":[]}}]},"destroy":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"Struct":{"address":"0x3","module":"validator_wrapper","name":"ValidatorWrapper","typeArguments":[]}}],"return":[{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}]},"load_validator_maybe_upgrade":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator_wrapper","name":"ValidatorWrapper","typeArguments":[]}}}],"return":[{"MutableReference":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}]}}},"voting_power":{"fileFormatVersion":6,"address":"0x3","name":"voting_power","friends":[{"address":"0x3","name":"validator_set"}],"structs":{"VotingPowerInfo":{"abilities":{"abilities":["Drop"]},"typeParameters":[],"fields":[{"name":"validator_index","type":"U64"},{"name":"voting_power","type":"U64"}]},"VotingPowerInfoV2":{"abilities":{"abilities":["Drop"]},"typeParameters":[],"fields":[{"name":"validator_index","type":"U64"},{"name":"voting_power","type":"U64"},{"name":"stake","type":"U64"}]}},"exposedFunctions":{"quorum_threshold":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[],"return":["U64"]},"set_voting_power":{"visibility":"Friend","isEntry":false,"typeParameters":[],"parameters":[{"MutableReference":{"Vector":{"Struct":{"address":"0x3","module":"validator","name":"Validator","typeArguments":[]}}}}],"return":[]},"total_voting_power":{"visibility":"Public","isEntry":false,"typeParameters":[],"parameters":[],"return":["U64"]}}}}',
);

export function loadAllTypes(coder: MoveCoder) {
  _0x1.loadAllTypes(coder);
  _0x2.loadAllTypes(coder);
  for (const m of Object.values(MODULES)) {
    coder.load(m as any, "0x3");
  }
}

loadAllTypes(defaultMoveCoder(SuiNetwork.MAIN_NET));
loadAllTypes(defaultMoveCoder(SuiNetwork.TEST_NET));
loadAllTypes(defaultMoveCoder(SuiNetwork.MOVEMENT_TEST_NET));
loadAllTypes(defaultMoveCoder(SuiNetwork.MOVEMENT_MAIN_NET));
