// @sentio/sdk/lib/eth/builtin/internal/erc721-test-utils.js
import { EthChainId } from "@sentio/sdk/eth";
import { getERC721Contract } from "./erc721-processor.js";
const mockField = {
    blockHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
    blockNumber: 0,
    logIndex: 0,
    removed: false,
    transactionHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
    transactionIndex: 0,
};
export function mockApprovalLog(contractAddress, event) {
    const contract = getERC721Contract(EthChainId.ETHEREUM, contractAddress);
    const encodedLog = contract.rawContract.interface.encodeEventLog("Approval(address,address,uint256)", [event.owner, event.approved, event.tokenId]);
    return {
        ...mockField,
        index: 0,
        address: contractAddress,
        data: encodedLog.data,
        topics: encodedLog.topics,
    };
}
export function mockApprovalForAllLog(contractAddress, event) {
    const contract = getERC721Contract(EthChainId.ETHEREUM, contractAddress);
    const encodedLog = contract.rawContract.interface.encodeEventLog("ApprovalForAll(address,address,bool)", [event.owner, event.operator, event.approved]);
    return {
        ...mockField,
        index: 0,
        address: contractAddress,
        data: encodedLog.data,
        topics: encodedLog.topics,
    };
}
export function mockTransferLog(contractAddress, event) {
    const contract = getERC721Contract(EthChainId.ETHEREUM, contractAddress);
    const encodedLog = contract.rawContract.interface.encodeEventLog("Transfer(address,address,uint256)", [event.from, event.to, event.tokenId]);
    return {
        ...mockField,
        index: 0,
        address: contractAddress,
        data: encodedLog.data,
        topics: encodedLog.topics,
    };
}
