// @sentio/sdk/lib/eth/builtin/internal/weth9-test-utils.js
import { EthChainId } from "@sentio/sdk/eth";
import { getWETH9Contract } from "./weth9-processor.js";
const mockField = {
    blockHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
    blockNumber: 0,
    logIndex: 0,
    removed: false,
    transactionHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
    transactionIndex: 0,
};
export function mockApprovalLog(contractAddress, event) {
    const contract = getWETH9Contract(EthChainId.ETHEREUM, contractAddress);
    const encodedLog = contract.rawContract.interface.encodeEventLog("Approval(address,address,uint256)", [event.src, event.guy, event.wad]);
    return {
        ...mockField,
        index: 0,
        address: contractAddress,
        data: encodedLog.data,
        topics: encodedLog.topics,
    };
}
export function mockTransferLog(contractAddress, event) {
    const contract = getWETH9Contract(EthChainId.ETHEREUM, contractAddress);
    const encodedLog = contract.rawContract.interface.encodeEventLog("Transfer(address,address,uint256)", [event.src, event.dst, event.wad]);
    return {
        ...mockField,
        index: 0,
        address: contractAddress,
        data: encodedLog.data,
        topics: encodedLog.topics,
    };
}
export function mockDepositLog(contractAddress, event) {
    const contract = getWETH9Contract(EthChainId.ETHEREUM, contractAddress);
    const encodedLog = contract.rawContract.interface.encodeEventLog("Deposit(address,uint256)", [event.dst, event.wad]);
    return {
        ...mockField,
        index: 0,
        address: contractAddress,
        data: encodedLog.data,
        topics: encodedLog.topics,
    };
}
export function mockWithdrawalLog(contractAddress, event) {
    const contract = getWETH9Contract(EthChainId.ETHEREUM, contractAddress);
    const encodedLog = contract.rawContract.interface.encodeEventLog("Withdrawal(address,uint256)", [event.src, event.wad]);
    return {
        ...mockField,
        index: 0,
        address: contractAddress,
        data: encodedLog.data,
        topics: encodedLog.topics,
    };
}
