// @typemove/sui/dist/cjs/sui-chain-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuiChainAdapter = void 0;
const to_internal_js_1 = require("./to-internal.js");
const move_1 = require("@typemove/move");
const client_1 = require("@mysten/sui.js/client");
class SuiChainAdapter extends move_1.ChainAdapter {
    async getChainId() {
        return this.client.getChainIdentifier();
    }
    // static INSTANCE = new SuiChainAdapter()
    client;
    constructor(client) {
        super();
        this.client = client;
    }
    async fetchModule(account, module
    // network: SuiNetwork
    ) {
        return await this.client.getNormalizedMoveModule({ package: account, module });
    }
    async fetchModules(account
    // network: SuiNetwork
    ) {
        const modules = await this.client.getNormalizedMoveModulesByPackage({
            package: account
        });
        return Object.values(modules);
    }
    getMeaningfulFunctionParams(params) {
        return params;
        // if (params.length === 0) {
        //   return params
        // }
        // return params.slice(0, params.length - 1)
    }
    toInternalModules(modules) {
        return Object.values(modules).map(to_internal_js_1.toInternalModule);
    }
    getAllEventStructs(modules) {
        const eventMap = new Map();
        for (const module of modules) {
            const qname = (0, move_1.moduleQname)(module);
            for (const struct of module.structs) {
                const abilities = new Set(struct.abilities);
                if (abilities.has('Drop') && abilities.has('Copy')) {
                    eventMap.set(qname + move_1.SPLITTER + struct.name, struct);
                }
            }
        }
        return eventMap;
    }
    getType(base) {
        return base.type;
    }
    getData(val) {
        // if (val.parsedJson) {
        //   return val.parsedJson as any
        // }
        if (val === undefined) {
            throw Error('val is undefined');
        }
        if ('parsedJson' in val) {
            return val.parsedJson;
        }
        // if (SuiParsedData.is(val)) {
        //   return val.fields as any
        // }
        if (val.dataType === 'moveObject') {
            return val.fields;
        }
        // if (SuiMoveObject.is(val)) {
        //   return val.fields as any
        // }
        // This may not be perfect, just think everything has
        if ('fields' in val) {
            if ('type' in val && Object.keys(val).length === 2) {
                return val.fields;
            }
        }
        return val;
    }
}
exports.SuiChainAdapter = SuiChainAdapter;
// function getRpcEndpoint(network: SuiNetwork): string {
//   switch (network) {
//     case SuiNetwork.TEST_NET:
//       return 'https://fullnode.testnet.sui.io/'
//   }
//   return 'https://fullnode.mainnet.sui.io/'
// }
function getRpcClient(endpoint) {
    return new client_1.SuiClient({ url: endpoint });
}
