// @sentio/chain/dist/chain-detail.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getChainBlockscoutUrl = exports.getChainExternalUrl = exports.EthChainInfo = exports.ExplorerApiType = void 0;
const chain_id_1 = require("./chain-id");
var ExplorerApiType;
(function (ExplorerApiType) {
    ExplorerApiType["ETHERSCAN"] = "etherscan";
    ExplorerApiType["BLOCKSCOUT"] = "blockscout";
    ExplorerApiType["OKLINK"] = "oklink";
})(ExplorerApiType || (exports.ExplorerApiType = ExplorerApiType = {}));
// TODO add all this details
var EthChainName;
(function (EthChainName) {
    EthChainName["OPTIMISM"] = "Optimism";
    EthChainName["CRONOS"] = "Cronos Mainnet";
    EthChainName["MANTA_PACIFIC"] = "Manta Pacific";
    EthChainName["B2_MAINNET"] = "B2 Mainnet";
    EthChainName["FANTOM"] = "Fantom Opera";
    EthChainName["FRAXTAL"] = "Fraxtal";
    EthChainName["KUCOIN"] = "KuCoin";
    EthChainName["CRONOS_TESTNET"] = "Cronos Testnet";
    EthChainName["CONFLUX"] = "Conflux";
    EthChainName["METIS"] = "Metis";
    EthChainName["BEVM_CANARY"] = "BEVM Canary";
    EthChainName["BEVM"] = "BEVM";
    EthChainName["MERLIN_MAINNET"] = "Merlin Mainnet";
    EthChainName["MANTLE"] = "Mantle";
    EthChainName["LUMIO_TESTNET"] = "Lumio Testnet";
    EthChainName["MOVE_EVM_TESTNET"] = "Move EVM Testnet";
    EthChainName["MODE"] = "Mode Mainnet";
    EthChainName["BOB"] = "Bob Mainnet";
    EthChainName["BASE_GOERLI"] = "Base Goerli";
    EthChainName["CHILIZ"] = "Chiliz";
    EthChainName["BITLAYER"] = "Bitlayer";
    EthChainName["ZKLINK_NOVA"] = "zkLink Nova";
    EthChainName["AURORA"] = "Auroa";
})(EthChainName || (EthChainName = {}));
exports.EthChainInfo = {
    [chain_id_1.EthChainId.ETHEREUM]: {
        name: 'Ethereum',
        chainId: chain_id_1.EthChainId.ETHEREUM,
        priceTokenAddress: '0x0000000000000000000000000000000000000000',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerUrl: 'https://etherscan.io',
        explorerApi: 'https://api.etherscan.io',
        blockscoutUrl: 'https://eth.blockscout.com'
    },
    [chain_id_1.EthChainId.SEPOLIA]: {
        name: 'Sepolia',
        chainId: chain_id_1.EthChainId.SEPOLIA,
        priceTokenAddress: '0x0000000000000000000000000000000000000000',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0x1ef5f52bdbe11af2377c58ecc914a8c72ea807cf',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerUrl: 'https://sepolia.etherscan.io',
        explorerApi: 'https://api-sepolia.etherscan.io',
        blockscoutUrl: 'https://eth-sepolia.blockscout.com'
    },
    [chain_id_1.EthChainId.HOLESKY]: {
        name: 'Holesky',
        chainId: chain_id_1.EthChainId.HOLESKY,
        priceTokenAddress: '0x0000000000000000000000000000000000000000',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0x94373a4919B3240D86eA41593D5eBa789FEF3848',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerUrl: 'https://holesky.etherscan.io',
        explorerApi: 'https://api-holesky.etherscan.io',
        blockscoutUrl: 'https://eth-holesky.blockscout.com'
    },
    [chain_id_1.EthChainId.BINANCE]: {
        name: 'Binance',
        chainId: chain_id_1.EthChainId.BINANCE,
        priceTokenAddress: '0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c',
        tokenSymbol: 'BNB',
        tokenDecimals: 18,
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerUrl: 'https://bscscan.com',
        explorerApi: 'https://api.bscscan.com'
    },
    [chain_id_1.EthChainId.POLYGON]: {
        name: 'Polygon',
        chainId: chain_id_1.EthChainId.POLYGON,
        priceTokenAddress: '0x0d500b1d8e8ef31e21c99d1db9a6444d3adf1270',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0x0d500b1d8e8ef31e21c99d1db9a6444d3adf1270',
        tokenSymbol: 'MATIC',
        tokenDecimals: 18,
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerUrl: 'https://polygonscan.com',
        explorerApi: 'https://api.polygonscan.com'
        // blockscoutBlockPrefix: 'https://polygon.blockscout.com/block/',
    },
    [chain_id_1.EthChainId.ARBITRUM]: {
        name: 'Arbitrum',
        chainId: chain_id_1.EthChainId.ARBITRUM,
        priceTokenAddress: '0x0000000000000000000000000000000000000000',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0x82aF49447D8a07e3bd95BD0d56f35241523fBab1',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerUrl: 'https://arbiscan.io',
        explorerApi: 'https://api.arbiscan.io'
    },
    [chain_id_1.EthChainId.AVALANCHE]: {
        name: 'Avalanche',
        chainId: chain_id_1.EthChainId.AVALANCHE,
        priceTokenAddress: '0x0000000000000000000000000000000000000000',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0xB31f66AA3C1e785363F0875A1B74E27b85FD66c7',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerUrl: 'https://snowtrace.io',
        explorerApi: 'https://api.routescan.io/v2/network/mainnet/evm/43114/etherscan'
    },
    [chain_id_1.EthChainId.POLYGON_ZKEVM]: {
        name: 'Polygon zkEVM',
        chainId: chain_id_1.EthChainId.POLYGON_ZKEVM,
        priceTokenAddress: '0x0000000000000000000000000000000000000000',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0x4F9A0e7FD2Bf6067db6994CF12E4495Df938E6e9',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerUrl: 'https://zkevm.polygonscan.com',
        explorerApi: 'https://api-zkevm.polygonscan.com'
    },
    [chain_id_1.EthChainId.MOONBEAM]: {
        name: 'Moonbeam',
        chainId: chain_id_1.EthChainId.MOONBEAM,
        priceTokenAddress: '0xacc15dc74880c9944775448304b263d191c6077f',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0xacc15dc74880c9944775448304b263d191c6077f',
        tokenSymbol: 'GLMR',
        tokenDecimals: 18,
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerUrl: 'https://moonscan.io',
        explorerApi: 'https://api-moonbeam.moonscan.io'
    },
    [chain_id_1.EthChainId.ASTAR]: {
        name: 'Astar',
        chainId: chain_id_1.EthChainId.ASTAR,
        priceTokenAddress: '0xaeaaf0e2c81af264101b9129c00f4440ccf0f720',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0xaeaaf0e2c81af264101b9129c00f4440ccf0f720',
        tokenSymbol: 'ASTR',
        tokenDecimals: 18,
        explorerApiType: ExplorerApiType.BLOCKSCOUT,
        explorerUrl: 'https://astar.blockscout.com',
        explorerApi: 'https://astar.blockscout.com'
    },
    [chain_id_1.EthChainId.ASTAR_ZKEVM]: {
        name: 'Astar zkEVM',
        chainId: chain_id_1.EthChainId.ASTAR_ZKEVM,
        priceTokenAddress: '0x0000000000000000000000000000000000000000',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0xE9CC37904875B459Fa5D0FE37680d36F1ED55e38',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerUrl: 'https://astar-zkevm.explorer.startale.com',
        explorerApiType: ExplorerApiType.BLOCKSCOUT,
        explorerApi: 'https://astar-zkevm.explorer.startale.com'
    },
    [chain_id_1.EthChainId.LINEA]: {
        name: 'Linea',
        chainId: chain_id_1.EthChainId.LINEA,
        priceTokenAddress: '0xe5d7c2a44ffddf6b295a15c148167daaaf5cf34f',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0xe5D7C2a44FfDDf6b295A15c148167daaAf5Cf34f',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerUrl: 'https://lineascan.build',
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerApi: 'https://api.lineascan.build'
    },
    [chain_id_1.EthChainId.SCROLL]: {
        name: 'Scroll',
        chainId: chain_id_1.EthChainId.SCROLL,
        priceTokenAddress: '0x0000000000000000000000000000000000000000',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0x5300000000000000000000000000000000000004',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerUrl: 'https://scrollscan.com',
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerApi: 'https://api.scrollscan.com'
    },
    [chain_id_1.EthChainId.TAIKO]: {
        name: 'Taiko Mainnet',
        chainId: chain_id_1.EthChainId.TAIKO,
        priceTokenAddress: '0x0000000000000000000000000000000000000000',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0xA51894664A773981C6C112C43ce576f315d5b1B6',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerUrl: 'https://taikoscan.io',
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerApi: 'https://api.taikoscan.io'
    },
    [chain_id_1.EthChainId.TAIKO_TESTNET]: {
        name: 'Taiko Testnet',
        chainId: chain_id_1.EthChainId.TAIKO_TESTNET,
        priceTokenAddress: '0x0000000000000000000000000000000000000000',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0xae2C46ddb314B9Ba743C6dEE4878F151881333D9',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerUrl: 'https://hekla.taikoscan.io',
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerApi: 'https://api-hekla.taikoscan.io'
    },
    [chain_id_1.EthChainId.XLAYER_TESTNET]: {
        name: 'X Layer Testnet',
        chainId: chain_id_1.EthChainId.XLAYER_TESTNET,
        priceTokenAddress: '0x0000000000000000000000000000000000000000',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0xa749ad81913cdc19881ebeb64631df72be708335',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerUrl: 'https://www.oklink.com/xlayer-test',
        explorerApiType: ExplorerApiType.OKLINK,
        explorerApi: 'https://www.oklink.com/api/v5/explorer'
    },
    [chain_id_1.EthChainId.XLAYER_MAINNET]: {
        name: 'X Layer Mainnet',
        chainId: chain_id_1.EthChainId.XLAYER_MAINNET,
        priceTokenAddress: '0x0000000000000000000000000000000000000000',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0x5a77f1443d16ee5761d310e38b62f77f726bc71c',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerUrl: 'https://www.oklink.com/xlayer',
        explorerApiType: ExplorerApiType.OKLINK,
        explorerApi: 'https://www.oklink.com/api/v5/explorer'
    },
    [chain_id_1.EthChainId.BLAST]: {
        name: 'Blast Mainnet',
        chainId: chain_id_1.EthChainId.BLAST,
        priceTokenAddress: '0x4300000000000000000000000000000000000004',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0x4300000000000000000000000000000000000004',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerUrl: 'https://blastscan.io',
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerApi: 'https://api.blastscan.io'
    },
    [chain_id_1.EthChainId.BLAST_SEPOLIA]: {
        name: 'Blast Mainnet',
        chainId: chain_id_1.EthChainId.BLAST,
        priceTokenAddress: '0x4200000000000000000000000000000000000023',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0x4200000000000000000000000000000000000023',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerUrl: 'https://sepolia.blastscan.io',
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerApi: 'https://api-sepolia.blastscan.io'
    },
    [chain_id_1.EthChainId.BASE]: {
        name: 'Base',
        chainId: chain_id_1.EthChainId.BASE,
        priceTokenAddress: '0x4200000000000000000000000000000000000006',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0x4200000000000000000000000000000000000006',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerUrl: 'https://basescan.org',
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerApi: 'https://api.basescan.org'
    },
    [chain_id_1.EthChainId.ZKSYNC_ERA]: {
        name: 'zkSync Era',
        chainId: chain_id_1.EthChainId.ZKSYNC_ERA,
        priceTokenAddress: '0x000000000000000000000000000000000000800A',
        tokenAddress: '0x000000000000000000000000000000000000800A',
        wrappedTokenAddress: '0x000000000000000000000000000000000000800A',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerUrl: 'https://explorer.zksync.io',
        explorerApiType: ExplorerApiType.ETHERSCAN,
        explorerApi: 'https://block-explorer-api.mainnet.zksync.io'
    },
    [chain_id_1.EthChainId.ZIRCUIT_TESTNET]: {
        // TODO update the wrappedTokenAddress and explorer
        name: 'Zircuit Testnet',
        chainId: chain_id_1.EthChainId.ZIRCUIT_TESTNET,
        priceTokenAddress: '0x0000000000000000000000000000000000000000',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0xc4A8b48b1dB6584FB446Fc3BaE230dD3EF3C85DB',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerUrl: 'https://explorer.zircuit.com'
    },
    [chain_id_1.EthChainId.ZIRCUIT_MAINNET]: {
        // TODO update the wrappedTokenAddress
        name: 'Zircuit Mainnet',
        chainId: chain_id_1.EthChainId.ZIRCUIT_MAINNET,
        priceTokenAddress: '0x0000000000000000000000000000000000000000',
        tokenAddress: '0x0000000000000000000000000000000000000000',
        wrappedTokenAddress: '0x0000000000000000000000000000000000000000',
        tokenSymbol: 'ETH',
        tokenDecimals: 18,
        explorerUrl: 'https://explorer.zircuit.com'
    }
};
function getChainExternalUrl(chainId, hash, subtype) {
    if (!chainId || !hash || !subtype) {
        return;
    }
    const supportedChain = exports.EthChainInfo[chainId];
    if (!supportedChain) {
        return;
    }
    return `${supportedChain.explorerUrl}/${subtype}/${hash}`;
}
exports.getChainExternalUrl = getChainExternalUrl;
function getChainBlockscoutUrl(chainId, hash, subtype) {
    if (!chainId || !hash || !subtype) {
        return;
    }
    const supportedChain = exports.EthChainInfo[chainId];
    if (!supportedChain) {
        return;
    }
    if (!supportedChain.blockscoutUrl) {
        return;
    }
    return `${supportedChain.blockscoutUrl}/${subtype}/${hash}`;
}
exports.getChainBlockscoutUrl = getChainBlockscoutUrl;
