// @sentio/chain/dist/chain-id.js
"use strict";
// copy from https://github.com/DefiLlama/chainlist/blob/main/constants/chainIds.js
// and https://besu.hyperledger.org/en/stable/Concepts/NetworkID-And-ChainID/
Object.defineProperty(exports, "__esModule", { value: true });
exports.getChainType = exports.ChainId = exports.BTCChainId = exports.StarknetChainId = exports.CosmosChainId = exports.FuelChainId = exports.SolanaChainId = exports.SuiChainId = exports.AptosChainId = exports.EthChainId = exports.ChainType = void 0;
var ChainType;
(function (ChainType) {
    ChainType["SOLANA"] = "solana";
    ChainType["SUI"] = "sui";
    ChainType["APTOS"] = "aptos";
    ChainType["ETH"] = "evm";
    ChainType["BTC"] = "btc";
})(ChainType || (exports.ChainType = ChainType = {}));
var EthChainId;
(function (EthChainId) {
    EthChainId["ETHEREUM"] = "1";
    EthChainId["OPTIMISM"] = "10";
    EthChainId["CRONOS"] = "25";
    EthChainId["BINANCE"] = "56";
    EthChainId["POLYGON"] = "137";
    EthChainId["MANTA_PACIFIC"] = "169";
    EthChainId["XLAYER_TESTNET"] = "195";
    EthChainId["XLAYER_MAINNET"] = "196";
    EthChainId["B2_MAINNET"] = "223";
    EthChainId["FANTOM"] = "250";
    EthChainId["FRAXTAL"] = "252";
    EthChainId["KUCOIN"] = "321";
    EthChainId["CRONOS_ZKEVM"] = "388";
    EthChainId["ZKSYNC_ERA"] = "324";
    EthChainId["CRONOS_TESTNET"] = "338";
    EthChainId["ASTAR"] = "592";
    EthChainId["LYRA"] = "957";
    EthChainId["POLYGON_ZKEVM"] = "1101";
    EthChainId["CONFLUX"] = "1030";
    EthChainId["METIS"] = "1088";
    EthChainId["MOONBEAM"] = "1284";
    EthChainId["UNCHAIN_SEPOLIA"] = "1301";
    EthChainId["SONEIUM_TESTNET"] = "1946";
    EthChainId["BEVM"] = "11501";
    EthChainId["MERLIN_MAINNET"] = "4200";
    EthChainId["MANTLE"] = "5000";
    EthChainId["BASE"] = "8453";
    EthChainId["LUMIO_TESTNET"] = "9990";
    EthChainId["HOLESKY"] = "17000";
    EthChainId["ASTAR_ZKEVM"] = "3776";
    EthChainId["MOVEMENT_EVM_TESTNET"] = "30732";
    EthChainId["MODE"] = "34443";
    EthChainId["ARBITRUM"] = "42161";
    EthChainId["AVALANCHE"] = "43114";
    EthChainId["ZIRCUIT_TESTNET"] = "48899";
    EthChainId["ZIRCUIT_MAINNET"] = "48900";
    EthChainId["BOB"] = "60808";
    EthChainId["LINEA"] = "59144";
    EthChainId["SONIC_TESTNET"] = "64165";
    EthChainId["BLAST"] = "81457";
    EthChainId["CHILIZ"] = "88888";
    EthChainId["TAIKO"] = "167000";
    EthChainId["TAIKO_TESTNET"] = "167008";
    EthChainId["BITLAYER"] = "200901";
    EthChainId["SCROLL"] = "534352";
    EthChainId["ZKLINK_NOVA"] = "810180";
    EthChainId["SEPOLIA"] = "11155111";
    EthChainId["AURORA"] = "1313161554";
    EthChainId["BLAST_SEPOLIA"] = "168587773";
})(EthChainId || (exports.EthChainId = EthChainId = {}));
var AptosChainId;
(function (AptosChainId) {
    AptosChainId["APTOS_MAINNET"] = "aptos_mainnet";
    AptosChainId["APTOS_TESTNET"] = "aptos_testnet";
    // APTOS_DEVNET = 'aptos_devnet',
    AptosChainId["APTOS_MOVEMENT_TESTNET"] = "aptos_m2_testnet";
    AptosChainId["APTOS_MOVEMENT_MAINNET"] = "aptos_m2_mainnet";
})(AptosChainId || (exports.AptosChainId = AptosChainId = {}));
var SuiChainId;
(function (SuiChainId) {
    SuiChainId["SUI_MAINNET"] = "sui_mainnet";
    SuiChainId["SUI_TESTNET"] = "sui_testnet";
    // SUI_DEVNET = 'sui_devnet',
    SuiChainId["SUI_MOVEMENT_TESTNET"] = "sui_m2_testnet";
    SuiChainId["SUI_MOVEMENT_MAINNET"] = "sui_m2_mainnet";
})(SuiChainId || (exports.SuiChainId = SuiChainId = {}));
var SolanaChainId;
(function (SolanaChainId) {
    SolanaChainId["SOLANA_MAINNET"] = "sol_mainnet";
    SolanaChainId["SOLANA_DEVNET"] = "sol_devnet";
    SolanaChainId["SOLANA_TESTNET"] = "sol_testnet";
    SolanaChainId["SOLANA_PYTH"] = "sol_pyth";
})(SolanaChainId || (exports.SolanaChainId = SolanaChainId = {}));
var FuelChainId;
(function (FuelChainId) {
    FuelChainId["FUEL_MAINNET"] = "fuel_mainnet";
    FuelChainId["FUEL_TESTNET"] = "fuel_testnet";
})(FuelChainId || (exports.FuelChainId = FuelChainId = {}));
var CosmosChainId;
(function (CosmosChainId) {
    CosmosChainId["INJECTIVE_MAINNET"] = "injective_mainnet";
    CosmosChainId["INJECTIVE_TESTNET"] = "injective_testnet";
})(CosmosChainId || (exports.CosmosChainId = CosmosChainId = {}));
var StarknetChainId;
(function (StarknetChainId) {
    StarknetChainId["STARKNET_MAINNET"] = "starknet_mainnet";
    StarknetChainId["STARKNET_SEPOLIA"] = "starknet_sepolia";
})(StarknetChainId || (exports.StarknetChainId = StarknetChainId = {}));
var BTCChainId;
(function (BTCChainId) {
    BTCChainId["BTC_MAINNET"] = "btc_mainnet";
    BTCChainId["BTC_TESTNET"] = "btc_testnet";
})(BTCChainId || (exports.BTCChainId = BTCChainId = {}));
exports.ChainId = {
    ...EthChainId,
    ...AptosChainId,
    ...SuiChainId,
    ...SolanaChainId,
    ...FuelChainId,
    ...CosmosChainId,
    ...StarknetChainId,
    ...BTCChainId
};
function getChainType(chainId) {
    const id = String(chainId).toLowerCase();
    if (id.startsWith('sol')) {
        return ChainType.SOLANA;
    }
    if (id.startsWith('sui')) {
        return ChainType.SUI;
    }
    if (id.startsWith('apt')) {
        return ChainType.APTOS;
    }
    if (id.startsWith('btc')) {
        return ChainType.BTC;
    }
    return ChainType.ETH;
}
exports.getChainType = getChainType;
