// @sentio/api/dist/src/models/AnalyticServiceLogQueryResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonAnyFromJSON, CommonAnyToJSON, } from './CommonAny.js';
import { CommonEventLogEntryFromJSON, CommonEventLogEntryToJSON, } from './CommonEventLogEntry.js';
/**
 * Check if a given object implements the AnalyticServiceLogQueryResponse interface.
 */
export function instanceOfAnalyticServiceLogQueryResponse(value) {
    return true;
}
export function AnalyticServiceLogQueryResponseFromJSON(json) {
    return AnalyticServiceLogQueryResponseFromJSONTyped(json, false);
}
export function AnalyticServiceLogQueryResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'entries': json['entries'] == null ? undefined : (json['entries'].map(CommonEventLogEntryFromJSON)),
        'after': json['after'] == null ? undefined : (json['after'].map(CommonAnyFromJSON)),
        'total': json['total'] == null ? undefined : json['total'],
    };
}
export function AnalyticServiceLogQueryResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'entries': value['entries'] == null ? undefined : (value['entries'].map(CommonEventLogEntryToJSON)),
        'after': value['after'] == null ? undefined : (value['after'].map(CommonAnyToJSON)),
        'total': value['total'],
    };
}
