// @sentio/api/dist/src/models/AnalyticServiceAnalyticServiceExecuteSQLBody.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { AnalyticServiceSQLQueryFromJSON, AnalyticServiceSQLQueryToJSON, } from './AnalyticServiceSQLQuery.js';
import { CommonCachePolicyFromJSON, CommonCachePolicyToJSON, } from './CommonCachePolicy.js';
/**
 * Check if a given object implements the AnalyticServiceAnalyticServiceExecuteSQLBody interface.
 */
export function instanceOfAnalyticServiceAnalyticServiceExecuteSQLBody(value) {
    return true;
}
export function AnalyticServiceAnalyticServiceExecuteSQLBodyFromJSON(json) {
    return AnalyticServiceAnalyticServiceExecuteSQLBodyFromJSONTyped(json, false);
}
export function AnalyticServiceAnalyticServiceExecuteSQLBodyFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'projectId': json['projectId'] == null ? undefined : json['projectId'],
        'version': json['version'] == null ? undefined : json['version'],
        'sqlQuery': json['sqlQuery'] == null ? undefined : AnalyticServiceSQLQueryFromJSON(json['sqlQuery']),
        'cursor': json['cursor'] == null ? undefined : json['cursor'],
        'bypassCache': json['bypassCache'] == null ? undefined : json['bypassCache'],
        'cachePolicy': json['cachePolicy'] == null ? undefined : CommonCachePolicyFromJSON(json['cachePolicy']),
    };
}
export function AnalyticServiceAnalyticServiceExecuteSQLBodyToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'projectId': value['projectId'],
        'version': value['version'],
        'sqlQuery': AnalyticServiceSQLQueryToJSON(value['sqlQuery']),
        'cursor': value['cursor'],
        'bypassCache': value['bypassCache'],
        'cachePolicy': CommonCachePolicyToJSON(value['cachePolicy']),
    };
}
