// @sentio/api/dist/src/models/AnalyticServiceLogQueryRequestSort.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the AnalyticServiceLogQueryRequestSort interface.
 */
export function instanceOfAnalyticServiceLogQueryRequestSort(value) {
    return true;
}
export function AnalyticServiceLogQueryRequestSortFromJSON(json) {
    return AnalyticServiceLogQueryRequestSortFromJSONTyped(json, false);
}
export function AnalyticServiceLogQueryRequestSortFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'field': json['field'] == null ? undefined : json['field'],
        'desc': json['desc'] == null ? undefined : json['desc'],
    };
}
export function AnalyticServiceLogQueryRequestSortToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'field': value['field'],
        'desc': value['desc'],
    };
}
