// @sentio/api/dist/src/models/AnalyticServiceSyncExecuteSQLResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonComputeStatsFromJSON, CommonComputeStatsToJSON, } from './CommonComputeStats.js';
import { CommonTabularDataFromJSON, CommonTabularDataToJSON, } from './CommonTabularData.js';
/**
 * Check if a given object implements the AnalyticServiceSyncExecuteSQLResponse interface.
 */
export function instanceOfAnalyticServiceSyncExecuteSQLResponse(value) {
    return true;
}
export function AnalyticServiceSyncExecuteSQLResponseFromJSON(json) {
    return AnalyticServiceSyncExecuteSQLResponseFromJSONTyped(json, false);
}
export function AnalyticServiceSyncExecuteSQLResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'runtimeCost': json['runtimeCost'] == null ? undefined : json['runtimeCost'],
        'result': json['result'] == null ? undefined : CommonTabularDataFromJSON(json['result']),
        'error': json['error'] == null ? undefined : json['error'],
        'computeStats': json['computeStats'] == null ? undefined : CommonComputeStatsFromJSON(json['computeStats']),
    };
}
export function AnalyticServiceSyncExecuteSQLResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'runtimeCost': value['runtimeCost'],
        'result': CommonTabularDataToJSON(value['result']),
        'error': value['error'],
        'computeStats': CommonComputeStatsToJSON(value['computeStats']),
    };
}
