// @sentio/api/dist/src/models/AnalyticServiceSQLQuery.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the AnalyticServiceSQLQuery interface.
 */
export function instanceOfAnalyticServiceSQLQuery(value) {
    return true;
}
export function AnalyticServiceSQLQueryFromJSON(json) {
    return AnalyticServiceSQLQueryFromJSONTyped(json, false);
}
export function AnalyticServiceSQLQueryFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'sql': json['sql'] == null ? undefined : json['sql'],
        'size': json['size'] == null ? undefined : json['size'],
    };
}
export function AnalyticServiceSQLQueryToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'sql': value['sql'],
        'size': value['size'],
    };
}
