// @sentio/api/dist/src/models/AnalyticServiceSearchServiceQueryLogBody.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { AnalyticServiceLogQueryRequestFilterFromJSON, AnalyticServiceLogQueryRequestFilterToJSON, } from './AnalyticServiceLogQueryRequestFilter.js';
import { AnalyticServiceLogQueryRequestSortFromJSON, AnalyticServiceLogQueryRequestSortToJSON, } from './AnalyticServiceLogQueryRequestSort.js';
import { CommonAnyFromJSON, CommonAnyToJSON, } from './CommonAny.js';
import { CommonTimeRangeFromJSON, CommonTimeRangeToJSON, } from './CommonTimeRange.js';
/**
 * Check if a given object implements the AnalyticServiceSearchServiceQueryLogBody interface.
 */
export function instanceOfAnalyticServiceSearchServiceQueryLogBody(value) {
    return true;
}
export function AnalyticServiceSearchServiceQueryLogBodyFromJSON(json) {
    return AnalyticServiceSearchServiceQueryLogBodyFromJSONTyped(json, false);
}
export function AnalyticServiceSearchServiceQueryLogBodyFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'projectId': json['projectId'] == null ? undefined : json['projectId'],
        'query': json['query'] == null ? undefined : json['query'],
        'timeRange': json['timeRange'] == null ? undefined : CommonTimeRangeFromJSON(json['timeRange']),
        'sorts': json['sorts'] == null ? undefined : (json['sorts'].map(AnalyticServiceLogQueryRequestSortFromJSON)),
        'after': json['after'] == null ? undefined : (json['after'].map(CommonAnyFromJSON)),
        'limit': json['limit'] == null ? undefined : json['limit'],
        'offset': json['offset'] == null ? undefined : json['offset'],
        'filters': json['filters'] == null ? undefined : (json['filters'].map(AnalyticServiceLogQueryRequestFilterFromJSON)),
        'version': json['version'] == null ? undefined : json['version'],
    };
}
export function AnalyticServiceSearchServiceQueryLogBodyToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'projectId': value['projectId'],
        'query': value['query'],
        'timeRange': CommonTimeRangeToJSON(value['timeRange']),
        'sorts': value['sorts'] == null ? undefined : (value['sorts'].map(AnalyticServiceLogQueryRequestSortToJSON)),
        'after': value['after'] == null ? undefined : (value['after'].map(CommonAnyToJSON)),
        'limit': value['limit'],
        'offset': value['offset'],
        'filters': value['filters'] == null ? undefined : (value['filters'].map(AnalyticServiceLogQueryRequestFilterToJSON)),
        'version': value['version'],
    };
}
