// @sentio/api/dist/src/models/CommonUserInfo.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the CommonUserInfo interface.
 */
export function instanceOfCommonUserInfo(value) {
    return true;
}
export function CommonUserInfoFromJSON(json) {
    return CommonUserInfoFromJSONTyped(json, false);
}
export function CommonUserInfoFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'id': json['id'] == null ? undefined : json['id'],
        'lastName': json['lastName'] == null ? undefined : json['lastName'],
        'firstName': json['firstName'] == null ? undefined : json['firstName'],
        'nickname': json['nickname'] == null ? undefined : json['nickname'],
        'picture': json['picture'] == null ? undefined : json['picture'],
        'username': json['username'] == null ? undefined : json['username'],
    };
}
export function CommonUserInfoToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'id': value['id'],
        'lastName': value['lastName'],
        'firstName': value['firstName'],
        'nickname': value['nickname'],
        'picture': value['picture'],
        'username': value['username'],
    };
}
