// @sentio/api/dist/src/models/CommonSegmentationQuerySelectorExprLogicExpr.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonJoinOperatorFromJSON, CommonJoinOperatorToJSON, } from './CommonJoinOperator.js';
import { CommonSegmentationQuerySelectorExprFromJSON, CommonSegmentationQuerySelectorExprToJSON, } from './CommonSegmentationQuerySelectorExpr.js';
/**
 * Check if a given object implements the CommonSegmentationQuerySelectorExprLogicExpr interface.
 */
export function instanceOfCommonSegmentationQuerySelectorExprLogicExpr(value) {
    return true;
}
export function CommonSegmentationQuerySelectorExprLogicExprFromJSON(json) {
    return CommonSegmentationQuerySelectorExprLogicExprFromJSONTyped(json, false);
}
export function CommonSegmentationQuerySelectorExprLogicExprFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'expressions': json['expressions'] == null ? undefined : (json['expressions'].map(CommonSegmentationQuerySelectorExprFromJSON)),
        'operator': json['operator'] == null ? undefined : CommonJoinOperatorFromJSON(json['operator']),
    };
}
export function CommonSegmentationQuerySelectorExprLogicExprToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'expressions': value['expressions'] == null ? undefined : (value['expressions'].map(CommonSegmentationQuerySelectorExprToJSON)),
        'operator': CommonJoinOperatorToJSON(value['operator']),
    };
}
