// @sentio/api/dist/src/models/CommonRetentionQueryResource.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonRetentionQueryFilterFromJSON, CommonRetentionQueryFilterToJSON, } from './CommonRetentionQueryFilter.js';
/**
 * Check if a given object implements the CommonRetentionQueryResource interface.
 */
export function instanceOfCommonRetentionQueryResource(value) {
    return true;
}
export function CommonRetentionQueryResourceFromJSON(json) {
    return CommonRetentionQueryResourceFromJSONTyped(json, false);
}
export function CommonRetentionQueryResourceFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'eventNames': json['eventNames'] == null ? undefined : json['eventNames'],
        'filter': json['filter'] == null ? undefined : CommonRetentionQueryFilterFromJSON(json['filter']),
    };
}
export function CommonRetentionQueryResourceToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'eventNames': value['eventNames'],
        'filter': CommonRetentionQueryFilterToJSON(value['filter']),
    };
}
