// @sentio/api/dist/src/models/CommonMatrix.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonMatrixSampleFromJSON, CommonMatrixSampleToJSON, } from './CommonMatrixSample.js';
/**
 * Check if a given object implements the CommonMatrix interface.
 */
export function instanceOfCommonMatrix(value) {
    return true;
}
export function CommonMatrixFromJSON(json) {
    return CommonMatrixFromJSONTyped(json, false);
}
export function CommonMatrixFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'samples': json['samples'] == null ? undefined : (json['samples'].map(CommonMatrixSampleFromJSON)),
        'totalSamples': json['totalSamples'] == null ? undefined : json['totalSamples'],
    };
}
export function CommonMatrixToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'samples': value['samples'] == null ? undefined : (value['samples'].map(CommonMatrixSampleToJSON)),
        'totalSamples': value['totalSamples'],
    };
}
