// @sentio/api/dist/src/models/CommonColumnState.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonColumnStateSortFromJSON, CommonColumnStateSortToJSON, } from './CommonColumnStateSort.js';
/**
 * Check if a given object implements the CommonColumnState interface.
 */
export function instanceOfCommonColumnState(value) {
    return true;
}
export function CommonColumnStateFromJSON(json) {
    return CommonColumnStateFromJSONTyped(json, false);
}
export function CommonColumnStateFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'columnSizing': json['columnSizing'] == null ? undefined : json['columnSizing'],
        'columnVisibility': json['columnVisibility'] == null ? undefined : json['columnVisibility'],
        'columnOrder': json['columnOrder'] == null ? undefined : json['columnOrder'],
        'sorting': json['sorting'] == null ? undefined : (json['sorting'].map(CommonColumnStateSortFromJSON)),
    };
}
export function CommonColumnStateToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'columnSizing': value['columnSizing'],
        'columnVisibility': value['columnVisibility'],
        'columnOrder': value['columnOrder'],
        'sorting': value['sorting'] == null ? undefined : (value['sorting'].map(CommonColumnStateSortToJSON)),
    };
}
