// @sentio/api/dist/src/models/CommonCoinIDAddressIdentifier.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the CommonCoinIDAddressIdentifier interface.
 */
export function instanceOfCommonCoinIDAddressIdentifier(value) {
    return true;
}
export function CommonCoinIDAddressIdentifierFromJSON(json) {
    return CommonCoinIDAddressIdentifierFromJSONTyped(json, false);
}
export function CommonCoinIDAddressIdentifierFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'address': json['address'] == null ? undefined : json['address'],
        'chain': json['chain'] == null ? undefined : json['chain'],
    };
}
export function CommonCoinIDAddressIdentifierToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'address': value['address'],
        'chain': value['chain'],
    };
}
