// @sentio/api/dist/src/models/CommonFunction.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonArgumentFromJSON, CommonArgumentToJSON, } from './CommonArgument.js';
/**
 * Check if a given object implements the CommonFunction interface.
 */
export function instanceOfCommonFunction(value) {
    return true;
}
export function CommonFunctionFromJSON(json) {
    return CommonFunctionFromJSONTyped(json, false);
}
export function CommonFunctionFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'name': json['name'] == null ? undefined : json['name'],
        'arguments': json['arguments'] == null ? undefined : (json['arguments'].map(CommonArgumentFromJSON)),
    };
}
export function CommonFunctionToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'name': value['name'],
        'arguments': value['arguments'] == null ? undefined : (value['arguments'].map(CommonArgumentToJSON)),
    };
}
