// @sentio/api/dist/src/models/CommonSegmentationQuery.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonFunctionFromJSON, CommonFunctionToJSON, } from './CommonFunction.js';
import { CommonSegmentationQueryAggregationFromJSON, CommonSegmentationQueryAggregationToJSON, } from './CommonSegmentationQueryAggregation.js';
import { CommonSegmentationQueryResourceFromJSON, CommonSegmentationQueryResourceToJSON, } from './CommonSegmentationQueryResource.js';
import { CommonSegmentationQuerySelectorExprFromJSON, CommonSegmentationQuerySelectorExprToJSON, } from './CommonSegmentationQuerySelectorExpr.js';
/**
 * Check if a given object implements the CommonSegmentationQuery interface.
 */
export function instanceOfCommonSegmentationQuery(value) {
    return true;
}
export function CommonSegmentationQueryFromJSON(json) {
    return CommonSegmentationQueryFromJSONTyped(json, false);
}
export function CommonSegmentationQueryFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'resource': json['resource'] == null ? undefined : CommonSegmentationQueryResourceFromJSON(json['resource']),
        'alias': json['alias'] == null ? undefined : json['alias'],
        'id': json['id'] == null ? undefined : json['id'],
        'aggregation': json['aggregation'] == null ? undefined : CommonSegmentationQueryAggregationFromJSON(json['aggregation']),
        'selectorExpr': json['selectorExpr'] == null ? undefined : CommonSegmentationQuerySelectorExprFromJSON(json['selectorExpr']),
        'groupBy': json['groupBy'] == null ? undefined : json['groupBy'],
        'limit': json['limit'] == null ? undefined : json['limit'],
        'functions': json['functions'] == null ? undefined : (json['functions'].map(CommonFunctionFromJSON)),
        'disabled': json['disabled'] == null ? undefined : json['disabled'],
    };
}
export function CommonSegmentationQueryToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'resource': CommonSegmentationQueryResourceToJSON(value['resource']),
        'alias': value['alias'],
        'id': value['id'],
        'aggregation': CommonSegmentationQueryAggregationToJSON(value['aggregation']),
        'selectorExpr': CommonSegmentationQuerySelectorExprToJSON(value['selectorExpr']),
        'groupBy': value['groupBy'],
        'limit': value['limit'],
        'functions': value['functions'] == null ? undefined : (value['functions'].map(CommonFunctionToJSON)),
        'disabled': value['disabled'],
    };
}
