// @sentio/api/dist/src/models/CommonRetentionQueryInterval.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonRetentionQueryIntervalUnitFromJSON, CommonRetentionQueryIntervalUnitToJSON, } from './CommonRetentionQueryIntervalUnit.js';
/**
 * Check if a given object implements the CommonRetentionQueryInterval interface.
 */
export function instanceOfCommonRetentionQueryInterval(value) {
    return true;
}
export function CommonRetentionQueryIntervalFromJSON(json) {
    return CommonRetentionQueryIntervalFromJSONTyped(json, false);
}
export function CommonRetentionQueryIntervalFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'value': json['value'] == null ? undefined : json['value'],
        'unit': json['unit'] == null ? undefined : CommonRetentionQueryIntervalUnitFromJSON(json['unit']),
    };
}
export function CommonRetentionQueryIntervalToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'value': value['value'],
        'unit': CommonRetentionQueryIntervalUnitToJSON(value['unit']),
    };
}
