// @sentio/api/dist/src/models/CommonCohortsFilterAggregationAggregateProperties.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonCohortsFilterAggregationAggregatePropertiesAggregationTypeFromJSON, CommonCohortsFilterAggregationAggregatePropertiesAggregationTypeToJSON, } from './CommonCohortsFilterAggregationAggregatePropertiesAggregationType.js';
/**
 * Check if a given object implements the CommonCohortsFilterAggregationAggregateProperties interface.
 */
export function instanceOfCommonCohortsFilterAggregationAggregateProperties(value) {
    return true;
}
export function CommonCohortsFilterAggregationAggregatePropertiesFromJSON(json) {
    return CommonCohortsFilterAggregationAggregatePropertiesFromJSONTyped(json, false);
}
export function CommonCohortsFilterAggregationAggregatePropertiesFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'type': json['type'] == null ? undefined : CommonCohortsFilterAggregationAggregatePropertiesAggregationTypeFromJSON(json['type']),
        'propertyName': json['propertyName'] == null ? undefined : json['propertyName'],
    };
}
export function CommonCohortsFilterAggregationAggregatePropertiesToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'type': CommonCohortsFilterAggregationAggregatePropertiesAggregationTypeToJSON(value['type']),
        'propertyName': value['propertyName'],
    };
}
