// @sentio/api/dist/src/models/CommonProjectSuperset.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the CommonProjectSuperset interface.
 */
export function instanceOfCommonProjectSuperset(value) {
    return true;
}
export function CommonProjectSupersetFromJSON(json) {
    return CommonProjectSupersetFromJSONTyped(json, false);
}
export function CommonProjectSupersetFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'projectId': json['projectId'] == null ? undefined : json['projectId'],
        'createdAt': json['createdAt'] == null ? undefined : (new Date(json['createdAt'])),
        'syncAt': json['syncAt'] == null ? undefined : (new Date(json['syncAt'])),
    };
}
export function CommonProjectSupersetToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'projectId': value['projectId'],
        'createdAt': value['createdAt'] == null ? undefined : ((value['createdAt']).toISOString()),
        'syncAt': value['syncAt'] == null ? undefined : ((value['syncAt']).toISOString()),
    };
}
