// @sentio/api/dist/src/models/CommonSegmentParameter.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the CommonSegmentParameter interface.
 */
export function instanceOfCommonSegmentParameter(value) {
    return true;
}
export function CommonSegmentParameterFromJSON(json) {
    return CommonSegmentParameterFromJSONTyped(json, false);
}
export function CommonSegmentParameterFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'cohortId': json['cohortId'] == null ? undefined : json['cohortId'],
        'allUsers': json['allUsers'] == null ? undefined : json['allUsers'],
    };
}
export function CommonSegmentParameterToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'cohortId': value['cohortId'],
        'allUsers': value['allUsers'],
    };
}
