// @sentio/api/dist/src/models/CommonCachePolicy.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the CommonCachePolicy interface.
 */
export function instanceOfCommonCachePolicy(value) {
    return true;
}
export function CommonCachePolicyFromJSON(json) {
    return CommonCachePolicyFromJSONTyped(json, false);
}
export function CommonCachePolicyFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'cacheTtlSecs': json['cacheTtlSecs'] == null ? undefined : json['cacheTtlSecs'],
        'cacheRefreshTtlSecs': json['cacheRefreshTtlSecs'] == null ? undefined : json['cacheRefreshTtlSecs'],
        'forceRefresh': json['forceRefresh'] == null ? undefined : json['forceRefresh'],
        'noCache': json['noCache'] == null ? undefined : json['noCache'],
    };
}
export function CommonCachePolicyToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'cacheTtlSecs': value['cacheTtlSecs'],
        'cacheRefreshTtlSecs': value['cacheRefreshTtlSecs'],
        'forceRefresh': value['forceRefresh'],
        'noCache': value['noCache'],
    };
}
