// @sentio/api/dist/src/models/CommonOwner.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonOrganizationFromJSON, CommonOrganizationToJSON, } from './CommonOrganization.js';
import { CommonTierFromJSON, CommonTierToJSON, } from './CommonTier.js';
import { CommonUserFromJSON, CommonUserToJSON, } from './CommonUser.js';
/**
 * Check if a given object implements the CommonOwner interface.
 */
export function instanceOfCommonOwner(value) {
    return true;
}
export function CommonOwnerFromJSON(json) {
    return CommonOwnerFromJSONTyped(json, false);
}
export function CommonOwnerFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'user': json['user'] == null ? undefined : CommonUserFromJSON(json['user']),
        'organization': json['organization'] == null ? undefined : CommonOrganizationFromJSON(json['organization']),
        'tier': json['tier'] == null ? undefined : CommonTierFromJSON(json['tier']),
    };
}
export function CommonOwnerToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'user': CommonUserToJSON(value['user']),
        'organization': CommonOrganizationToJSON(value['organization']),
        'tier': CommonTierToJSON(value['tier']),
    };
}
