// @sentio/api/dist/src/models/CommonDuration.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the CommonDuration interface.
 */
export function instanceOfCommonDuration(value) {
    return true;
}
export function CommonDurationFromJSON(json) {
    return CommonDurationFromJSONTyped(json, false);
}
export function CommonDurationFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'value': json['value'] == null ? undefined : json['value'],
        'unit': json['unit'] == null ? undefined : json['unit'],
    };
}
export function CommonDurationToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'value': value['value'],
        'unit': value['unit'],
    };
}
