// @sentio/api/dist/src/models/CommonSelector.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonAnyFromJSON, CommonAnyToJSON, } from './CommonAny.js';
import { CommonSelectorOperatorTypeFromJSON, CommonSelectorOperatorTypeToJSON, } from './CommonSelectorOperatorType.js';
/**
 * Check if a given object implements the CommonSelector interface.
 */
export function instanceOfCommonSelector(value) {
    return true;
}
export function CommonSelectorFromJSON(json) {
    return CommonSelectorFromJSONTyped(json, false);
}
export function CommonSelectorFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'key': json['key'] == null ? undefined : json['key'],
        'operator': json['operator'] == null ? undefined : CommonSelectorOperatorTypeFromJSON(json['operator']),
        'value': json['value'] == null ? undefined : (json['value'].map(CommonAnyFromJSON)),
    };
}
export function CommonSelectorToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'key': value['key'],
        'operator': CommonSelectorOperatorTypeToJSON(value['operator']),
        'value': value['value'] == null ? undefined : (value['value'].map(CommonAnyToJSON)),
    };
}
