// @sentio/api/dist/src/models/CommonSelectorOperatorType.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * - GT: Numeric operators
 *  - CONTAINS: String operators
 * @export
 */
export const CommonSelectorOperatorType = {
    Eq: 'EQ',
    Neq: 'NEQ',
    Exists: 'EXISTS',
    NotExists: 'NOT_EXISTS',
    Gt: 'GT',
    Gte: 'GTE',
    Lt: 'LT',
    Lte: 'LTE',
    Between: 'BETWEEN',
    NotBetween: 'NOT_BETWEEN',
    Contains: 'CONTAINS',
    NotContains: 'NOT_CONTAINS',
    InCohorts: 'IN_COHORTS',
    NotInCohorts: 'NOT_IN_COHORTS'
};
export function instanceOfCommonSelectorOperatorType(value) {
    return Object.values(CommonSelectorOperatorType).includes(value);
}
export function CommonSelectorOperatorTypeFromJSON(json) {
    return CommonSelectorOperatorTypeFromJSONTyped(json, false);
}
export function CommonSelectorOperatorTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
export function CommonSelectorOperatorTypeToJSON(value) {
    return value;
}
