// @sentio/api/dist/src/models/CommonOrganization.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonOrganizationMemberFromJSON, CommonOrganizationMemberToJSON, } from './CommonOrganizationMember.js';
import { CommonProjectInfoFromJSON, CommonProjectInfoToJSON, } from './CommonProjectInfo.js';
import { CommonTierFromJSON, CommonTierToJSON, } from './CommonTier.js';
/**
 * Check if a given object implements the CommonOrganization interface.
 */
export function instanceOfCommonOrganization(value) {
    return true;
}
export function CommonOrganizationFromJSON(json) {
    return CommonOrganizationFromJSONTyped(json, false);
}
export function CommonOrganizationFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'id': json['id'] == null ? undefined : json['id'],
        'oid': json['oid'] == null ? undefined : json['oid'],
        'name': json['name'] == null ? undefined : json['name'],
        'createdAt': json['createdAt'] == null ? undefined : json['createdAt'],
        'updatedAt': json['updatedAt'] == null ? undefined : json['updatedAt'],
        'members': json['members'] == null ? undefined : (json['members'].map(CommonOrganizationMemberFromJSON)),
        'displayName': json['displayName'] == null ? undefined : json['displayName'],
        'logoUrl': json['logoUrl'] == null ? undefined : json['logoUrl'],
        'projects': json['projects'] == null ? undefined : (json['projects'].map(CommonProjectInfoFromJSON)),
        'tier': json['tier'] == null ? undefined : CommonTierFromJSON(json['tier']),
    };
}
export function CommonOrganizationToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'id': value['id'],
        'oid': value['oid'],
        'name': value['name'],
        'createdAt': value['createdAt'],
        'updatedAt': value['updatedAt'],
        'members': value['members'] == null ? undefined : (value['members'].map(CommonOrganizationMemberToJSON)),
        'displayName': value['displayName'],
        'logoUrl': value['logoUrl'],
        'projects': value['projects'] == null ? undefined : (value['projects'].map(CommonProjectInfoToJSON)),
        'tier': CommonTierToJSON(value['tier']),
    };
}
