// @sentio/api/dist/src/models/CommonSegmentationQueryAggregation.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonSegmentationQueryAggregationAggregatePropertiesFromJSON, CommonSegmentationQueryAggregationAggregatePropertiesToJSON, } from './CommonSegmentationQueryAggregationAggregateProperties.js';
import { CommonSegmentationQueryAggregationCountUniqueFromJSON, CommonSegmentationQueryAggregationCountUniqueToJSON, } from './CommonSegmentationQueryAggregationCountUnique.js';
/**
 * Check if a given object implements the CommonSegmentationQueryAggregation interface.
 */
export function instanceOfCommonSegmentationQueryAggregation(value) {
    return true;
}
export function CommonSegmentationQueryAggregationFromJSON(json) {
    return CommonSegmentationQueryAggregationFromJSONTyped(json, false);
}
export function CommonSegmentationQueryAggregationFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'total': json['total'] == null ? undefined : json['total'],
        'unique': json['unique'] == null ? undefined : json['unique'],
        'countUnique': json['countUnique'] == null ? undefined : CommonSegmentationQueryAggregationCountUniqueFromJSON(json['countUnique']),
        'aggregateProperties': json['aggregateProperties'] == null ? undefined : CommonSegmentationQueryAggregationAggregatePropertiesFromJSON(json['aggregateProperties']),
    };
}
export function CommonSegmentationQueryAggregationToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'total': value['total'],
        'unique': value['unique'],
        'countUnique': CommonSegmentationQueryAggregationCountUniqueToJSON(value['countUnique']),
        'aggregateProperties': CommonSegmentationQueryAggregationAggregatePropertiesToJSON(value['aggregateProperties']),
    };
}
