// @sentio/api/dist/src/models/CommonCohortsFilter.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonCohortsFilterAggregationFromJSON, CommonCohortsFilterAggregationToJSON, } from './CommonCohortsFilterAggregation.js';
import { CommonSelectorExprFromJSON, CommonSelectorExprToJSON, } from './CommonSelectorExpr.js';
import { CommonTimeRangeLiteFromJSON, CommonTimeRangeLiteToJSON, } from './CommonTimeRangeLite.js';
/**
 * Check if a given object implements the CommonCohortsFilter interface.
 */
export function instanceOfCommonCohortsFilter(value) {
    return true;
}
export function CommonCohortsFilterFromJSON(json) {
    return CommonCohortsFilterFromJSONTyped(json, false);
}
export function CommonCohortsFilterFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'symbol': json['symbol'] == null ? undefined : json['symbol'],
        'name': json['name'] == null ? undefined : json['name'],
        'aggregation': json['aggregation'] == null ? undefined : CommonCohortsFilterAggregationFromJSON(json['aggregation']),
        'selectorExpr': json['selectorExpr'] == null ? undefined : CommonSelectorExprFromJSON(json['selectorExpr']),
        'timeRange': json['timeRange'] == null ? undefined : CommonTimeRangeLiteFromJSON(json['timeRange']),
    };
}
export function CommonCohortsFilterToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'symbol': value['symbol'],
        'name': value['name'],
        'aggregation': CommonCohortsFilterAggregationToJSON(value['aggregation']),
        'selectorExpr': CommonSelectorExprToJSON(value['selectorExpr']),
        'timeRange': CommonTimeRangeLiteToJSON(value['timeRange']),
    };
}
