// @sentio/api/dist/src/models/CommonCohortsFilterAggregation.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonAnyFromJSON, CommonAnyToJSON, } from './CommonAny.js';
import { CommonCohortsFilterAggregationAggregatePropertiesFromJSON, CommonCohortsFilterAggregationAggregatePropertiesToJSON, } from './CommonCohortsFilterAggregationAggregateProperties.js';
import { CommonCohortsFilterAggregationOperatorTypeFromJSON, CommonCohortsFilterAggregationOperatorTypeToJSON, } from './CommonCohortsFilterAggregationOperatorType.js';
/**
 * Check if a given object implements the CommonCohortsFilterAggregation interface.
 */
export function instanceOfCommonCohortsFilterAggregation(value) {
    return true;
}
export function CommonCohortsFilterAggregationFromJSON(json) {
    return CommonCohortsFilterAggregationFromJSONTyped(json, false);
}
export function CommonCohortsFilterAggregationFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'total': json['total'] == null ? undefined : json['total'],
        'aggregateProperties': json['aggregateProperties'] == null ? undefined : CommonCohortsFilterAggregationAggregatePropertiesFromJSON(json['aggregateProperties']),
        'operator': json['operator'] == null ? undefined : CommonCohortsFilterAggregationOperatorTypeFromJSON(json['operator']),
        'value': json['value'] == null ? undefined : (json['value'].map(CommonAnyFromJSON)),
    };
}
export function CommonCohortsFilterAggregationToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'total': value['total'],
        'aggregateProperties': CommonCohortsFilterAggregationAggregatePropertiesToJSON(value['aggregateProperties']),
        'operator': CommonCohortsFilterAggregationOperatorTypeToJSON(value['operator']),
        'value': value['value'] == null ? undefined : (value['value'].map(CommonAnyToJSON)),
    };
}
