// @sentio/api/dist/src/models/CommonProjectViewProjectViewConfig.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonEventLogConfigFromJSON, CommonEventLogConfigToJSON, } from './CommonEventLogConfig.js';
/**
 * Check if a given object implements the CommonProjectViewProjectViewConfig interface.
 */
export function instanceOfCommonProjectViewProjectViewConfig(value) {
    return true;
}
export function CommonProjectViewProjectViewConfigFromJSON(json) {
    return CommonProjectViewProjectViewConfigFromJSONTyped(json, false);
}
export function CommonProjectViewProjectViewConfigFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'eventLog': json['eventLog'] == null ? undefined : CommonEventLogConfigFromJSON(json['eventLog']),
    };
}
export function CommonProjectViewProjectViewConfigToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'eventLog': CommonEventLogConfigToJSON(value['eventLog']),
    };
}
