// @sentio/api/dist/src/models/CommonAggregate.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonAggregateAggregateOpsFromJSON, CommonAggregateAggregateOpsToJSON, } from './CommonAggregateAggregateOps.js';
/**
 * Check if a given object implements the CommonAggregate interface.
 */
export function instanceOfCommonAggregate(value) {
    return true;
}
export function CommonAggregateFromJSON(json) {
    return CommonAggregateFromJSONTyped(json, false);
}
export function CommonAggregateFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'op': json['op'] == null ? undefined : CommonAggregateAggregateOpsFromJSON(json['op']),
        'grouping': json['grouping'] == null ? undefined : json['grouping'],
    };
}
export function CommonAggregateToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'op': CommonAggregateAggregateOpsToJSON(value['op']),
        'grouping': value['grouping'],
    };
}
