// @sentio/api/dist/src/models/CommonCohortsFilterAggregationAggregatePropertiesAggregationType.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 */
export const CommonCohortsFilterAggregationAggregatePropertiesAggregationType = {
    Sum: 'SUM',
    Avg: 'AVG',
    Median: 'MEDIAN',
    Min: 'MIN',
    Max: 'MAX',
    DistinctCount: 'DISTINCT_COUNT',
    Last: 'LAST',
    First: 'FIRST'
};
export function instanceOfCommonCohortsFilterAggregationAggregatePropertiesAggregationType(value) {
    return Object.values(CommonCohortsFilterAggregationAggregatePropertiesAggregationType).includes(value);
}
export function CommonCohortsFilterAggregationAggregatePropertiesAggregationTypeFromJSON(json) {
    return CommonCohortsFilterAggregationAggregatePropertiesAggregationTypeFromJSONTyped(json, false);
}
export function CommonCohortsFilterAggregationAggregatePropertiesAggregationTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
export function CommonCohortsFilterAggregationAggregatePropertiesAggregationTypeToJSON(value) {
    return value;
}
