// @sentio/api/dist/src/models/CommonSegmentationQueryAggregationCountUnique.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonDurationFromJSON, CommonDurationToJSON, } from './CommonDuration.js';
/**
 * Check if a given object implements the CommonSegmentationQueryAggregationCountUnique interface.
 */
export function instanceOfCommonSegmentationQueryAggregationCountUnique(value) {
    return true;
}
export function CommonSegmentationQueryAggregationCountUniqueFromJSON(json) {
    return CommonSegmentationQueryAggregationCountUniqueFromJSONTyped(json, false);
}
export function CommonSegmentationQueryAggregationCountUniqueFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'duration': json['duration'] == null ? undefined : CommonDurationFromJSON(json['duration']),
    };
}
export function CommonSegmentationQueryAggregationCountUniqueToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'duration': CommonDurationToJSON(value['duration']),
    };
}
