// @sentio/api/dist/src/models/CommonPriceSegmentationQuery.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonCoinIDFromJSON, CommonCoinIDToJSON, } from './CommonCoinID.js';
/**
 * Check if a given object implements the CommonPriceSegmentationQuery interface.
 */
export function instanceOfCommonPriceSegmentationQuery(value) {
    return true;
}
export function CommonPriceSegmentationQueryFromJSON(json) {
    return CommonPriceSegmentationQueryFromJSONTyped(json, false);
}
export function CommonPriceSegmentationQueryFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'id': json['id'] == null ? undefined : json['id'],
        'alias': json['alias'] == null ? undefined : json['alias'],
        'coinId': json['coinId'] == null ? undefined : (json['coinId'].map(CommonCoinIDFromJSON)),
        'disabled': json['disabled'] == null ? undefined : json['disabled'],
    };
}
export function CommonPriceSegmentationQueryToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'id': value['id'],
        'alias': value['alias'],
        'coinId': value['coinId'] == null ? undefined : (value['coinId'].map(CommonCoinIDToJSON)),
        'disabled': value['disabled'],
    };
}
