// @sentio/api/dist/src/models/CommonEventLogConfig.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonColumnStateFromJSON, CommonColumnStateToJSON, } from './CommonColumnState.js';
import { CommonEventLogColumnFromJSON, CommonEventLogColumnToJSON, } from './CommonEventLogColumn.js';
/**
 * Check if a given object implements the CommonEventLogConfig interface.
 */
export function instanceOfCommonEventLogConfig(value) {
    return true;
}
export function CommonEventLogConfigFromJSON(json) {
    return CommonEventLogConfigFromJSONTyped(json, false);
}
export function CommonEventLogConfigFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'columns': json['columns'] == null ? undefined : (json['columns'].map(CommonEventLogColumnFromJSON)),
        'state': json['state'] == null ? undefined : CommonColumnStateFromJSON(json['state']),
    };
}
export function CommonEventLogConfigToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'columns': value['columns'] == null ? undefined : (value['columns'].map(CommonEventLogColumnToJSON)),
        'state': CommonColumnStateToJSON(value['state']),
    };
}
