// @sentio/api/dist/src/models/CommonSegmentationQuerySelectorExpr.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonSegmentationQuerySelectorExprLogicExprFromJSON, CommonSegmentationQuerySelectorExprLogicExprToJSON, } from './CommonSegmentationQuerySelectorExprLogicExpr.js';
import { CommonSelectorFromJSON, CommonSelectorToJSON, } from './CommonSelector.js';
/**
 * Check if a given object implements the CommonSegmentationQuerySelectorExpr interface.
 */
export function instanceOfCommonSegmentationQuerySelectorExpr(value) {
    return true;
}
export function CommonSegmentationQuerySelectorExprFromJSON(json) {
    return CommonSegmentationQuerySelectorExprFromJSONTyped(json, false);
}
export function CommonSegmentationQuerySelectorExprFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'selector': json['selector'] == null ? undefined : CommonSelectorFromJSON(json['selector']),
        'logicExpr': json['logicExpr'] == null ? undefined : CommonSegmentationQuerySelectorExprLogicExprFromJSON(json['logicExpr']),
    };
}
export function CommonSegmentationQuerySelectorExprToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'selector': CommonSelectorToJSON(value['selector']),
        'logicExpr': CommonSegmentationQuerySelectorExprLogicExprToJSON(value['logicExpr']),
    };
}
