// @sentio/api/dist/src/models/CommonProjectView.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonProjectViewProjectViewConfigFromJSON, CommonProjectViewProjectViewConfigToJSON, } from './CommonProjectViewProjectViewConfig.js';
/**
 * Check if a given object implements the CommonProjectView interface.
 */
export function instanceOfCommonProjectView(value) {
    return true;
}
export function CommonProjectViewFromJSON(json) {
    return CommonProjectViewFromJSONTyped(json, false);
}
export function CommonProjectViewFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'id': json['id'] == null ? undefined : json['id'],
        'projectId': json['projectId'] == null ? undefined : json['projectId'],
        'name': json['name'] == null ? undefined : json['name'],
        'config': json['config'] == null ? undefined : CommonProjectViewProjectViewConfigFromJSON(json['config']),
    };
}
export function CommonProjectViewToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'id': value['id'],
        'projectId': value['projectId'],
        'name': value['name'],
        'config': CommonProjectViewProjectViewConfigToJSON(value['config']),
    };
}
