// @sentio/api/dist/src/models/CommonSegmentationQueryAggregationAggregateProperties.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonSegmentationQueryAggregationAggregatePropertiesAggregationTypeFromJSON, CommonSegmentationQueryAggregationAggregatePropertiesAggregationTypeToJSON, } from './CommonSegmentationQueryAggregationAggregatePropertiesAggregationType.js';
/**
 * Check if a given object implements the CommonSegmentationQueryAggregationAggregateProperties interface.
 */
export function instanceOfCommonSegmentationQueryAggregationAggregateProperties(value) {
    return true;
}
export function CommonSegmentationQueryAggregationAggregatePropertiesFromJSON(json) {
    return CommonSegmentationQueryAggregationAggregatePropertiesFromJSONTyped(json, false);
}
export function CommonSegmentationQueryAggregationAggregatePropertiesFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'type': json['type'] == null ? undefined : CommonSegmentationQueryAggregationAggregatePropertiesAggregationTypeFromJSON(json['type']),
        'propertyName': json['propertyName'] == null ? undefined : json['propertyName'],
    };
}
export function CommonSegmentationQueryAggregationAggregatePropertiesToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'type': CommonSegmentationQueryAggregationAggregatePropertiesAggregationTypeToJSON(value['type']),
        'propertyName': value['propertyName'],
    };
}
