// @sentio/api/dist/src/models/CommonRetentionMatrixSample.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonSegmentParameterFromJSON, CommonSegmentParameterToJSON, } from './CommonSegmentParameter.js';
/**
 * Check if a given object implements the CommonRetentionMatrixSample interface.
 */
export function instanceOfCommonRetentionMatrixSample(value) {
    return true;
}
export function CommonRetentionMatrixSampleFromJSON(json) {
    return CommonRetentionMatrixSampleFromJSONTyped(json, false);
}
export function CommonRetentionMatrixSampleFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'time': json['time'] == null ? undefined : (new Date(json['time'])),
        'segmentParameter': json['segmentParameter'] == null ? undefined : CommonSegmentParameterFromJSON(json['segmentParameter']),
        'labels': json['labels'] == null ? undefined : json['labels'],
        'totalCount': json['totalCount'] == null ? undefined : json['totalCount'],
        'counts': json['counts'] == null ? undefined : json['counts'],
        'rates': json['rates'] == null ? undefined : json['rates'],
    };
}
export function CommonRetentionMatrixSampleToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'time': value['time'] == null ? undefined : ((value['time']).toISOString()),
        'segmentParameter': CommonSegmentParameterToJSON(value['segmentParameter']),
        'labels': value['labels'],
        'totalCount': value['totalCount'],
        'counts': value['counts'],
        'rates': value['rates'],
    };
}
