// @sentio/api/dist/src/models/CommonOrganizationMember.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonOrganizationRoleFromJSON, CommonOrganizationRoleToJSON, } from './CommonOrganizationRole.js';
import { CommonUserInfoFromJSON, CommonUserInfoToJSON, } from './CommonUserInfo.js';
/**
 * Check if a given object implements the CommonOrganizationMember interface.
 */
export function instanceOfCommonOrganizationMember(value) {
    return true;
}
export function CommonOrganizationMemberFromJSON(json) {
    return CommonOrganizationMemberFromJSONTyped(json, false);
}
export function CommonOrganizationMemberFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'user': json['user'] == null ? undefined : CommonUserInfoFromJSON(json['user']),
        'role': json['role'] == null ? undefined : CommonOrganizationRoleFromJSON(json['role']),
    };
}
export function CommonOrganizationMemberToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'user': CommonUserInfoToJSON(value['user']),
        'role': CommonOrganizationRoleToJSON(value['role']),
    };
}
