// @sentio/api/dist/src/models/CommonSelectorExpr.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonSelectorFromJSON, CommonSelectorToJSON, } from './CommonSelector.js';
import { CommonSelectorExprLogicExprFromJSON, CommonSelectorExprLogicExprToJSON, } from './CommonSelectorExprLogicExpr.js';
/**
 * Check if a given object implements the CommonSelectorExpr interface.
 */
export function instanceOfCommonSelectorExpr(value) {
    return true;
}
export function CommonSelectorExprFromJSON(json) {
    return CommonSelectorExprFromJSONTyped(json, false);
}
export function CommonSelectorExprFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'selector': json['selector'] == null ? undefined : CommonSelectorFromJSON(json['selector']),
        'logicExpr': json['logicExpr'] == null ? undefined : CommonSelectorExprLogicExprFromJSON(json['logicExpr']),
    };
}
export function CommonSelectorExprToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'selector': CommonSelectorToJSON(value['selector']),
        'logicExpr': CommonSelectorExprLogicExprToJSON(value['logicExpr']),
    };
}
