// @sentio/api/dist/src/models/CommonChannel.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonChannelTypeFromJSON, CommonChannelTypeToJSON, } from './CommonChannelType.js';
/**
 * Check if a given object implements the CommonChannel interface.
 */
export function instanceOfCommonChannel(value) {
    return true;
}
export function CommonChannelFromJSON(json) {
    return CommonChannelFromJSONTyped(json, false);
}
export function CommonChannelFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'id': json['id'] == null ? undefined : json['id'],
        'projectId': json['projectId'] == null ? undefined : json['projectId'],
        'type': json['type'] == null ? undefined : CommonChannelTypeFromJSON(json['type']),
        'slackWebhookUrl': json['slackWebhookUrl'] == null ? undefined : json['slackWebhookUrl'],
        'emailAddress': json['emailAddress'] == null ? undefined : json['emailAddress'],
        'name': json['name'] == null ? undefined : json['name'],
        'customWebhookUrl': json['customWebhookUrl'] == null ? undefined : json['customWebhookUrl'],
        'customHeaders': json['customHeaders'] == null ? undefined : json['customHeaders'],
        'telegramReference': json['telegramReference'] == null ? undefined : json['telegramReference'],
        'telegramChatId': json['telegramChatId'] == null ? undefined : json['telegramChatId'],
        'slackTeam': json['slackTeam'] == null ? undefined : json['slackTeam'],
        'slackChannel': json['slackChannel'] == null ? undefined : json['slackChannel'],
        'pagerdutyConfig': json['pagerdutyConfig'] == null ? undefined : json['pagerdutyConfig'],
    };
}
export function CommonChannelToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'id': value['id'],
        'projectId': value['projectId'],
        'type': CommonChannelTypeToJSON(value['type']),
        'slackWebhookUrl': value['slackWebhookUrl'],
        'emailAddress': value['emailAddress'],
        'name': value['name'],
        'customWebhookUrl': value['customWebhookUrl'],
        'customHeaders': value['customHeaders'],
        'telegramReference': value['telegramReference'],
        'telegramChatId': value['telegramChatId'],
        'slackTeam': value['slackTeam'],
        'slackChannel': value['slackChannel'],
        'pagerdutyConfig': value['pagerdutyConfig'],
    };
}
