// @sentio/api/dist/src/models/CommonQuery.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonAggregateFromJSON, CommonAggregateToJSON, } from './CommonAggregate.js';
import { CommonFunctionFromJSON, CommonFunctionToJSON, } from './CommonFunction.js';
/**
 * Check if a given object implements the CommonQuery interface.
 */
export function instanceOfCommonQuery(value) {
    return true;
}
export function CommonQueryFromJSON(json) {
    return CommonQueryFromJSONTyped(json, false);
}
export function CommonQueryFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'query': json['query'] == null ? undefined : json['query'],
        'alias': json['alias'] == null ? undefined : json['alias'],
        'id': json['id'] == null ? undefined : json['id'],
        'labelSelector': json['labelSelector'] == null ? undefined : json['labelSelector'],
        'aggregate': json['aggregate'] == null ? undefined : CommonAggregateFromJSON(json['aggregate']),
        'functions': json['functions'] == null ? undefined : (json['functions'].map(CommonFunctionFromJSON)),
        'disabled': json['disabled'] == null ? undefined : json['disabled'],
    };
}
export function CommonQueryToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'query': value['query'],
        'alias': value['alias'],
        'id': value['id'],
        'labelSelector': value['labelSelector'],
        'aggregate': CommonAggregateToJSON(value['aggregate']),
        'functions': value['functions'] == null ? undefined : (value['functions'].map(CommonFunctionToJSON)),
        'disabled': value['disabled'],
    };
}
