// @sentio/api/dist/src/models/CommonTabularData.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { mapValues } from '../runtime.js';
import { CommonTabularDataColumnTypeFromJSON, CommonTabularDataColumnTypeToJSON, } from './CommonTabularDataColumnType.js';
/**
 * Check if a given object implements the CommonTabularData interface.
 */
export function instanceOfCommonTabularData(value) {
    return true;
}
export function CommonTabularDataFromJSON(json) {
    return CommonTabularDataFromJSONTyped(json, false);
}
export function CommonTabularDataFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'columns': json['columns'] == null ? undefined : json['columns'],
        'columnTypes': json['columnTypes'] == null ? undefined : (mapValues(json['columnTypes'], CommonTabularDataColumnTypeFromJSON)),
        'rows': json['rows'] == null ? undefined : json['rows'],
        'generatedAt': json['generatedAt'] == null ? undefined : (new Date(json['generatedAt'])),
        'cursor': json['cursor'] == null ? undefined : json['cursor'],
    };
}
export function CommonTabularDataToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'columns': value['columns'],
        'columnTypes': value['columnTypes'] == null ? undefined : (mapValues(value['columnTypes'], CommonTabularDataColumnTypeToJSON)),
        'rows': value['rows'],
        'generatedAt': value['generatedAt'] == null ? undefined : ((value['generatedAt']).toISOString()),
        'cursor': value['cursor'],
    };
}
