// @sentio/api/dist/src/models/CommonTimeRangeTimeLike.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonTimeRangeRelativeTimeFromJSON, CommonTimeRangeRelativeTimeToJSON, } from './CommonTimeRangeRelativeTime.js';
/**
 * Check if a given object implements the CommonTimeRangeTimeLike interface.
 */
export function instanceOfCommonTimeRangeTimeLike(value) {
    return true;
}
export function CommonTimeRangeTimeLikeFromJSON(json) {
    return CommonTimeRangeTimeLikeFromJSONTyped(json, false);
}
export function CommonTimeRangeTimeLikeFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'relativeTime': json['relativeTime'] == null ? undefined : CommonTimeRangeRelativeTimeFromJSON(json['relativeTime']),
        'absoluteTime': json['absoluteTime'] == null ? undefined : json['absoluteTime'],
    };
}
export function CommonTimeRangeTimeLikeToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'relativeTime': CommonTimeRangeRelativeTimeToJSON(value['relativeTime']),
        'absoluteTime': value['absoluteTime'],
    };
}
