// @sentio/api/dist/src/models/CommonArgument.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonDurationFromJSON, CommonDurationToJSON, } from './CommonDuration.js';
/**
 * Check if a given object implements the CommonArgument interface.
 */
export function instanceOfCommonArgument(value) {
    return true;
}
export function CommonArgumentFromJSON(json) {
    return CommonArgumentFromJSONTyped(json, false);
}
export function CommonArgumentFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'stringValue': json['stringValue'] == null ? undefined : json['stringValue'],
        'intValue': json['intValue'] == null ? undefined : json['intValue'],
        'doubleValue': json['doubleValue'] == null ? undefined : json['doubleValue'],
        'boolValue': json['boolValue'] == null ? undefined : json['boolValue'],
        'durationValue': json['durationValue'] == null ? undefined : CommonDurationFromJSON(json['durationValue']),
    };
}
export function CommonArgumentToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'stringValue': value['stringValue'],
        'intValue': value['intValue'],
        'doubleValue': value['doubleValue'],
        'boolValue': value['boolValue'],
        'durationValue': CommonDurationToJSON(value['durationValue']),
    };
}
