// @sentio/api/dist/src/models/CommonRetentionQuery.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonRetentionQueryCriteriaFromJSON, CommonRetentionQueryCriteriaToJSON, } from './CommonRetentionQueryCriteria.js';
import { CommonRetentionQueryIntervalFromJSON, CommonRetentionQueryIntervalToJSON, } from './CommonRetentionQueryInterval.js';
import { CommonRetentionQueryResourceFromJSON, CommonRetentionQueryResourceToJSON, } from './CommonRetentionQueryResource.js';
import { CommonSegmentParameterFromJSON, CommonSegmentParameterToJSON, } from './CommonSegmentParameter.js';
import { CommonSelectorExprFromJSON, CommonSelectorExprToJSON, } from './CommonSelectorExpr.js';
/**
 * Check if a given object implements the CommonRetentionQuery interface.
 */
export function instanceOfCommonRetentionQuery(value) {
    return true;
}
export function CommonRetentionQueryFromJSON(json) {
    return CommonRetentionQueryFromJSONTyped(json, false);
}
export function CommonRetentionQueryFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'resources': json['resources'] == null ? undefined : (json['resources'].map(CommonRetentionQueryResourceFromJSON)),
        'criteria': json['criteria'] == null ? undefined : CommonRetentionQueryCriteriaFromJSON(json['criteria']),
        'interval': json['interval'] == null ? undefined : CommonRetentionQueryIntervalFromJSON(json['interval']),
        'selectorExpr': json['selectorExpr'] == null ? undefined : CommonSelectorExprFromJSON(json['selectorExpr']),
        'groupBy': json['groupBy'] == null ? undefined : json['groupBy'],
        'segmentBy': json['segmentBy'] == null ? undefined : (json['segmentBy'].map(CommonSegmentParameterFromJSON)),
        'windowSize': json['windowSize'] == null ? undefined : json['windowSize'],
    };
}
export function CommonRetentionQueryToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'resources': value['resources'] == null ? undefined : (value['resources'].map(CommonRetentionQueryResourceToJSON)),
        'criteria': CommonRetentionQueryCriteriaToJSON(value['criteria']),
        'interval': CommonRetentionQueryIntervalToJSON(value['interval']),
        'selectorExpr': CommonSelectorExprToJSON(value['selectorExpr']),
        'groupBy': value['groupBy'],
        'segmentBy': value['segmentBy'] == null ? undefined : (value['segmentBy'].map(CommonSegmentParameterToJSON)),
        'windowSize': value['windowSize'],
    };
}
