// @sentio/api/dist/src/models/CommonComputeStats.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the CommonComputeStats interface.
 */
export function instanceOfCommonComputeStats(value) {
    return true;
}
export function CommonComputeStatsFromJSON(json) {
    return CommonComputeStatsFromJSONTyped(json, false);
}
export function CommonComputeStatsFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'computedAt': json['computedAt'] == null ? undefined : (new Date(json['computedAt'])),
        'computeCostMs': json['computeCostMs'] == null ? undefined : json['computeCostMs'],
        'binaryVersionHash': json['binaryVersionHash'] == null ? undefined : json['binaryVersionHash'],
        'computedBy': json['computedBy'] == null ? undefined : json['computedBy'],
        'isCached': json['isCached'] == null ? undefined : json['isCached'],
        'isRefreshing': json['isRefreshing'] == null ? undefined : json['isRefreshing'],
    };
}
export function CommonComputeStatsToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'computedAt': value['computedAt'] == null ? undefined : ((value['computedAt']).toISOString()),
        'computeCostMs': value['computeCostMs'],
        'binaryVersionHash': value['binaryVersionHash'],
        'computedBy': value['computedBy'],
        'isCached': value['isCached'],
        'isRefreshing': value['isRefreshing'],
    };
}
