// @sentio/api/dist/src/models/CommonColumnStateSort.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the CommonColumnStateSort interface.
 */
export function instanceOfCommonColumnStateSort(value) {
    return true;
}
export function CommonColumnStateSortFromJSON(json) {
    return CommonColumnStateSortFromJSONTyped(json, false);
}
export function CommonColumnStateSortFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'id': json['id'] == null ? undefined : json['id'],
        'desc': json['desc'] == null ? undefined : json['desc'],
    };
}
export function CommonColumnStateSortToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'id': value['id'],
        'desc': value['desc'],
    };
}
