// @sentio/api/dist/src/models/CommonTimeRange.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonDurationFromJSON, CommonDurationToJSON, } from './CommonDuration.js';
import { CommonTimeRangeTimeLikeFromJSON, CommonTimeRangeTimeLikeToJSON, } from './CommonTimeRangeTimeLike.js';
/**
 * Check if a given object implements the CommonTimeRange interface.
 */
export function instanceOfCommonTimeRange(value) {
    return true;
}
export function CommonTimeRangeFromJSON(json) {
    return CommonTimeRangeFromJSONTyped(json, false);
}
export function CommonTimeRangeFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'start': json['start'] == null ? undefined : CommonTimeRangeTimeLikeFromJSON(json['start']),
        'end': json['end'] == null ? undefined : CommonTimeRangeTimeLikeFromJSON(json['end']),
        'step': json['step'] == null ? undefined : json['step'],
        'interval': json['interval'] == null ? undefined : CommonDurationFromJSON(json['interval']),
        'timezone': json['timezone'] == null ? undefined : json['timezone'],
    };
}
export function CommonTimeRangeToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'start': CommonTimeRangeTimeLikeToJSON(value['start']),
        'end': CommonTimeRangeTimeLikeToJSON(value['end']),
        'step': value['step'],
        'interval': CommonDurationToJSON(value['interval']),
        'timezone': value['timezone'],
    };
}
