// @sentio/api/dist/src/models/CommonSelectorExprLogicExpr.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonJoinOperatorFromJSON, CommonJoinOperatorToJSON, } from './CommonJoinOperator.js';
import { CommonSelectorExprFromJSON, CommonSelectorExprToJSON, } from './CommonSelectorExpr.js';
/**
 * Check if a given object implements the CommonSelectorExprLogicExpr interface.
 */
export function instanceOfCommonSelectorExprLogicExpr(value) {
    return true;
}
export function CommonSelectorExprLogicExprFromJSON(json) {
    return CommonSelectorExprLogicExprFromJSONTyped(json, false);
}
export function CommonSelectorExprLogicExprFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'expressions': json['expressions'] == null ? undefined : (json['expressions'].map(CommonSelectorExprFromJSON)),
        'operator': json['operator'] == null ? undefined : CommonJoinOperatorFromJSON(json['operator']),
    };
}
export function CommonSelectorExprLogicExprToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'expressions': value['expressions'] == null ? undefined : (value['expressions'].map(CommonSelectorExprToJSON)),
        'operator': CommonJoinOperatorToJSON(value['operator']),
    };
}
