// @sentio/api/dist/src/models/CommonCohortsFilterAggregationOperatorType.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 */
export const CommonCohortsFilterAggregationOperatorType = {
    Eq: 'EQ',
    Neq: 'NEQ',
    Gt: 'GT',
    Gte: 'GTE',
    Lt: 'LT',
    Lte: 'LTE',
    Between: 'BETWEEN',
    NotBetween: 'NOT_BETWEEN'
};
export function instanceOfCommonCohortsFilterAggregationOperatorType(value) {
    return Object.values(CommonCohortsFilterAggregationOperatorType).includes(value);
}
export function CommonCohortsFilterAggregationOperatorTypeFromJSON(json) {
    return CommonCohortsFilterAggregationOperatorTypeFromJSONTyped(json, false);
}
export function CommonCohortsFilterAggregationOperatorTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
export function CommonCohortsFilterAggregationOperatorTypeToJSON(value) {
    return value;
}
