// @sentio/api/dist/src/models/CommonMatrixSample.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonMatrixMetricFromJSON, CommonMatrixMetricToJSON, } from './CommonMatrixMetric.js';
import { CommonMatrixValueFromJSON, CommonMatrixValueToJSON, } from './CommonMatrixValue.js';
/**
 * Check if a given object implements the CommonMatrixSample interface.
 */
export function instanceOfCommonMatrixSample(value) {
    return true;
}
export function CommonMatrixSampleFromJSON(json) {
    return CommonMatrixSampleFromJSONTyped(json, false);
}
export function CommonMatrixSampleFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'metric': json['metric'] == null ? undefined : CommonMatrixMetricFromJSON(json['metric']),
        'values': json['values'] == null ? undefined : (json['values'].map(CommonMatrixValueFromJSON)),
    };
}
export function CommonMatrixSampleToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'metric': CommonMatrixMetricToJSON(value['metric']),
        'values': value['values'] == null ? undefined : (value['values'].map(CommonMatrixValueToJSON)),
    };
}
