// @sentio/api/dist/src/models/CommonProjectProjectMember.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonUserInfoFromJSON, CommonUserInfoToJSON, } from './CommonUserInfo.js';
/**
 * Check if a given object implements the CommonProjectProjectMember interface.
 */
export function instanceOfCommonProjectProjectMember(value) {
    return true;
}
export function CommonProjectProjectMemberFromJSON(json) {
    return CommonProjectProjectMemberFromJSONTyped(json, false);
}
export function CommonProjectProjectMemberFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'user': json['user'] == null ? undefined : CommonUserInfoFromJSON(json['user']),
        'role': json['role'] == null ? undefined : json['role'],
    };
}
export function CommonProjectProjectMemberToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'user': CommonUserInfoToJSON(value['user']),
        'role': value['role'],
    };
}
