// @sentio/api/dist/src/models/CommonRetentionQueryFilter.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonRetentionQueryFilterTimeFilterFromJSON, CommonRetentionQueryFilterTimeFilterToJSON, } from './CommonRetentionQueryFilterTimeFilter.js';
import { CommonSelectorExprFromJSON, CommonSelectorExprToJSON, } from './CommonSelectorExpr.js';
/**
 * Check if a given object implements the CommonRetentionQueryFilter interface.
 */
export function instanceOfCommonRetentionQueryFilter(value) {
    return true;
}
export function CommonRetentionQueryFilterFromJSON(json) {
    return CommonRetentionQueryFilterFromJSONTyped(json, false);
}
export function CommonRetentionQueryFilterFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'propertyFilter': json['propertyFilter'] == null ? undefined : CommonSelectorExprFromJSON(json['propertyFilter']),
        'timeFilter': json['timeFilter'] == null ? undefined : CommonRetentionQueryFilterTimeFilterFromJSON(json['timeFilter']),
    };
}
export function CommonRetentionQueryFilterToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'propertyFilter': CommonSelectorExprToJSON(value['propertyFilter']),
        'timeFilter': CommonRetentionQueryFilterTimeFilterToJSON(value['timeFilter']),
    };
}
