// @sentio/api/dist/src/models/CommonRetentionMatrix.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonRetentionMatrixSampleFromJSON, CommonRetentionMatrixSampleToJSON, } from './CommonRetentionMatrixSample.js';
/**
 * Check if a given object implements the CommonRetentionMatrix interface.
 */
export function instanceOfCommonRetentionMatrix(value) {
    return true;
}
export function CommonRetentionMatrixFromJSON(json) {
    return CommonRetentionMatrixFromJSONTyped(json, false);
}
export function CommonRetentionMatrixFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'samples': json['samples'] == null ? undefined : (json['samples'].map(CommonRetentionMatrixSampleFromJSON)),
    };
}
export function CommonRetentionMatrixToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'samples': value['samples'] == null ? undefined : (value['samples'].map(CommonRetentionMatrixSampleToJSON)),
    };
}
