// @sentio/api/dist/src/models/CommonFormula.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonFunctionFromJSON, CommonFunctionToJSON, } from './CommonFunction.js';
/**
 * Check if a given object implements the CommonFormula interface.
 */
export function instanceOfCommonFormula(value) {
    return true;
}
export function CommonFormulaFromJSON(json) {
    return CommonFormulaFromJSONTyped(json, false);
}
export function CommonFormulaFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'expression': json['expression'] == null ? undefined : json['expression'],
        'alias': json['alias'] == null ? undefined : json['alias'],
        'id': json['id'] == null ? undefined : json['id'],
        'disabled': json['disabled'] == null ? undefined : json['disabled'],
        'functions': json['functions'] == null ? undefined : (json['functions'].map(CommonFunctionFromJSON)),
    };
}
export function CommonFormulaToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'expression': value['expression'],
        'alias': value['alias'],
        'id': value['id'],
        'disabled': value['disabled'],
        'functions': value['functions'] == null ? undefined : (value['functions'].map(CommonFunctionToJSON)),
    };
}
