// @sentio/api/dist/src/models/CommonRetentionQueryFilterTimeFilter.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonRetentionQueryFilterTimeFilterTypeFromJSON, CommonRetentionQueryFilterTimeFilterTypeToJSON, } from './CommonRetentionQueryFilterTimeFilterType.js';
/**
 * Check if a given object implements the CommonRetentionQueryFilterTimeFilter interface.
 */
export function instanceOfCommonRetentionQueryFilterTimeFilter(value) {
    return true;
}
export function CommonRetentionQueryFilterTimeFilterFromJSON(json) {
    return CommonRetentionQueryFilterTimeFilterFromJSONTyped(json, false);
}
export function CommonRetentionQueryFilterTimeFilterFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'type': json['type'] == null ? undefined : CommonRetentionQueryFilterTimeFilterTypeFromJSON(json['type']),
    };
}
export function CommonRetentionQueryFilterTimeFilterToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'type': CommonRetentionQueryFilterTimeFilterTypeToJSON(value['type']),
    };
}
