// @sentio/api/dist/src/models/CommonCoinID.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonCoinIDAddressIdentifierFromJSON, CommonCoinIDAddressIdentifierToJSON, } from './CommonCoinIDAddressIdentifier.js';
/**
 * Check if a given object implements the CommonCoinID interface.
 */
export function instanceOfCommonCoinID(value) {
    return true;
}
export function CommonCoinIDFromJSON(json) {
    return CommonCoinIDFromJSONTyped(json, false);
}
export function CommonCoinIDFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'symbol': json['symbol'] == null ? undefined : json['symbol'],
        'address': json['address'] == null ? undefined : CommonCoinIDAddressIdentifierFromJSON(json['address']),
    };
}
export function CommonCoinIDToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'symbol': value['symbol'],
        'address': CommonCoinIDAddressIdentifierToJSON(value['address']),
    };
}
