// @sentio/api/dist/src/models/CommonSegmentationQueryResource.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonCohortsQueryFromJSON, CommonCohortsQueryToJSON, } from './CommonCohortsQuery.js';
import { CommonSegmentationQueryResourceTypeFromJSON, CommonSegmentationQueryResourceTypeToJSON, } from './CommonSegmentationQueryResourceType.js';
/**
 * Check if a given object implements the CommonSegmentationQueryResource interface.
 */
export function instanceOfCommonSegmentationQueryResource(value) {
    return true;
}
export function CommonSegmentationQueryResourceFromJSON(json) {
    return CommonSegmentationQueryResourceFromJSONTyped(json, false);
}
export function CommonSegmentationQueryResourceFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'name': json['name'] == null ? undefined : json['name'],
        'type': json['type'] == null ? undefined : CommonSegmentationQueryResourceTypeFromJSON(json['type']),
        'cohortsId': json['cohortsId'] == null ? undefined : json['cohortsId'],
        'cohortsQuery': json['cohortsQuery'] == null ? undefined : CommonCohortsQueryFromJSON(json['cohortsQuery']),
    };
}
export function CommonSegmentationQueryResourceToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'name': value['name'],
        'type': CommonSegmentationQueryResourceTypeToJSON(value['type']),
        'cohortsId': value['cohortsId'],
        'cohortsQuery': CommonCohortsQueryToJSON(value['cohortsQuery']),
    };
}
