// @sentio/api/dist/src/models/CommonCohortsGroup.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonCohortsFilterFromJSON, CommonCohortsFilterToJSON, } from './CommonCohortsFilter.js';
import { CommonJoinOperatorFromJSON, CommonJoinOperatorToJSON, } from './CommonJoinOperator.js';
/**
 * Check if a given object implements the CommonCohortsGroup interface.
 */
export function instanceOfCommonCohortsGroup(value) {
    return true;
}
export function CommonCohortsGroupFromJSON(json) {
    return CommonCohortsGroupFromJSONTyped(json, false);
}
export function CommonCohortsGroupFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'joinOperator': json['joinOperator'] == null ? undefined : CommonJoinOperatorFromJSON(json['joinOperator']),
        'filters': json['filters'] == null ? undefined : (json['filters'].map(CommonCohortsFilterFromJSON)),
    };
}
export function CommonCohortsGroupToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'joinOperator': CommonJoinOperatorToJSON(value['joinOperator']),
        'filters': value['filters'] == null ? undefined : (value['filters'].map(CommonCohortsFilterToJSON)),
    };
}
