// @sentio/api/dist/src/models/CommonRetentionQueryFilterTimeFilterType.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 */
export const CommonRetentionQueryFilterTimeFilterType = {
    Disable: 'Disable',
    FirstInTimeRange: 'FirstInTimeRange',
    FirstInGlobal: 'FirstInGlobal'
};
export function instanceOfCommonRetentionQueryFilterTimeFilterType(value) {
    return Object.values(CommonRetentionQueryFilterTimeFilterType).includes(value);
}
export function CommonRetentionQueryFilterTimeFilterTypeFromJSON(json) {
    return CommonRetentionQueryFilterTimeFilterTypeFromJSONTyped(json, false);
}
export function CommonRetentionQueryFilterTimeFilterTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
export function CommonRetentionQueryFilterTimeFilterTypeToJSON(value) {
    return value;
}
