// @sentio/api/dist/src/models/CommonEventLogColumn.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the CommonEventLogColumn interface.
 */
export function instanceOfCommonEventLogColumn(value) {
    return true;
}
export function CommonEventLogColumnFromJSON(json) {
    return CommonEventLogColumnFromJSONTyped(json, false);
}
export function CommonEventLogColumnFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'id': json['id'] == null ? undefined : json['id'],
        'size': json['size'] == null ? undefined : json['size'],
        'name': json['name'] == null ? undefined : json['name'],
        'accessorKey': json['accessorKey'] == null ? undefined : json['accessorKey'],
        'enableHiding': json['enableHiding'] == null ? undefined : json['enableHiding'],
        'enableSorting': json['enableSorting'] == null ? undefined : json['enableSorting'],
        'enableResizing': json['enableResizing'] == null ? undefined : json['enableResizing'],
    };
}
export function CommonEventLogColumnToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'id': value['id'],
        'size': value['size'],
        'name': value['name'],
        'accessorKey': value['accessorKey'],
        'enableHiding': value['enableHiding'],
        'enableSorting': value['enableSorting'],
        'enableResizing': value['enableResizing'],
    };
}
