// @sentio/api/dist/src/models/CommonTimeRangeLite.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the CommonTimeRangeLite interface.
 */
export function instanceOfCommonTimeRangeLite(value) {
    if (!('start' in value))
        return false;
    if (!('end' in value))
        return false;
    if (!('step' in value))
        return false;
    return true;
}
export function CommonTimeRangeLiteFromJSON(json) {
    return CommonTimeRangeLiteFromJSONTyped(json, false);
}
export function CommonTimeRangeLiteFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'start': json['start'],
        'end': json['end'],
        'step': json['step'],
        'timezone': json['timezone'] == null ? undefined : json['timezone'],
    };
}
export function CommonTimeRangeLiteToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'start': value['start'],
        'end': value['end'],
        'step': value['step'],
        'timezone': value['timezone'],
    };
}
