// @sentio/api/dist/src/models/CommonCohortsQuery.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonCohortsGroupFromJSON, CommonCohortsGroupToJSON, } from './CommonCohortsGroup.js';
import { CommonJoinOperatorFromJSON, CommonJoinOperatorToJSON, } from './CommonJoinOperator.js';
/**
 * Check if a given object implements the CommonCohortsQuery interface.
 */
export function instanceOfCommonCohortsQuery(value) {
    return true;
}
export function CommonCohortsQueryFromJSON(json) {
    return CommonCohortsQueryFromJSONTyped(json, false);
}
export function CommonCohortsQueryFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'joinOperator': json['joinOperator'] == null ? undefined : CommonJoinOperatorFromJSON(json['joinOperator']),
        'groups': json['groups'] == null ? undefined : (json['groups'].map(CommonCohortsGroupFromJSON)),
        'name': json['name'] == null ? undefined : json['name'],
        'id': json['id'] == null ? undefined : json['id'],
    };
}
export function CommonCohortsQueryToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'joinOperator': CommonJoinOperatorToJSON(value['joinOperator']),
        'groups': value['groups'] == null ? undefined : (value['groups'].map(CommonCohortsGroupToJSON)),
        'name': value['name'],
        'id': value['id'],
    };
}
