// @sentio/api/dist/src/models/EvmAccessListItem.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the EvmAccessListItem interface.
 */
export function instanceOfEvmAccessListItem(value) {
    return true;
}
export function EvmAccessListItemFromJSON(json) {
    return EvmAccessListItemFromJSONTyped(json, false);
}
export function EvmAccessListItemFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'address': json['address'] == null ? undefined : json['address'],
        'storageKeys': json['storageKeys'] == null ? undefined : json['storageKeys'],
    };
}
export function EvmAccessListItemToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'address': value['address'],
        'storageKeys': value['storageKeys'],
    };
}
