// @sentio/api/dist/src/models/EvmTransactionReceipt.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the EvmTransactionReceipt interface.
 */
export function instanceOfEvmTransactionReceipt(value) {
    return true;
}
export function EvmTransactionReceiptFromJSON(json) {
    return EvmTransactionReceiptFromJSONTyped(json, false);
}
export function EvmTransactionReceiptFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'gasUsed': json['gasUsed'] == null ? undefined : json['gasUsed'],
        'cumulativeGasUsed': json['cumulativeGasUsed'] == null ? undefined : json['cumulativeGasUsed'],
        'effectiveGasPrice': json['effectiveGasPrice'] == null ? undefined : json['effectiveGasPrice'],
        'status': json['status'] == null ? undefined : json['status'],
        'error': json['error'] == null ? undefined : json['error'],
        'revertReason': json['revertReason'] == null ? undefined : json['revertReason'],
        'logs': json['logs'] == null ? undefined : json['logs'],
    };
}
export function EvmTransactionReceiptToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'gasUsed': value['gasUsed'],
        'cumulativeGasUsed': value['cumulativeGasUsed'],
        'effectiveGasPrice': value['effectiveGasPrice'],
        'status': value['status'],
        'error': value['error'],
        'revertReason': value['revertReason'],
        'logs': value['logs'],
    };
}
