// @sentio/api/dist/src/models/GoogleProtobufNullValue.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * `NullValue` is a singleton enumeration to represent the null value for the
 * `Value` type union.
 *
 * The JSON representation for `NullValue` is JSON `null`.
 *
 *  - NULL_VALUE: Null value.
 * @export
 */
export const GoogleProtobufNullValue = {
    NullValue: 'NULL_VALUE'
};
export function instanceOfGoogleProtobufNullValue(value) {
    return Object.values(GoogleProtobufNullValue).includes(value);
}
export function GoogleProtobufNullValueFromJSON(json) {
    return GoogleProtobufNullValueFromJSONTyped(json, false);
}
export function GoogleProtobufNullValueFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
export function GoogleProtobufNullValueToJSON(value) {
    return value;
}
