// @sentio/api/dist/src/models/GoogleApiHttpBody.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { GoogleProtobufAnyFromJSON, GoogleProtobufAnyToJSON, } from './GoogleProtobufAny.js';
/**
 * Check if a given object implements the GoogleApiHttpBody interface.
 */
export function instanceOfGoogleApiHttpBody(value) {
    return true;
}
export function GoogleApiHttpBodyFromJSON(json) {
    return GoogleApiHttpBodyFromJSONTyped(json, false);
}
export function GoogleApiHttpBodyFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'contentType': json['contentType'] == null ? undefined : json['contentType'],
        'data': json['data'] == null ? undefined : json['data'],
        'extensions': json['extensions'] == null ? undefined : (json['extensions'].map(GoogleProtobufAnyFromJSON)),
    };
}
export function GoogleApiHttpBodyToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'contentType': value['contentType'],
        'data': value['data'],
        'extensions': value['extensions'] == null ? undefined : (value['extensions'].map(GoogleProtobufAnyToJSON)),
    };
}
