// @sentio/api/dist/src/models/InsightsServiceQueryResponseResult.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonComputeStatsFromJSON, CommonComputeStatsToJSON, } from './CommonComputeStats.js';
import { CommonMatrixFromJSON, CommonMatrixToJSON, } from './CommonMatrix.js';
import { InsightsServiceDataSourceFromJSON, InsightsServiceDataSourceToJSON, } from './InsightsServiceDataSource.js';
/**
 * Check if a given object implements the InsightsServiceQueryResponseResult interface.
 */
export function instanceOfInsightsServiceQueryResponseResult(value) {
    return true;
}
export function InsightsServiceQueryResponseResultFromJSON(json) {
    return InsightsServiceQueryResponseResultFromJSONTyped(json, false);
}
export function InsightsServiceQueryResponseResultFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'id': json['id'] == null ? undefined : json['id'],
        'alias': json['alias'] == null ? undefined : json['alias'],
        'dataSource': json['dataSource'] == null ? undefined : InsightsServiceDataSourceFromJSON(json['dataSource']),
        'matrix': json['matrix'] == null ? undefined : CommonMatrixFromJSON(json['matrix']),
        'error': json['error'] == null ? undefined : json['error'],
        'computeStats': json['computeStats'] == null ? undefined : CommonComputeStatsFromJSON(json['computeStats']),
    };
}
export function InsightsServiceQueryResponseResultToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'id': value['id'],
        'alias': value['alias'],
        'dataSource': InsightsServiceDataSourceToJSON(value['dataSource']),
        'matrix': CommonMatrixToJSON(value['matrix']),
        'error': value['error'],
        'computeStats': CommonComputeStatsToJSON(value['computeStats']),
    };
}
