// @sentio/api/dist/src/models/InsightsServiceRetentionRequest.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonRetentionQueryFromJSON, CommonRetentionQueryToJSON, } from './CommonRetentionQuery.js';
import { CommonTimeRangeLiteFromJSON, CommonTimeRangeLiteToJSON, } from './CommonTimeRangeLite.js';
/**
 * Check if a given object implements the InsightsServiceRetentionRequest interface.
 */
export function instanceOfInsightsServiceRetentionRequest(value) {
    return true;
}
export function InsightsServiceRetentionRequestFromJSON(json) {
    return InsightsServiceRetentionRequestFromJSONTyped(json, false);
}
export function InsightsServiceRetentionRequestFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'projectOwner': json['projectOwner'] == null ? undefined : json['projectOwner'],
        'projectSlug': json['projectSlug'] == null ? undefined : json['projectSlug'],
        'projectId': json['projectId'] == null ? undefined : json['projectId'],
        'version': json['version'] == null ? undefined : json['version'],
        'timeRange': json['timeRange'] == null ? undefined : CommonTimeRangeLiteFromJSON(json['timeRange']),
        'query': json['query'] == null ? undefined : CommonRetentionQueryFromJSON(json['query']),
    };
}
export function InsightsServiceRetentionRequestToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'projectOwner': value['projectOwner'],
        'projectSlug': value['projectSlug'],
        'projectId': value['projectId'],
        'version': value['version'],
        'timeRange': CommonTimeRangeLiteToJSON(value['timeRange']),
        'query': CommonRetentionQueryToJSON(value['query']),
    };
}
