// @sentio/api/dist/src/models/InsightsServiceQueryResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { InsightsServiceQueryResponseResultFromJSON, InsightsServiceQueryResponseResultToJSON, } from './InsightsServiceQueryResponseResult.js';
/**
 * Check if a given object implements the InsightsServiceQueryResponse interface.
 */
export function instanceOfInsightsServiceQueryResponse(value) {
    return true;
}
export function InsightsServiceQueryResponseFromJSON(json) {
    return InsightsServiceQueryResponseFromJSONTyped(json, false);
}
export function InsightsServiceQueryResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'results': json['results'] == null ? undefined : (json['results'].map(InsightsServiceQueryResponseResultFromJSON)),
    };
}
export function InsightsServiceQueryResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'results': value['results'] == null ? undefined : (value['results'].map(InsightsServiceQueryResponseResultToJSON)),
    };
}
