// @sentio/api/dist/src/models/InsightsServiceRetentionResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonComputeStatsFromJSON, CommonComputeStatsToJSON, } from './CommonComputeStats.js';
import { CommonRetentionMatrixFromJSON, CommonRetentionMatrixToJSON, } from './CommonRetentionMatrix.js';
/**
 * Check if a given object implements the InsightsServiceRetentionResponse interface.
 */
export function instanceOfInsightsServiceRetentionResponse(value) {
    return true;
}
export function InsightsServiceRetentionResponseFromJSON(json) {
    return InsightsServiceRetentionResponseFromJSONTyped(json, false);
}
export function InsightsServiceRetentionResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'computeStats': json['computeStats'] == null ? undefined : CommonComputeStatsFromJSON(json['computeStats']),
        'results': json['results'] == null ? undefined : CommonRetentionMatrixFromJSON(json['results']),
        'error': json['error'] == null ? undefined : json['error'],
    };
}
export function InsightsServiceRetentionResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'computeStats': CommonComputeStatsToJSON(value['computeStats']),
        'results': CommonRetentionMatrixToJSON(value['results']),
        'error': value['error'],
    };
}
