// @sentio/api/dist/src/models/InsightsServiceQueryRequestQuery.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonPriceSegmentationQueryFromJSON, CommonPriceSegmentationQueryToJSON, } from './CommonPriceSegmentationQuery.js';
import { CommonQueryFromJSON, CommonQueryToJSON, } from './CommonQuery.js';
import { CommonSegmentationQueryFromJSON, CommonSegmentationQueryToJSON, } from './CommonSegmentationQuery.js';
import { InsightsServiceDataSourceFromJSON, InsightsServiceDataSourceToJSON, } from './InsightsServiceDataSource.js';
/**
 * Check if a given object implements the InsightsServiceQueryRequestQuery interface.
 */
export function instanceOfInsightsServiceQueryRequestQuery(value) {
    return true;
}
export function InsightsServiceQueryRequestQueryFromJSON(json) {
    return InsightsServiceQueryRequestQueryFromJSONTyped(json, false);
}
export function InsightsServiceQueryRequestQueryFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'metricsQuery': json['metricsQuery'] == null ? undefined : CommonQueryFromJSON(json['metricsQuery']),
        'eventsQuery': json['eventsQuery'] == null ? undefined : CommonSegmentationQueryFromJSON(json['eventsQuery']),
        'priceQuery': json['priceQuery'] == null ? undefined : CommonPriceSegmentationQueryFromJSON(json['priceQuery']),
        'dataSource': json['dataSource'] == null ? undefined : InsightsServiceDataSourceFromJSON(json['dataSource']),
        'sourceName': json['sourceName'] == null ? undefined : json['sourceName'],
    };
}
export function InsightsServiceQueryRequestQueryToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'metricsQuery': CommonQueryToJSON(value['metricsQuery']),
        'eventsQuery': CommonSegmentationQueryToJSON(value['eventsQuery']),
        'priceQuery': CommonPriceSegmentationQueryToJSON(value['priceQuery']),
        'dataSource': InsightsServiceDataSourceToJSON(value['dataSource']),
        'sourceName': value['sourceName'],
    };
}
