// @sentio/api/dist/src/models/InsightsServiceListCoinsResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonCoinIDFromJSON, CommonCoinIDToJSON, } from './CommonCoinID.js';
import { CommonComputeStatsFromJSON, CommonComputeStatsToJSON, } from './CommonComputeStats.js';
/**
 * Check if a given object implements the InsightsServiceListCoinsResponse interface.
 */
export function instanceOfInsightsServiceListCoinsResponse(value) {
    return true;
}
export function InsightsServiceListCoinsResponseFromJSON(json) {
    return InsightsServiceListCoinsResponseFromJSONTyped(json, false);
}
export function InsightsServiceListCoinsResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'coins': json['coins'] == null ? undefined : (json['coins'].map(CommonCoinIDFromJSON)),
        'computeStats': json['computeStats'] == null ? undefined : CommonComputeStatsFromJSON(json['computeStats']),
    };
}
export function InsightsServiceListCoinsResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'coins': value['coins'] == null ? undefined : (value['coins'].map(CommonCoinIDToJSON)),
        'computeStats': CommonComputeStatsToJSON(value['computeStats']),
    };
}
