// @sentio/api/dist/src/models/InsightsServiceInsightsServiceQueryBody.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonCachePolicyFromJSON, CommonCachePolicyToJSON, } from './CommonCachePolicy.js';
import { CommonFormulaFromJSON, CommonFormulaToJSON, } from './CommonFormula.js';
import { CommonTimeRangeLiteFromJSON, CommonTimeRangeLiteToJSON, } from './CommonTimeRangeLite.js';
import { InsightsServiceQueryRequestQueryFromJSON, InsightsServiceQueryRequestQueryToJSON, } from './InsightsServiceQueryRequestQuery.js';
/**
 * Check if a given object implements the InsightsServiceInsightsServiceQueryBody interface.
 */
export function instanceOfInsightsServiceInsightsServiceQueryBody(value) {
    return true;
}
export function InsightsServiceInsightsServiceQueryBodyFromJSON(json) {
    return InsightsServiceInsightsServiceQueryBodyFromJSONTyped(json, false);
}
export function InsightsServiceInsightsServiceQueryBodyFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'projectId': json['projectId'] == null ? undefined : json['projectId'],
        'version': json['version'] == null ? undefined : json['version'],
        'timeRange': json['timeRange'] == null ? undefined : CommonTimeRangeLiteFromJSON(json['timeRange']),
        'queries': json['queries'] == null ? undefined : (json['queries'].map(InsightsServiceQueryRequestQueryFromJSON)),
        'formulas': json['formulas'] == null ? undefined : (json['formulas'].map(CommonFormulaFromJSON)),
        'limit': json['limit'] == null ? undefined : json['limit'],
        'offset': json['offset'] == null ? undefined : json['offset'],
        'bypassCache': json['bypassCache'] == null ? undefined : json['bypassCache'],
        'cachePolicy': json['cachePolicy'] == null ? undefined : CommonCachePolicyFromJSON(json['cachePolicy']),
    };
}
export function InsightsServiceInsightsServiceQueryBodyToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'projectId': value['projectId'],
        'version': value['version'],
        'timeRange': CommonTimeRangeLiteToJSON(value['timeRange']),
        'queries': value['queries'] == null ? undefined : (value['queries'].map(InsightsServiceQueryRequestQueryToJSON)),
        'formulas': value['formulas'] == null ? undefined : (value['formulas'].map(CommonFormulaToJSON)),
        'limit': value['limit'],
        'offset': value['offset'],
        'bypassCache': value['bypassCache'],
        'cachePolicy': CommonCachePolicyToJSON(value['cachePolicy']),
    };
}
