// @sentio/api/dist/src/models/MetricsServiceMetricMetadata.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the MetricsServiceMetricMetadata interface.
 */
export function instanceOfMetricsServiceMetricMetadata(value) {
    return true;
}
export function MetricsServiceMetricMetadataFromJSON(json) {
    return MetricsServiceMetricMetadataFromJSONTyped(json, false);
}
export function MetricsServiceMetricMetadataFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'type': json['type'] == null ? undefined : json['type'],
        'unit': json['unit'] == null ? undefined : json['unit'],
        'help': json['help'] == null ? undefined : json['help'],
        'lastSeen': json['lastSeen'] == null ? undefined : json['lastSeen'],
    };
}
export function MetricsServiceMetricMetadataToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'type': value['type'],
        'unit': value['unit'],
        'help': value['help'],
        'lastSeen': value['lastSeen'],
    };
}
