// @sentio/api/dist/src/models/MetricsServiceMetricsQueryResponseMatrix.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { MetricsServiceMetricsQueryResponseSampleFromJSON, MetricsServiceMetricsQueryResponseSampleToJSON, } from './MetricsServiceMetricsQueryResponseSample.js';
/**
 * Check if a given object implements the MetricsServiceMetricsQueryResponseMatrix interface.
 */
export function instanceOfMetricsServiceMetricsQueryResponseMatrix(value) {
    return true;
}
export function MetricsServiceMetricsQueryResponseMatrixFromJSON(json) {
    return MetricsServiceMetricsQueryResponseMatrixFromJSONTyped(json, false);
}
export function MetricsServiceMetricsQueryResponseMatrixFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'samples': json['samples'] == null ? undefined : (json['samples'].map(MetricsServiceMetricsQueryResponseSampleFromJSON)),
        'totalSamples': json['totalSamples'] == null ? undefined : json['totalSamples'],
    };
}
export function MetricsServiceMetricsQueryResponseMatrixToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'samples': value['samples'] == null ? undefined : (value['samples'].map(MetricsServiceMetricsQueryResponseSampleToJSON)),
        'totalSamples': value['totalSamples'],
    };
}
