// @sentio/api/dist/src/models/MetricsServiceMetricsQueryResponseResult.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonComputeStatsFromJSON, CommonComputeStatsToJSON, } from './CommonComputeStats.js';
import { MetricsServiceMetricsQueryResponseMatrixFromJSON, MetricsServiceMetricsQueryResponseMatrixToJSON, } from './MetricsServiceMetricsQueryResponseMatrix.js';
/**
 * Check if a given object implements the MetricsServiceMetricsQueryResponseResult interface.
 */
export function instanceOfMetricsServiceMetricsQueryResponseResult(value) {
    return true;
}
export function MetricsServiceMetricsQueryResponseResultFromJSON(json) {
    return MetricsServiceMetricsQueryResponseResultFromJSONTyped(json, false);
}
export function MetricsServiceMetricsQueryResponseResultFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'matrix': json['matrix'] == null ? undefined : MetricsServiceMetricsQueryResponseMatrixFromJSON(json['matrix']),
        'error': json['error'] == null ? undefined : json['error'],
        'alias': json['alias'] == null ? undefined : json['alias'],
        'id': json['id'] == null ? undefined : json['id'],
        'computeStats': json['computeStats'] == null ? undefined : CommonComputeStatsFromJSON(json['computeStats']),
    };
}
export function MetricsServiceMetricsQueryResponseResultToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'matrix': MetricsServiceMetricsQueryResponseMatrixToJSON(value['matrix']),
        'error': value['error'],
        'alias': value['alias'],
        'id': value['id'],
        'computeStats': CommonComputeStatsToJSON(value['computeStats']),
    };
}
