// @sentio/api/dist/src/models/MetricsServiceMetricsQueryResponseMetric.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the MetricsServiceMetricsQueryResponseMetric interface.
 */
export function instanceOfMetricsServiceMetricsQueryResponseMetric(value) {
    return true;
}
export function MetricsServiceMetricsQueryResponseMetricFromJSON(json) {
    return MetricsServiceMetricsQueryResponseMetricFromJSONTyped(json, false);
}
export function MetricsServiceMetricsQueryResponseMetricFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'name': json['name'] == null ? undefined : json['name'],
        'labels': json['labels'] == null ? undefined : json['labels'],
        'displayName': json['displayName'] == null ? undefined : json['displayName'],
    };
}
export function MetricsServiceMetricsQueryResponseMetricToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'name': value['name'],
        'labels': value['labels'],
        'displayName': value['displayName'],
    };
}
