// @sentio/api/dist/src/models/MetricsServiceQueryValueResponseResult.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { MetricsServiceMetricsQueryResponseSampleFromJSON, MetricsServiceMetricsQueryResponseSampleToJSON, } from './MetricsServiceMetricsQueryResponseSample.js';
/**
 * Check if a given object implements the MetricsServiceQueryValueResponseResult interface.
 */
export function instanceOfMetricsServiceQueryValueResponseResult(value) {
    return true;
}
export function MetricsServiceQueryValueResponseResultFromJSON(json) {
    return MetricsServiceQueryValueResponseResultFromJSONTyped(json, false);
}
export function MetricsServiceQueryValueResponseResultFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'sample': json['sample'] == null ? undefined : (json['sample'].map(MetricsServiceMetricsQueryResponseSampleFromJSON)),
        'error': json['error'] == null ? undefined : json['error'],
        'alias': json['alias'] == null ? undefined : json['alias'],
        'id': json['id'] == null ? undefined : json['id'],
    };
}
export function MetricsServiceQueryValueResponseResultToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'sample': value['sample'] == null ? undefined : (value['sample'].map(MetricsServiceMetricsQueryResponseSampleToJSON)),
        'error': value['error'],
        'alias': value['alias'],
        'id': value['id'],
    };
}
