// @sentio/api/dist/src/models/MetricsServiceObservabilityServiceQueryRangeBody.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonFormulaFromJSON, CommonFormulaToJSON, } from './CommonFormula.js';
import { CommonQueryFromJSON, CommonQueryToJSON, } from './CommonQuery.js';
import { CommonTimeRangeLiteFromJSON, CommonTimeRangeLiteToJSON, } from './CommonTimeRangeLite.js';
/**
 * Check if a given object implements the MetricsServiceObservabilityServiceQueryRangeBody interface.
 */
export function instanceOfMetricsServiceObservabilityServiceQueryRangeBody(value) {
    if (!('timeRange' in value))
        return false;
    return true;
}
export function MetricsServiceObservabilityServiceQueryRangeBodyFromJSON(json) {
    return MetricsServiceObservabilityServiceQueryRangeBodyFromJSONTyped(json, false);
}
export function MetricsServiceObservabilityServiceQueryRangeBodyFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'queries': json['queries'] == null ? undefined : (json['queries'].map(CommonQueryFromJSON)),
        'formulas': json['formulas'] == null ? undefined : (json['formulas'].map(CommonFormulaFromJSON)),
        'samplesLimit': json['samplesLimit'] == null ? undefined : json['samplesLimit'],
        'timeRange': CommonTimeRangeLiteFromJSON(json['timeRange']),
        'projectId': json['projectId'] == null ? undefined : json['projectId'],
        'version': json['version'] == null ? undefined : json['version'],
        'samplesOffset': json['samplesOffset'] == null ? undefined : json['samplesOffset'],
    };
}
export function MetricsServiceObservabilityServiceQueryRangeBodyToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'queries': value['queries'] == null ? undefined : (value['queries'].map(CommonQueryToJSON)),
        'formulas': value['formulas'] == null ? undefined : (value['formulas'].map(CommonFormulaToJSON)),
        'samplesLimit': value['samplesLimit'],
        'timeRange': CommonTimeRangeLiteToJSON(value['timeRange']),
        'projectId': value['projectId'],
        'version': value['version'],
        'samplesOffset': value['samplesOffset'],
    };
}
