// @sentio/api/dist/src/models/MetricsServiceMetricInfo.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { mapValues } from '../runtime.js';
import { MetricsServiceMetricInfoLabelValuesFromJSON, MetricsServiceMetricInfoLabelValuesToJSON, } from './MetricsServiceMetricInfoLabelValues.js';
import { MetricsServiceMetricMetadataFromJSON, MetricsServiceMetricMetadataToJSON, } from './MetricsServiceMetricMetadata.js';
/**
 * Check if a given object implements the MetricsServiceMetricInfo interface.
 */
export function instanceOfMetricsServiceMetricInfo(value) {
    return true;
}
export function MetricsServiceMetricInfoFromJSON(json) {
    return MetricsServiceMetricInfoFromJSONTyped(json, false);
}
export function MetricsServiceMetricInfoFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'name': json['name'] == null ? undefined : json['name'],
        'displayName': json['displayName'] == null ? undefined : json['displayName'],
        'projectId': json['projectId'] == null ? undefined : json['projectId'],
        'contractName': json['contractName'] == null ? undefined : json['contractName'],
        'contractAddress': json['contractAddress'] == null ? undefined : json['contractAddress'],
        'chainId': json['chainId'] == null ? undefined : json['chainId'],
        'labels': json['labels'] == null ? undefined : (mapValues(json['labels'], MetricsServiceMetricInfoLabelValuesFromJSON)),
        'metadata': json['metadata'] == null ? undefined : MetricsServiceMetricMetadataFromJSON(json['metadata']),
    };
}
export function MetricsServiceMetricInfoToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'name': value['name'],
        'displayName': value['displayName'],
        'projectId': value['projectId'],
        'contractName': value['contractName'],
        'contractAddress': value['contractAddress'],
        'chainId': value['chainId'],
        'labels': value['labels'] == null ? undefined : (mapValues(value['labels'], MetricsServiceMetricInfoLabelValuesToJSON)),
        'metadata': MetricsServiceMetricMetadataToJSON(value['metadata']),
    };
}
