// @sentio/api/dist/src/models/MetricsServiceMetricsQueryResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { MetricsServiceMetricsQueryResponseResultFromJSON, MetricsServiceMetricsQueryResponseResultToJSON, } from './MetricsServiceMetricsQueryResponseResult.js';
/**
 * Check if a given object implements the MetricsServiceMetricsQueryResponse interface.
 */
export function instanceOfMetricsServiceMetricsQueryResponse(value) {
    return true;
}
export function MetricsServiceMetricsQueryResponseFromJSON(json) {
    return MetricsServiceMetricsQueryResponseFromJSONTyped(json, false);
}
export function MetricsServiceMetricsQueryResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'results': json['results'] == null ? undefined : (json['results'].map(MetricsServiceMetricsQueryResponseResultFromJSON)),
    };
}
export function MetricsServiceMetricsQueryResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'results': value['results'] == null ? undefined : (value['results'].map(MetricsServiceMetricsQueryResponseResultToJSON)),
    };
}
