// @sentio/api/dist/src/models/MetricsServiceMetricsQueryResponseValue.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the MetricsServiceMetricsQueryResponseValue interface.
 */
export function instanceOfMetricsServiceMetricsQueryResponseValue(value) {
    return true;
}
export function MetricsServiceMetricsQueryResponseValueFromJSON(json) {
    return MetricsServiceMetricsQueryResponseValueFromJSONTyped(json, false);
}
export function MetricsServiceMetricsQueryResponseValueFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'timestamp': json['timestamp'] == null ? undefined : json['timestamp'],
        'value': json['value'] == null ? undefined : json['value'],
    };
}
export function MetricsServiceMetricsQueryResponseValueToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'timestamp': value['timestamp'],
        'value': value['value'],
    };
}
