// @sentio/api/dist/src/models/MetricsServiceObservabilityServiceQueryBody.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonFormulaFromJSON, CommonFormulaToJSON, } from './CommonFormula.js';
import { CommonQueryFromJSON, CommonQueryToJSON, } from './CommonQuery.js';
/**
 * Check if a given object implements the MetricsServiceObservabilityServiceQueryBody interface.
 */
export function instanceOfMetricsServiceObservabilityServiceQueryBody(value) {
    return true;
}
export function MetricsServiceObservabilityServiceQueryBodyFromJSON(json) {
    return MetricsServiceObservabilityServiceQueryBodyFromJSONTyped(json, false);
}
export function MetricsServiceObservabilityServiceQueryBodyFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'queries': json['queries'] == null ? undefined : (json['queries'].map(CommonQueryFromJSON)),
        'formulas': json['formulas'] == null ? undefined : (json['formulas'].map(CommonFormulaFromJSON)),
        'time': json['time'] == null ? undefined : json['time'],
        'samplesLimit': json['samplesLimit'] == null ? undefined : json['samplesLimit'],
        'version': json['version'] == null ? undefined : json['version'],
        'timezone': json['timezone'] == null ? undefined : json['timezone'],
        'samplesOffset': json['samplesOffset'] == null ? undefined : json['samplesOffset'],
    };
}
export function MetricsServiceObservabilityServiceQueryBodyToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'queries': value['queries'] == null ? undefined : (value['queries'].map(CommonQueryToJSON)),
        'formulas': value['formulas'] == null ? undefined : (value['formulas'].map(CommonFormulaToJSON)),
        'time': value['time'],
        'samplesLimit': value['samplesLimit'],
        'version': value['version'],
        'timezone': value['timezone'],
        'samplesOffset': value['samplesOffset'],
    };
}
