// @sentio/api/dist/src/models/MetricsServiceMetricsQueryResponseSample.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { MetricsServiceMetricsQueryResponseMetricFromJSON, MetricsServiceMetricsQueryResponseMetricToJSON, } from './MetricsServiceMetricsQueryResponseMetric.js';
import { MetricsServiceMetricsQueryResponseValueFromJSON, MetricsServiceMetricsQueryResponseValueToJSON, } from './MetricsServiceMetricsQueryResponseValue.js';
/**
 * Check if a given object implements the MetricsServiceMetricsQueryResponseSample interface.
 */
export function instanceOfMetricsServiceMetricsQueryResponseSample(value) {
    return true;
}
export function MetricsServiceMetricsQueryResponseSampleFromJSON(json) {
    return MetricsServiceMetricsQueryResponseSampleFromJSONTyped(json, false);
}
export function MetricsServiceMetricsQueryResponseSampleFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'metric': json['metric'] == null ? undefined : MetricsServiceMetricsQueryResponseMetricFromJSON(json['metric']),
        'values': json['values'] == null ? undefined : (json['values'].map(MetricsServiceMetricsQueryResponseValueFromJSON)),
    };
}
export function MetricsServiceMetricsQueryResponseSampleToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'metric': MetricsServiceMetricsQueryResponseMetricToJSON(value['metric']),
        'values': value['values'] == null ? undefined : (value['values'].map(MetricsServiceMetricsQueryResponseValueToJSON)),
    };
}
