// @sentio/api/dist/src/models/MetricsServiceGetMetricsResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { MetricsServiceMetricInfoFromJSON, MetricsServiceMetricInfoToJSON, } from './MetricsServiceMetricInfo.js';
/**
 * Check if a given object implements the MetricsServiceGetMetricsResponse interface.
 */
export function instanceOfMetricsServiceGetMetricsResponse(value) {
    return true;
}
export function MetricsServiceGetMetricsResponseFromJSON(json) {
    return MetricsServiceGetMetricsResponseFromJSONTyped(json, false);
}
export function MetricsServiceGetMetricsResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'metrics': json['metrics'] == null ? undefined : (json['metrics'].map(MetricsServiceMetricInfoFromJSON)),
    };
}
export function MetricsServiceGetMetricsResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'metrics': value['metrics'] == null ? undefined : (value['metrics'].map(MetricsServiceMetricInfoToJSON)),
    };
}
