// @sentio/api/dist/src/models/MetricsServiceQueryValueResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { MetricsServiceQueryValueResponseResultFromJSON, MetricsServiceQueryValueResponseResultToJSON, } from './MetricsServiceQueryValueResponseResult.js';
/**
 * Check if a given object implements the MetricsServiceQueryValueResponse interface.
 */
export function instanceOfMetricsServiceQueryValueResponse(value) {
    return true;
}
export function MetricsServiceQueryValueResponseFromJSON(json) {
    return MetricsServiceQueryValueResponseFromJSONTyped(json, false);
}
export function MetricsServiceQueryValueResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'results': json['results'] == null ? undefined : (json['results'].map(MetricsServiceQueryValueResponseResultFromJSON)),
    };
}
export function MetricsServiceQueryValueResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'results': value['results'] == null ? undefined : (value['results'].map(MetricsServiceQueryValueResponseResultToJSON)),
    };
}
