// @sentio/api/dist/src/models/PriceServiceBatchGetPricesResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { PriceServiceBatchGetPricesResponseCoinPriceFromJSON, PriceServiceBatchGetPricesResponseCoinPriceToJSON, } from './PriceServiceBatchGetPricesResponseCoinPrice.js';
/**
 * Check if a given object implements the PriceServiceBatchGetPricesResponse interface.
 */
export function instanceOfPriceServiceBatchGetPricesResponse(value) {
    return true;
}
export function PriceServiceBatchGetPricesResponseFromJSON(json) {
    return PriceServiceBatchGetPricesResponseFromJSONTyped(json, false);
}
export function PriceServiceBatchGetPricesResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'prices': json['prices'] == null ? undefined : (json['prices'].map(PriceServiceBatchGetPricesResponseCoinPriceFromJSON)),
    };
}
export function PriceServiceBatchGetPricesResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'prices': value['prices'] == null ? undefined : (value['prices'].map(PriceServiceBatchGetPricesResponseCoinPriceToJSON)),
    };
}
