// @sentio/api/dist/src/models/PriceServiceCheckLatestPriceResponseCoinPrice.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { PriceServiceCoinIDFromJSON, PriceServiceCoinIDToJSON, } from './PriceServiceCoinID.js';
/**
 * Check if a given object implements the PriceServiceCheckLatestPriceResponseCoinPrice interface.
 */
export function instanceOfPriceServiceCheckLatestPriceResponseCoinPrice(value) {
    return true;
}
export function PriceServiceCheckLatestPriceResponseCoinPriceFromJSON(json) {
    return PriceServiceCheckLatestPriceResponseCoinPriceFromJSONTyped(json, false);
}
export function PriceServiceCheckLatestPriceResponseCoinPriceFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'coinId': json['coinId'] == null ? undefined : PriceServiceCoinIDFromJSON(json['coinId']),
        'price': json['price'] == null ? undefined : json['price'],
        'timestamp': json['timestamp'] == null ? undefined : (new Date(json['timestamp'])),
    };
}
export function PriceServiceCheckLatestPriceResponseCoinPriceToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'coinId': PriceServiceCoinIDToJSON(value['coinId']),
        'price': value['price'],
        'timestamp': value['timestamp'] == null ? undefined : ((value['timestamp']).toISOString()),
    };
}
