// @sentio/api/dist/src/models/PriceServiceBatchGetPricesResponseCoinPricePrice.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { PriceServiceGetPriceResponseFromJSON, PriceServiceGetPriceResponseToJSON, } from './PriceServiceGetPriceResponse.js';
/**
 * Check if a given object implements the PriceServiceBatchGetPricesResponseCoinPricePrice interface.
 */
export function instanceOfPriceServiceBatchGetPricesResponseCoinPricePrice(value) {
    return true;
}
export function PriceServiceBatchGetPricesResponseCoinPricePriceFromJSON(json) {
    return PriceServiceBatchGetPricesResponseCoinPricePriceFromJSONTyped(json, false);
}
export function PriceServiceBatchGetPricesResponseCoinPricePriceFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'results': json['results'] == null ? undefined : (json['results'].map(PriceServiceGetPriceResponseFromJSON)),
    };
}
export function PriceServiceBatchGetPricesResponseCoinPricePriceToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'results': value['results'] == null ? undefined : (value['results'].map(PriceServiceGetPriceResponseToJSON)),
    };
}
