// @sentio/api/dist/src/models/PriceServiceBatchGetPricesResponseCoinPrice.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { PriceServiceBatchGetPricesResponseCoinPricePriceFromJSON, PriceServiceBatchGetPricesResponseCoinPricePriceToJSON, } from './PriceServiceBatchGetPricesResponseCoinPricePrice.js';
import { PriceServiceCoinIDFromJSON, PriceServiceCoinIDToJSON, } from './PriceServiceCoinID.js';
/**
 * Check if a given object implements the PriceServiceBatchGetPricesResponseCoinPrice interface.
 */
export function instanceOfPriceServiceBatchGetPricesResponseCoinPrice(value) {
    return true;
}
export function PriceServiceBatchGetPricesResponseCoinPriceFromJSON(json) {
    return PriceServiceBatchGetPricesResponseCoinPriceFromJSONTyped(json, false);
}
export function PriceServiceBatchGetPricesResponseCoinPriceFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'coinId': json['coinId'] == null ? undefined : PriceServiceCoinIDFromJSON(json['coinId']),
        'price': json['price'] == null ? undefined : PriceServiceBatchGetPricesResponseCoinPricePriceFromJSON(json['price']),
        'error': json['error'] == null ? undefined : json['error'],
    };
}
export function PriceServiceBatchGetPricesResponseCoinPriceToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'coinId': PriceServiceCoinIDToJSON(value['coinId']),
        'price': PriceServiceBatchGetPricesResponseCoinPricePriceToJSON(value['price']),
        'error': value['error'],
    };
}
