// @sentio/api/dist/src/models/PriceServiceCoinID.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { PriceServiceCoinIDAddressIdentifierFromJSON, PriceServiceCoinIDAddressIdentifierToJSON, } from './PriceServiceCoinIDAddressIdentifier.js';
/**
 * Check if a given object implements the PriceServiceCoinID interface.
 */
export function instanceOfPriceServiceCoinID(value) {
    return true;
}
export function PriceServiceCoinIDFromJSON(json) {
    return PriceServiceCoinIDFromJSONTyped(json, false);
}
export function PriceServiceCoinIDFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'symbol': json['symbol'] == null ? undefined : json['symbol'],
        'address': json['address'] == null ? undefined : PriceServiceCoinIDAddressIdentifierFromJSON(json['address']),
    };
}
export function PriceServiceCoinIDToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'symbol': value['symbol'],
        'address': PriceServiceCoinIDAddressIdentifierToJSON(value['address']),
    };
}
