// @sentio/api/dist/src/models/PriceServiceCheckLatestPriceResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { PriceServiceCheckLatestPriceResponseCoinPriceFromJSON, PriceServiceCheckLatestPriceResponseCoinPriceToJSON, } from './PriceServiceCheckLatestPriceResponseCoinPrice.js';
/**
 * Check if a given object implements the PriceServiceCheckLatestPriceResponse interface.
 */
export function instanceOfPriceServiceCheckLatestPriceResponse(value) {
    return true;
}
export function PriceServiceCheckLatestPriceResponseFromJSON(json) {
    return PriceServiceCheckLatestPriceResponseFromJSONTyped(json, false);
}
export function PriceServiceCheckLatestPriceResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'prices': json['prices'] == null ? undefined : (json['prices'].map(PriceServiceCheckLatestPriceResponseCoinPriceFromJSON)),
        'latestPrice': json['latestPrice'] == null ? undefined : PriceServiceCheckLatestPriceResponseCoinPriceFromJSON(json['latestPrice']),
    };
}
export function PriceServiceCheckLatestPriceResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'prices': value['prices'] == null ? undefined : (value['prices'].map(PriceServiceCheckLatestPriceResponseCoinPriceToJSON)),
        'latestPrice': PriceServiceCheckLatestPriceResponseCoinPriceToJSON(value['latestPrice']),
    };
}
