// @sentio/api/dist/src/models/PriceServiceListCoinsResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { mapValues } from '../runtime.js';
import { PriceServiceCoinIDFromJSON, PriceServiceCoinIDToJSON, } from './PriceServiceCoinID.js';
/**
 * Check if a given object implements the PriceServiceListCoinsResponse interface.
 */
export function instanceOfPriceServiceListCoinsResponse(value) {
    return true;
}
export function PriceServiceListCoinsResponseFromJSON(json) {
    return PriceServiceListCoinsResponseFromJSONTyped(json, false);
}
export function PriceServiceListCoinsResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'coins': json['coins'] == null ? undefined : (json['coins'].map(PriceServiceCoinIDFromJSON)),
        'coinAddressesInChain': json['coinAddressesInChain'] == null ? undefined : (mapValues(json['coinAddressesInChain'], PriceServiceCoinIDFromJSON)),
    };
}
export function PriceServiceListCoinsResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'coins': value['coins'] == null ? undefined : (value['coins'].map(PriceServiceCoinIDToJSON)),
        'coinAddressesInChain': value['coinAddressesInChain'] == null ? undefined : (mapValues(value['coinAddressesInChain'], PriceServiceCoinIDToJSON)),
    };
}
