// @sentio/api/dist/src/models/PriceServiceGetPriceResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the PriceServiceGetPriceResponse interface.
 */
export function instanceOfPriceServiceGetPriceResponse(value) {
    return true;
}
export function PriceServiceGetPriceResponseFromJSON(json) {
    return PriceServiceGetPriceResponseFromJSONTyped(json, false);
}
export function PriceServiceGetPriceResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'price': json['price'] == null ? undefined : json['price'],
        'timestamp': json['timestamp'] == null ? undefined : (new Date(json['timestamp'])),
    };
}
export function PriceServiceGetPriceResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'price': value['price'],
        'timestamp': value['timestamp'] == null ? undefined : ((value['timestamp']).toISOString()),
    };
}
