// @sentio/api/dist/src/models/SolidityServiceTxIdentifier.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the SolidityServiceTxIdentifier interface.
 */
export function instanceOfSolidityServiceTxIdentifier(value) {
    return true;
}
export function SolidityServiceTxIdentifierFromJSON(json) {
    return SolidityServiceTxIdentifierFromJSONTyped(json, false);
}
export function SolidityServiceTxIdentifierFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'txHash': json['txHash'] == null ? undefined : json['txHash'],
        'simulationId': json['simulationId'] == null ? undefined : json['simulationId'],
        'bundleId': json['bundleId'] == null ? undefined : json['bundleId'],
    };
}
export function SolidityServiceTxIdentifierToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'txHash': value['txHash'],
        'simulationId': value['simulationId'],
        'bundleId': value['bundleId'],
    };
}
