// @sentio/api/dist/src/models/SolidityServiceStorageSummaryResult.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the SolidityServiceStorageSummaryResult interface.
 */
export function instanceOfSolidityServiceStorageSummaryResult(value) {
    return true;
}
export function SolidityServiceStorageSummaryResultFromJSON(json) {
    return SolidityServiceStorageSummaryResultFromJSONTyped(json, false);
}
export function SolidityServiceStorageSummaryResultFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'address': json['address'] == null ? undefined : json['address'],
        'codeAddress': json['codeAddress'] == null ? undefined : json['codeAddress'],
        'stateVariables': json['stateVariables'] == null ? undefined : json['stateVariables'],
    };
}
export function SolidityServiceStorageSummaryResultToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'address': value['address'],
        'codeAddress': value['codeAddress'],
        'stateVariables': value['stateVariables'],
    };
}
