// @sentio/api/dist/src/models/SolidityServiceSettings.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { SolidityServiceMetadataFromJSON, SolidityServiceMetadataToJSON, } from './SolidityServiceMetadata.js';
import { SolidityServiceOptimizerFromJSON, SolidityServiceOptimizerToJSON, } from './SolidityServiceOptimizer.js';
/**
 * Check if a given object implements the SolidityServiceSettings interface.
 */
export function instanceOfSolidityServiceSettings(value) {
    return true;
}
export function SolidityServiceSettingsFromJSON(json) {
    return SolidityServiceSettingsFromJSONTyped(json, false);
}
export function SolidityServiceSettingsFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'remappings': json['remappings'] == null ? undefined : json['remappings'],
        'optimizer': json['optimizer'] == null ? undefined : SolidityServiceOptimizerFromJSON(json['optimizer']),
        'evmVersion': json['evmVersion'] == null ? undefined : json['evmVersion'],
        'metadata': json['metadata'] == null ? undefined : SolidityServiceMetadataFromJSON(json['metadata']),
        'viaIR': json['viaIR'] == null ? undefined : json['viaIR'],
        'compilationTarget': json['compilationTarget'] == null ? undefined : json['compilationTarget'],
    };
}
export function SolidityServiceSettingsToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'remappings': value['remappings'],
        'optimizer': SolidityServiceOptimizerToJSON(value['optimizer']),
        'evmVersion': value['evmVersion'],
        'metadata': SolidityServiceMetadataToJSON(value['metadata']),
        'viaIR': value['viaIR'],
        'compilationTarget': value['compilationTarget'],
    };
}
