// @sentio/api/dist/src/models/SolidityServiceOptimizer.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { SolidityServiceOptimizerDetailsFromJSON, SolidityServiceOptimizerDetailsToJSON, } from './SolidityServiceOptimizerDetails.js';
/**
 * Check if a given object implements the SolidityServiceOptimizer interface.
 */
export function instanceOfSolidityServiceOptimizer(value) {
    return true;
}
export function SolidityServiceOptimizerFromJSON(json) {
    return SolidityServiceOptimizerFromJSONTyped(json, false);
}
export function SolidityServiceOptimizerFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'enabled': json['enabled'] == null ? undefined : json['enabled'],
        'runs': json['runs'] == null ? undefined : json['runs'],
        'details': json['details'] == null ? undefined : SolidityServiceOptimizerDetailsFromJSON(json['details']),
    };
}
export function SolidityServiceOptimizerToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'enabled': value['enabled'],
        'runs': value['runs'],
        'details': SolidityServiceOptimizerDetailsToJSON(value['details']),
    };
}
