// @sentio/api/dist/src/models/SolidityServiceGetTraceCallInternalRequestStateContext.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the SolidityServiceGetTraceCallInternalRequestStateContext interface.
 */
export function instanceOfSolidityServiceGetTraceCallInternalRequestStateContext(value) {
    return true;
}
export function SolidityServiceGetTraceCallInternalRequestStateContextFromJSON(json) {
    return SolidityServiceGetTraceCallInternalRequestStateContextFromJSONTyped(json, false);
}
export function SolidityServiceGetTraceCallInternalRequestStateContextFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'blockNumber': json['blockNumber'] == null ? undefined : json['blockNumber'],
        'transactionIndex': json['transactionIndex'] == null ? undefined : json['transactionIndex'],
    };
}
export function SolidityServiceGetTraceCallInternalRequestStateContextToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'blockNumber': value['blockNumber'],
        'transactionIndex': value['transactionIndex'],
    };
}
