// @sentio/api/dist/src/models/SolidityServiceGetSimulationBundleResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { SolidityServiceSimulationFromJSON, SolidityServiceSimulationToJSON, } from './SolidityServiceSimulation.js';
/**
 * Check if a given object implements the SolidityServiceGetSimulationBundleResponse interface.
 */
export function instanceOfSolidityServiceGetSimulationBundleResponse(value) {
    return true;
}
export function SolidityServiceGetSimulationBundleResponseFromJSON(json) {
    return SolidityServiceGetSimulationBundleResponseFromJSONTyped(json, false);
}
export function SolidityServiceGetSimulationBundleResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'simulations': json['simulations'] == null ? undefined : (json['simulations'].map(SolidityServiceSimulationFromJSON)),
        'error': json['error'] == null ? undefined : json['error'],
    };
}
export function SolidityServiceGetSimulationBundleResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'simulations': value['simulations'] == null ? undefined : (value['simulations'].map(SolidityServiceSimulationToJSON)),
        'error': value['error'],
    };
}
