// @sentio/api/dist/src/models/SolidityServiceStateOverride.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the SolidityServiceStateOverride interface.
 */
export function instanceOfSolidityServiceStateOverride(value) {
    return true;
}
export function SolidityServiceStateOverrideFromJSON(json) {
    return SolidityServiceStateOverrideFromJSONTyped(json, false);
}
export function SolidityServiceStateOverrideFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'state': json['state'] == null ? undefined : json['state'],
        'balance': json['balance'] == null ? undefined : json['balance'],
        'code': json['code'] == null ? undefined : json['code'],
    };
}
export function SolidityServiceStateOverrideToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'state': value['state'],
        'balance': value['balance'],
        'code': value['code'],
    };
}
