// @sentio/api/dist/src/models/SolidityServiceSimulateTransactionBundleResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { SolidityServiceSimulationFromJSON, SolidityServiceSimulationToJSON, } from './SolidityServiceSimulation.js';
/**
 * Check if a given object implements the SolidityServiceSimulateTransactionBundleResponse interface.
 */
export function instanceOfSolidityServiceSimulateTransactionBundleResponse(value) {
    return true;
}
export function SolidityServiceSimulateTransactionBundleResponseFromJSON(json) {
    return SolidityServiceSimulateTransactionBundleResponseFromJSONTyped(json, false);
}
export function SolidityServiceSimulateTransactionBundleResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'bundleId': json['bundleId'] == null ? undefined : json['bundleId'],
        'simulations': json['simulations'] == null ? undefined : (json['simulations'].map(SolidityServiceSimulationFromJSON)),
        'error': json['error'] == null ? undefined : json['error'],
    };
}
export function SolidityServiceSimulateTransactionBundleResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'bundleId': value['bundleId'],
        'simulations': value['simulations'] == null ? undefined : (value['simulations'].map(SolidityServiceSimulationToJSON)),
        'error': value['error'],
    };
}
