// @sentio/api/dist/src/models/SolidityServiceDecodeStateDiffResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the SolidityServiceDecodeStateDiffResponse interface.
 */
export function instanceOfSolidityServiceDecodeStateDiffResponse(value) {
    return true;
}
export function SolidityServiceDecodeStateDiffResponseFromJSON(json) {
    return SolidityServiceDecodeStateDiffResponseFromJSONTyped(json, false);
}
export function SolidityServiceDecodeStateDiffResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'result': json['result'] == null ? undefined : json['result'],
    };
}
export function SolidityServiceDecodeStateDiffResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'result': value['result'],
    };
}
