// @sentio/api/dist/src/models/SolidityServiceOptimizerDetailsYulDetails.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the SolidityServiceOptimizerDetailsYulDetails interface.
 */
export function instanceOfSolidityServiceOptimizerDetailsYulDetails(value) {
    return true;
}
export function SolidityServiceOptimizerDetailsYulDetailsFromJSON(json) {
    return SolidityServiceOptimizerDetailsYulDetailsFromJSONTyped(json, false);
}
export function SolidityServiceOptimizerDetailsYulDetailsFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'stackAllocation': json['stackAllocation'] == null ? undefined : json['stackAllocation'],
        'optimizerSteps': json['optimizerSteps'] == null ? undefined : json['optimizerSteps'],
    };
}
export function SolidityServiceOptimizerDetailsYulDetailsToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'stackAllocation': value['stackAllocation'],
        'optimizerSteps': value['optimizerSteps'],
    };
}
