// @sentio/api/dist/src/models/SolidityServiceOptimizerDetails.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { SolidityServiceOptimizerDetailsYulDetailsFromJSON, SolidityServiceOptimizerDetailsYulDetailsToJSON, } from './SolidityServiceOptimizerDetailsYulDetails.js';
/**
 * Check if a given object implements the SolidityServiceOptimizerDetails interface.
 */
export function instanceOfSolidityServiceOptimizerDetails(value) {
    return true;
}
export function SolidityServiceOptimizerDetailsFromJSON(json) {
    return SolidityServiceOptimizerDetailsFromJSONTyped(json, false);
}
export function SolidityServiceOptimizerDetailsFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'peephole': json['peephole'] == null ? undefined : json['peephole'],
        'jumpdestRemover': json['jumpdestRemover'] == null ? undefined : json['jumpdestRemover'],
        'yul': json['yul'] == null ? undefined : json['yul'],
        'yulDetails': json['yulDetails'] == null ? undefined : SolidityServiceOptimizerDetailsYulDetailsFromJSON(json['yulDetails']),
    };
}
export function SolidityServiceOptimizerDetailsToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'peephole': value['peephole'],
        'jumpdestRemover': value['jumpdestRemover'],
        'yul': value['yul'],
        'yulDetails': SolidityServiceOptimizerDetailsYulDetailsToJSON(value['yulDetails']),
    };
}
