// @sentio/api/dist/src/models/SolidityServiceSourceSpec.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { SolidityServiceSourceMultiFileFromJSON, SolidityServiceSourceMultiFileToJSON, } from './SolidityServiceSourceMultiFile.js';
/**
 * Check if a given object implements the SolidityServiceSourceSpec interface.
 */
export function instanceOfSolidityServiceSourceSpec(value) {
    return true;
}
export function SolidityServiceSourceSpecFromJSON(json) {
    return SolidityServiceSourceSpecFromJSONTyped(json, false);
}
export function SolidityServiceSourceSpecFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'id': json['id'] == null ? undefined : json['id'],
        'multiFile': json['multiFile'] == null ? undefined : SolidityServiceSourceMultiFileFromJSON(json['multiFile']),
        'standardJson': json['standardJson'] == null ? undefined : json['standardJson'],
        'metadata': json['metadata'] == null ? undefined : json['metadata'],
        'solidityVersion': json['solidityVersion'] == null ? undefined : json['solidityVersion'],
        'contractName': json['contractName'] == null ? undefined : json['contractName'],
        'constructorArgs': json['constructorArgs'] == null ? undefined : json['constructorArgs'],
    };
}
export function SolidityServiceSourceSpecToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'id': value['id'],
        'multiFile': SolidityServiceSourceMultiFileToJSON(value['multiFile']),
        'standardJson': value['standardJson'],
        'metadata': value['metadata'],
        'solidityVersion': value['solidityVersion'],
        'contractName': value['contractName'],
        'constructorArgs': value['constructorArgs'],
    };
}
