// @sentio/api/dist/src/models/SolidityServiceSpecializations.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the SolidityServiceSpecializations interface.
 */
export function instanceOfSolidityServiceSpecializations(value) {
    return true;
}
export function SolidityServiceSpecializationsFromJSON(json) {
    return SolidityServiceSpecializationsFromJSONTyped(json, false);
}
export function SolidityServiceSpecializationsFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'constructorArguments': json['constructorArguments'] == null ? undefined : json['constructorArguments'],
    };
}
export function SolidityServiceSpecializationsToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'constructorArguments': value['constructorArguments'],
    };
}
