// @sentio/api/dist/src/models/SolidityServiceSourceMultiFile.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the SolidityServiceSourceMultiFile interface.
 */
export function instanceOfSolidityServiceSourceMultiFile(value) {
    return true;
}
export function SolidityServiceSourceMultiFileFromJSON(json) {
    return SolidityServiceSourceMultiFileFromJSONTyped(json, false);
}
export function SolidityServiceSourceMultiFileFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'source': json['source'] == null ? undefined : json['source'],
        'compilerSettings': json['compilerSettings'] == null ? undefined : json['compilerSettings'],
    };
}
export function SolidityServiceSourceMultiFileToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'source': value['source'],
        'compilerSettings': value['compilerSettings'],
    };
}
