// @sentio/api/dist/src/models/SolidityServiceCompilerOptions.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { SolidityServiceSettingsFromJSON, SolidityServiceSettingsToJSON, } from './SolidityServiceSettings.js';
import { SolidityServiceSpecializationsFromJSON, SolidityServiceSpecializationsToJSON, } from './SolidityServiceSpecializations.js';
/**
 * Check if a given object implements the SolidityServiceCompilerOptions interface.
 */
export function instanceOfSolidityServiceCompilerOptions(value) {
    return true;
}
export function SolidityServiceCompilerOptionsFromJSON(json) {
    return SolidityServiceCompilerOptionsFromJSONTyped(json, false);
}
export function SolidityServiceCompilerOptionsFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'language': json['language'] == null ? undefined : json['language'],
        'version': json['version'] == null ? undefined : json['version'],
        'settings': json['settings'] == null ? undefined : SolidityServiceSettingsFromJSON(json['settings']),
        'specializations': json['specializations'] == null ? undefined : SolidityServiceSpecializationsFromJSON(json['specializations']),
    };
}
export function SolidityServiceCompilerOptionsToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'language': value['language'],
        'version': value['version'],
        'settings': SolidityServiceSettingsToJSON(value['settings']),
        'specializations': SolidityServiceSpecializationsToJSON(value['specializations']),
    };
}
