// @sentio/api/dist/src/models/SolidityServiceGetTraceCallInternalRequestBundle.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { SolidityServiceBlockOverridesFromJSON, SolidityServiceBlockOverridesToJSON, } from './SolidityServiceBlockOverrides.js';
/**
 * Check if a given object implements the SolidityServiceGetTraceCallInternalRequestBundle interface.
 */
export function instanceOfSolidityServiceGetTraceCallInternalRequestBundle(value) {
    return true;
}
export function SolidityServiceGetTraceCallInternalRequestBundleFromJSON(json) {
    return SolidityServiceGetTraceCallInternalRequestBundleFromJSONTyped(json, false);
}
export function SolidityServiceGetTraceCallInternalRequestBundleFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'transactions': json['transactions'] == null ? undefined : json['transactions'],
        'blockOverride': json['blockOverride'] == null ? undefined : SolidityServiceBlockOverridesFromJSON(json['blockOverride']),
    };
}
export function SolidityServiceGetTraceCallInternalRequestBundleToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'transactions': value['transactions'],
        'blockOverride': SolidityServiceBlockOverridesToJSON(value['blockOverride']),
    };
}
