// @sentio/api/dist/src/models/SolidityServiceFetchAndCompileInternalResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { mapValues } from '../runtime.js';
import { SolidityServiceFailureFromJSON, SolidityServiceFailureToJSON, } from './SolidityServiceFailure.js';
/**
 * Check if a given object implements the SolidityServiceFetchAndCompileInternalResponse interface.
 */
export function instanceOfSolidityServiceFetchAndCompileInternalResponse(value) {
    return true;
}
export function SolidityServiceFetchAndCompileInternalResponseFromJSON(json) {
    return SolidityServiceFetchAndCompileInternalResponseFromJSONTyped(json, false);
}
export function SolidityServiceFetchAndCompileInternalResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'compiledAddresses': json['compiledAddresses'] == null ? undefined : json['compiledAddresses'],
        'failures': json['failures'] == null ? undefined : (mapValues(json['failures'], SolidityServiceFailureFromJSON)),
    };
}
export function SolidityServiceFetchAndCompileInternalResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'compiledAddresses': value['compiledAddresses'],
        'failures': value['failures'] == null ? undefined : (mapValues(value['failures'], SolidityServiceFailureToJSON)),
    };
}
