// @sentio/api/dist/src/models/SolidityServiceSimulateTransactionResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { SolidityServiceSimulationFromJSON, SolidityServiceSimulationToJSON, } from './SolidityServiceSimulation.js';
/**
 * Check if a given object implements the SolidityServiceSimulateTransactionResponse interface.
 */
export function instanceOfSolidityServiceSimulateTransactionResponse(value) {
    return true;
}
export function SolidityServiceSimulateTransactionResponseFromJSON(json) {
    return SolidityServiceSimulateTransactionResponseFromJSONTyped(json, false);
}
export function SolidityServiceSimulateTransactionResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'simulation': json['simulation'] == null ? undefined : SolidityServiceSimulationFromJSON(json['simulation']),
    };
}
export function SolidityServiceSimulateTransactionResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'simulation': SolidityServiceSimulationToJSON(value['simulation']),
    };
}
