// @sentio/api/dist/src/models/SolidityServiceCompileSourceInternalResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { SolidityServiceFailureFromJSON, SolidityServiceFailureToJSON, } from './SolidityServiceFailure.js';
/**
 * Check if a given object implements the SolidityServiceCompileSourceInternalResponse interface.
 */
export function instanceOfSolidityServiceCompileSourceInternalResponse(value) {
    return true;
}
export function SolidityServiceCompileSourceInternalResponseFromJSON(json) {
    return SolidityServiceCompileSourceInternalResponseFromJSONTyped(json, false);
}
export function SolidityServiceCompileSourceInternalResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'failure': json['failure'] == null ? undefined : SolidityServiceFailureFromJSON(json['failure']),
    };
}
export function SolidityServiceCompileSourceInternalResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'failure': SolidityServiceFailureToJSON(value['failure']),
    };
}
