// @sentio/api/dist/src/models/SolidityServiceSimulationResult.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { EvmTransactionFromJSON, EvmTransactionToJSON, } from './EvmTransaction.js';
import { EvmTransactionReceiptFromJSON, EvmTransactionReceiptToJSON, } from './EvmTransactionReceipt.js';
/**
 * Check if a given object implements the SolidityServiceSimulationResult interface.
 */
export function instanceOfSolidityServiceSimulationResult(value) {
    return true;
}
export function SolidityServiceSimulationResultFromJSON(json) {
    return SolidityServiceSimulationResultFromJSONTyped(json, false);
}
export function SolidityServiceSimulationResultFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'transaction': json['transaction'] == null ? undefined : EvmTransactionFromJSON(json['transaction']),
        'transactionReceipt': json['transactionReceipt'] == null ? undefined : EvmTransactionReceiptFromJSON(json['transactionReceipt']),
    };
}
export function SolidityServiceSimulationResultToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'transaction': EvmTransactionToJSON(value['transaction']),
        'transactionReceipt': EvmTransactionReceiptToJSON(value['transactionReceipt']),
    };
}
