// @sentio/api/dist/src/models/SolidityServiceContractKeyInfo.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the SolidityServiceContractKeyInfo interface.
 */
export function instanceOfSolidityServiceContractKeyInfo(value) {
    return true;
}
export function SolidityServiceContractKeyInfoFromJSON(json) {
    return SolidityServiceContractKeyInfoFromJSONTyped(json, false);
}
export function SolidityServiceContractKeyInfoFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'preprocessKey': json['preprocessKey'] == null ? undefined : json['preprocessKey'],
        'compilationKey': json['compilationKey'] == null ? undefined : json['compilationKey'],
    };
}
export function SolidityServiceContractKeyInfoToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'preprocessKey': value['preprocessKey'],
        'compilationKey': value['compilationKey'],
    };
}
