// @sentio/api/dist/src/models/SolidityServiceBlockOverrides.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the SolidityServiceBlockOverrides interface.
 */
export function instanceOfSolidityServiceBlockOverrides(value) {
    return true;
}
export function SolidityServiceBlockOverridesFromJSON(json) {
    return SolidityServiceBlockOverridesFromJSONTyped(json, false);
}
export function SolidityServiceBlockOverridesFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'blockNumber': json['blockNumber'] == null ? undefined : json['blockNumber'],
        'timestamp': json['timestamp'] == null ? undefined : json['timestamp'],
        'gasLimit': json['gasLimit'] == null ? undefined : json['gasLimit'],
        'difficulty': json['difficulty'] == null ? undefined : json['difficulty'],
        'baseFee': json['baseFee'] == null ? undefined : json['baseFee'],
        'blockHash': json['blockHash'] == null ? undefined : json['blockHash'],
    };
}
export function SolidityServiceBlockOverridesToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'blockNumber': value['blockNumber'],
        'timestamp': value['timestamp'],
        'gasLimit': value['gasLimit'],
        'difficulty': value['difficulty'],
        'baseFee': value['baseFee'],
        'blockHash': value['blockHash'],
    };
}
