// @sentio/api/dist/src/models/SolidityServiceSimulateTransactionRequest.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { SolidityServiceSimulationFromJSON, SolidityServiceSimulationToJSON, } from './SolidityServiceSimulation.js';
/**
 * Check if a given object implements the SolidityServiceSimulateTransactionRequest interface.
 */
export function instanceOfSolidityServiceSimulateTransactionRequest(value) {
    if (!('projectOwner' in value))
        return false;
    if (!('projectSlug' in value))
        return false;
    if (!('simulation' in value))
        return false;
    return true;
}
export function SolidityServiceSimulateTransactionRequestFromJSON(json) {
    return SolidityServiceSimulateTransactionRequestFromJSONTyped(json, false);
}
export function SolidityServiceSimulateTransactionRequestFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'projectOwner': json['projectOwner'],
        'projectSlug': json['projectSlug'],
        'simulation': SolidityServiceSimulationFromJSON(json['simulation']),
    };
}
export function SolidityServiceSimulateTransactionRequestToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'projectOwner': value['projectOwner'],
        'projectSlug': value['projectSlug'],
        'simulation': SolidityServiceSimulationToJSON(value['simulation']),
    };
}
