// @sentio/api/dist/src/models/SolidityServiceGetSimulationsResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { SolidityServiceSimulationFromJSON, SolidityServiceSimulationToJSON, } from './SolidityServiceSimulation.js';
/**
 * Check if a given object implements the SolidityServiceGetSimulationsResponse interface.
 */
export function instanceOfSolidityServiceGetSimulationsResponse(value) {
    return true;
}
export function SolidityServiceGetSimulationsResponseFromJSON(json) {
    return SolidityServiceGetSimulationsResponseFromJSONTyped(json, false);
}
export function SolidityServiceGetSimulationsResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'simulations': json['simulations'] == null ? undefined : (json['simulations'].map(SolidityServiceSimulationFromJSON)),
        'count': json['count'] == null ? undefined : json['count'],
        'page': json['page'] == null ? undefined : json['page'],
        'pageSize': json['pageSize'] == null ? undefined : json['pageSize'],
    };
}
export function SolidityServiceGetSimulationsResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'simulations': value['simulations'] == null ? undefined : (value['simulations'].map(SolidityServiceSimulationToJSON)),
        'count': value['count'],
        'page': value['page'],
        'pageSize': value['pageSize'],
    };
}
