// @sentio/api/dist/src/models/SolidityServiceGetTraceCallInternalResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the SolidityServiceGetTraceCallInternalResponse interface.
 */
export function instanceOfSolidityServiceGetTraceCallInternalResponse(value) {
    return true;
}
export function SolidityServiceGetTraceCallInternalResponseFromJSON(json) {
    return SolidityServiceGetTraceCallInternalResponseFromJSONTyped(json, false);
}
export function SolidityServiceGetTraceCallInternalResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'result': json['result'] == null ? undefined : json['result'],
        'outputs': json['outputs'] == null ? undefined : json['outputs'],
        'error': json['error'] == null ? undefined : json['error'],
    };
}
export function SolidityServiceGetTraceCallInternalResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'result': value['result'],
        'outputs': value['outputs'],
        'error': value['error'],
    };
}
