// @sentio/api/dist/src/models/SolidityServiceGetStorageSummaryInternalResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { SolidityServiceStorageSummaryResultFromJSON, SolidityServiceStorageSummaryResultToJSON, } from './SolidityServiceStorageSummaryResult.js';
/**
 * Check if a given object implements the SolidityServiceGetStorageSummaryInternalResponse interface.
 */
export function instanceOfSolidityServiceGetStorageSummaryInternalResponse(value) {
    return true;
}
export function SolidityServiceGetStorageSummaryInternalResponseFromJSON(json) {
    return SolidityServiceGetStorageSummaryInternalResponseFromJSONTyped(json, false);
}
export function SolidityServiceGetStorageSummaryInternalResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'results': json['results'] == null ? undefined : (json['results'].map(SolidityServiceStorageSummaryResultFromJSON)),
    };
}
export function SolidityServiceGetStorageSummaryInternalResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'results': value['results'] == null ? undefined : (value['results'].map(SolidityServiceStorageSummaryResultToJSON)),
    };
}
