// @sentio/api/dist/src/models/SolidityServiceFailure.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the SolidityServiceFailure interface.
 */
export function instanceOfSolidityServiceFailure(value) {
    return true;
}
export function SolidityServiceFailureFromJSON(json) {
    return SolidityServiceFailureFromJSONTyped(json, false);
}
export function SolidityServiceFailureFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'error': json['error'] == null ? undefined : json['error'],
        'reason': json['reason'] == null ? undefined : json['reason'],
    };
}
export function SolidityServiceFailureToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'error': value['error'],
        'reason': value['reason'],
    };
}
