// @sentio/api/dist/src/models/SolidityServiceSimulation.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { mapValues } from '../runtime.js';
import { EvmAccessListItemFromJSON, EvmAccessListItemToJSON, } from './EvmAccessListItem.js';
import { SolidityServiceBlockOverridesFromJSON, SolidityServiceBlockOverridesToJSON, } from './SolidityServiceBlockOverrides.js';
import { SolidityServiceSimulationResultFromJSON, SolidityServiceSimulationResultToJSON, } from './SolidityServiceSimulationResult.js';
import { SolidityServiceStateOverrideFromJSON, SolidityServiceStateOverrideToJSON, } from './SolidityServiceStateOverride.js';
/**
 * Check if a given object implements the SolidityServiceSimulation interface.
 */
export function instanceOfSolidityServiceSimulation(value) {
    if (!('networkId' in value))
        return false;
    if (!('to' in value))
        return false;
    if (!('input' in value))
        return false;
    if (!('blockNumber' in value))
        return false;
    if (!('transactionIndex' in value))
        return false;
    if (!('from' in value))
        return false;
    if (!('gas' in value))
        return false;
    if (!('gasPrice' in value))
        return false;
    if (!('value' in value))
        return false;
    return true;
}
export function SolidityServiceSimulationFromJSON(json) {
    return SolidityServiceSimulationFromJSONTyped(json, false);
}
export function SolidityServiceSimulationFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'id': json['id'] == null ? undefined : json['id'],
        'createAt': json['createAt'] == null ? undefined : json['createAt'],
        'bundleId': json['bundleId'] == null ? undefined : json['bundleId'],
        'networkId': json['networkId'],
        'chainId': json['chainId'] == null ? undefined : json['chainId'],
        'to': json['to'],
        'input': json['input'],
        'blockNumber': json['blockNumber'],
        'transactionIndex': json['transactionIndex'],
        'from': json['from'],
        'gas': json['gas'],
        'gasPrice': json['gasPrice'],
        'maxFeePerGas': json['maxFeePerGas'] == null ? undefined : json['maxFeePerGas'],
        'maxPriorityFeePerGas': json['maxPriorityFeePerGas'] == null ? undefined : json['maxPriorityFeePerGas'],
        'value': json['value'],
        'accessList': json['accessList'] == null ? undefined : (json['accessList'].map(EvmAccessListItemFromJSON)),
        'originTxHash': json['originTxHash'] == null ? undefined : json['originTxHash'],
        'label': json['label'] == null ? undefined : json['label'],
        'stateOverrides': json['stateOverrides'] == null ? undefined : (mapValues(json['stateOverrides'], SolidityServiceStateOverrideFromJSON)),
        'sourceOverrides': json['sourceOverrides'] == null ? undefined : json['sourceOverrides'],
        'blockOverride': json['blockOverride'] == null ? undefined : SolidityServiceBlockOverridesFromJSON(json['blockOverride']),
        'debugDeployment': json['debugDeployment'] == null ? undefined : json['debugDeployment'],
        'result': json['result'] == null ? undefined : SolidityServiceSimulationResultFromJSON(json['result']),
    };
}
export function SolidityServiceSimulationToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'id': value['id'],
        'createAt': value['createAt'],
        'bundleId': value['bundleId'],
        'networkId': value['networkId'],
        'chainId': value['chainId'],
        'to': value['to'],
        'input': value['input'],
        'blockNumber': value['blockNumber'],
        'transactionIndex': value['transactionIndex'],
        'from': value['from'],
        'gas': value['gas'],
        'gasPrice': value['gasPrice'],
        'maxFeePerGas': value['maxFeePerGas'],
        'maxPriorityFeePerGas': value['maxPriorityFeePerGas'],
        'value': value['value'],
        'accessList': value['accessList'] == null ? undefined : (value['accessList'].map(EvmAccessListItemToJSON)),
        'originTxHash': value['originTxHash'],
        'label': value['label'],
        'stateOverrides': value['stateOverrides'] == null ? undefined : (mapValues(value['stateOverrides'], SolidityServiceStateOverrideToJSON)),
        'sourceOverrides': value['sourceOverrides'],
        'blockOverride': SolidityServiceBlockOverridesToJSON(value['blockOverride']),
        'debugDeployment': value['debugDeployment'],
        'result': SolidityServiceSimulationResultToJSON(value['result']),
    };
}
