// @sentio/api/dist/src/models/SolidityServiceSourceInfo.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { SolidityServiceCompilerOptionsFromJSON, SolidityServiceCompilerOptionsToJSON, } from './SolidityServiceCompilerOptions.js';
/**
 * Check if a given object implements the SolidityServiceSourceInfo interface.
 */
export function instanceOfSolidityServiceSourceInfo(value) {
    return true;
}
export function SolidityServiceSourceInfoFromJSON(json) {
    return SolidityServiceSourceInfoFromJSONTyped(json, false);
}
export function SolidityServiceSourceInfoFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'contractName': json['contractName'] == null ? undefined : json['contractName'],
        'options': json['options'] == null ? undefined : SolidityServiceCompilerOptionsFromJSON(json['options']),
    };
}
export function SolidityServiceSourceInfoToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'contractName': value['contractName'],
        'options': SolidityServiceCompilerOptionsToJSON(value['options']),
    };
}
