// @sentio/api/dist/src/models/SolidityServiceMetadata.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the SolidityServiceMetadata interface.
 */
export function instanceOfSolidityServiceMetadata(value) {
    return true;
}
export function SolidityServiceMetadataFromJSON(json) {
    return SolidityServiceMetadataFromJSONTyped(json, false);
}
export function SolidityServiceMetadataFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'useLiteralContent': json['useLiteralContent'] == null ? undefined : json['useLiteralContent'],
        'bytecodeHash': json['bytecodeHash'] == null ? undefined : json['bytecodeHash'],
        'appendCBOR': json['appendCBOR'] == null ? undefined : json['appendCBOR'],
    };
}
export function SolidityServiceMetadataToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'useLiteralContent': value['useLiteralContent'],
        'bytecodeHash': value['bytecodeHash'],
        'appendCBOR': value['appendCBOR'],
    };
}
