// @sentio/api/dist/src/models/SolidityServicePreProcessCompilationInternalResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the SolidityServicePreProcessCompilationInternalResponse interface.
 */
export function instanceOfSolidityServicePreProcessCompilationInternalResponse(value) {
    return true;
}
export function SolidityServicePreProcessCompilationInternalResponseFromJSON(json) {
    return SolidityServicePreProcessCompilationInternalResponseFromJSONTyped(json, false);
}
export function SolidityServicePreProcessCompilationInternalResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'preprocessedAddresses': json['preprocessedAddresses'] == null ? undefined : json['preprocessedAddresses'],
    };
}
export function SolidityServicePreProcessCompilationInternalResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'preprocessedAddresses': value['preprocessedAddresses'],
    };
}
