// @sentio/api/dist/src/models/TxindexEvmSearchTransactionsResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { TxindexEvmRawTransactionFromJSON, TxindexEvmRawTransactionToJSON, } from './TxindexEvmRawTransaction.js';
/**
 * Check if a given object implements the TxindexEvmSearchTransactionsResponse interface.
 */
export function instanceOfTxindexEvmSearchTransactionsResponse(value) {
    return true;
}
export function TxindexEvmSearchTransactionsResponseFromJSON(json) {
    return TxindexEvmSearchTransactionsResponseFromJSONTyped(json, false);
}
export function TxindexEvmSearchTransactionsResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'transactions': json['transactions'] == null ? undefined : (json['transactions'].map(TxindexEvmRawTransactionFromJSON)),
        'nextPageToken': json['nextPageToken'] == null ? undefined : json['nextPageToken'],
    };
}
export function TxindexEvmSearchTransactionsResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'transactions': value['transactions'] == null ? undefined : (value['transactions'].map(TxindexEvmRawTransactionToJSON)),
        'nextPageToken': value['nextPageToken'],
    };
}
