// @sentio/api/dist/src/models/TxindexEvmRawTransaction.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { EvmTransactionFromJSON, EvmTransactionToJSON, } from './EvmTransaction.js';
/**
 * Check if a given object implements the TxindexEvmRawTransaction interface.
 */
export function instanceOfTxindexEvmRawTransaction(value) {
    return true;
}
export function TxindexEvmRawTransactionFromJSON(json) {
    return TxindexEvmRawTransactionFromJSONTyped(json, false);
}
export function TxindexEvmRawTransactionFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'hash': json['hash'] == null ? undefined : json['hash'],
        'blockNumber': json['blockNumber'] == null ? undefined : json['blockNumber'],
        'isIn': json['isIn'] == null ? undefined : json['isIn'],
        'trace': json['trace'] == null ? undefined : json['trace'],
        'tx': json['tx'] == null ? undefined : EvmTransactionFromJSON(json['tx']),
        'json': json['json'] == null ? undefined : json['json'],
        'timestamp': json['timestamp'] == null ? undefined : json['timestamp'],
        'transactionStatus': json['transactionStatus'] == null ? undefined : json['transactionStatus'],
        'methodSignature': json['methodSignature'] == null ? undefined : json['methodSignature'],
        'methodSignatureText': json['methodSignatureText'] == null ? undefined : json['methodSignatureText'],
        'abiItem': json['abiItem'] == null ? undefined : json['abiItem'],
    };
}
export function TxindexEvmRawTransactionToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'hash': value['hash'],
        'blockNumber': value['blockNumber'],
        'isIn': value['isIn'],
        'trace': value['trace'],
        'tx': EvmTransactionToJSON(value['tx']),
        'json': value['json'],
        'timestamp': value['timestamp'],
        'transactionStatus': value['transactionStatus'],
        'methodSignature': value['methodSignature'],
        'methodSignatureText': value['methodSignatureText'],
        'abiItem': value['abiItem'],
    };
}
