// @sentio/api/dist/src/models/WebServiceEventLogsConfig.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonEventLogConfigFromJSON, CommonEventLogConfigToJSON, } from './CommonEventLogConfig.js';
import { WebServiceEventLogsConfigTimeRangeOverrideFromJSON, WebServiceEventLogsConfigTimeRangeOverrideToJSON, } from './WebServiceEventLogsConfigTimeRangeOverride.js';
/**
 * Check if a given object implements the WebServiceEventLogsConfig interface.
 */
export function instanceOfWebServiceEventLogsConfig(value) {
    return true;
}
export function WebServiceEventLogsConfigFromJSON(json) {
    return WebServiceEventLogsConfigFromJSONTyped(json, false);
}
export function WebServiceEventLogsConfigFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'columnsConfig': json['columnsConfig'] == null ? undefined : CommonEventLogConfigFromJSON(json['columnsConfig']),
        'timeRangeOverride': json['timeRangeOverride'] == null ? undefined : WebServiceEventLogsConfigTimeRangeOverrideFromJSON(json['timeRangeOverride']),
        'query': json['query'] == null ? undefined : json['query'],
        'sourceName': json['sourceName'] == null ? undefined : json['sourceName'],
    };
}
export function WebServiceEventLogsConfigToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'columnsConfig': CommonEventLogConfigToJSON(value['columnsConfig']),
        'timeRangeOverride': WebServiceEventLogsConfigTimeRangeOverrideToJSON(value['timeRangeOverride']),
        'query': value['query'],
        'sourceName': value['sourceName'],
    };
}
