// @sentio/api/dist/src/models/WebServiceChartConfigPieConfig.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { WebServiceChartConfigCalculationFromJSON, WebServiceChartConfigCalculationToJSON, } from './WebServiceChartConfigCalculation.js';
import { WebServiceChartConfigPieConfigPieTypeFromJSON, WebServiceChartConfigPieConfigPieTypeToJSON, } from './WebServiceChartConfigPieConfigPieType.js';
/**
 * Check if a given object implements the WebServiceChartConfigPieConfig interface.
 */
export function instanceOfWebServiceChartConfigPieConfig(value) {
    return true;
}
export function WebServiceChartConfigPieConfigFromJSON(json) {
    return WebServiceChartConfigPieConfigFromJSONTyped(json, false);
}
export function WebServiceChartConfigPieConfigFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'pieType': json['pieType'] == null ? undefined : WebServiceChartConfigPieConfigPieTypeFromJSON(json['pieType']),
        'showPercent': json['showPercent'] == null ? undefined : json['showPercent'],
        'showValue': json['showValue'] == null ? undefined : json['showValue'],
        'calculation': json['calculation'] == null ? undefined : WebServiceChartConfigCalculationFromJSON(json['calculation']),
    };
}
export function WebServiceChartConfigPieConfigToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'pieType': WebServiceChartConfigPieConfigPieTypeToJSON(value['pieType']),
        'showPercent': value['showPercent'],
        'showValue': value['showValue'],
        'calculation': WebServiceChartConfigCalculationToJSON(value['calculation']),
    };
}
