// @sentio/api/dist/src/models/WebServiceEventLogsConfigTimeRangeOverride.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonTimeRangeFromJSON, CommonTimeRangeToJSON, } from './CommonTimeRange.js';
/**
 * Check if a given object implements the WebServiceEventLogsConfigTimeRangeOverride interface.
 */
export function instanceOfWebServiceEventLogsConfigTimeRangeOverride(value) {
    return true;
}
export function WebServiceEventLogsConfigTimeRangeOverrideFromJSON(json) {
    return WebServiceEventLogsConfigTimeRangeOverrideFromJSONTyped(json, false);
}
export function WebServiceEventLogsConfigTimeRangeOverrideFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'enabled': json['enabled'] == null ? undefined : json['enabled'],
        'timeRange': json['timeRange'] == null ? undefined : CommonTimeRangeFromJSON(json['timeRange']),
    };
}
export function WebServiceEventLogsConfigTimeRangeOverrideToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'enabled': value['enabled'],
        'timeRange': CommonTimeRangeToJSON(value['timeRange']),
    };
}
