// @sentio/api/dist/src/models/WebServiceDashboardLayoutsLayout.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the WebServiceDashboardLayoutsLayout interface.
 */
export function instanceOfWebServiceDashboardLayoutsLayout(value) {
    return true;
}
export function WebServiceDashboardLayoutsLayoutFromJSON(json) {
    return WebServiceDashboardLayoutsLayoutFromJSONTyped(json, false);
}
export function WebServiceDashboardLayoutsLayoutFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'i': json['i'] == null ? undefined : json['i'],
        'x': json['x'] == null ? undefined : json['x'],
        'y': json['y'] == null ? undefined : json['y'],
        'w': json['w'] == null ? undefined : json['w'],
        'h': json['h'] == null ? undefined : json['h'],
    };
}
export function WebServiceDashboardLayoutsLayoutToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'i': value['i'],
        'x': value['x'],
        'y': value['y'],
        'w': value['w'],
        'h': value['h'],
    };
}
