// @sentio/api/dist/src/models/WebServiceGetProjectListResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonProjectFromJSON, CommonProjectToJSON, } from './CommonProject.js';
/**
 * Check if a given object implements the WebServiceGetProjectListResponse interface.
 */
export function instanceOfWebServiceGetProjectListResponse(value) {
    return true;
}
export function WebServiceGetProjectListResponseFromJSON(json) {
    return WebServiceGetProjectListResponseFromJSONTyped(json, false);
}
export function WebServiceGetProjectListResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'projects': json['projects'] == null ? undefined : (json['projects'].map(CommonProjectFromJSON)),
        'sharedProjects': json['sharedProjects'] == null ? undefined : (json['sharedProjects'].map(CommonProjectFromJSON)),
        'orgProjects': json['orgProjects'] == null ? undefined : (json['orgProjects'].map(CommonProjectFromJSON)),
    };
}
export function WebServiceGetProjectListResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'projects': value['projects'] == null ? undefined : (value['projects'].map(CommonProjectToJSON)),
        'sharedProjects': value['sharedProjects'] == null ? undefined : (value['sharedProjects'].map(CommonProjectToJSON)),
        'orgProjects': value['orgProjects'] == null ? undefined : (value['orgProjects'].map(CommonProjectToJSON)),
    };
}
