// @sentio/api/dist/src/models/WebServiceChartConfigQueryValueConfig.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { WebServiceChartConfigCalculationFromJSON, WebServiceChartConfigCalculationToJSON, } from './WebServiceChartConfigCalculation.js';
import { WebServiceChartConfigColorThemeFromJSON, WebServiceChartConfigColorThemeToJSON, } from './WebServiceChartConfigColorTheme.js';
/**
 * Check if a given object implements the WebServiceChartConfigQueryValueConfig interface.
 */
export function instanceOfWebServiceChartConfigQueryValueConfig(value) {
    return true;
}
export function WebServiceChartConfigQueryValueConfigFromJSON(json) {
    return WebServiceChartConfigQueryValueConfigFromJSONTyped(json, false);
}
export function WebServiceChartConfigQueryValueConfigFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'colorTheme': json['colorTheme'] == null ? undefined : WebServiceChartConfigColorThemeFromJSON(json['colorTheme']),
        'showBackgroundChart': json['showBackgroundChart'] == null ? undefined : json['showBackgroundChart'],
        'calculation': json['calculation'] == null ? undefined : WebServiceChartConfigCalculationFromJSON(json['calculation']),
        'seriesCalculation': json['seriesCalculation'] == null ? undefined : WebServiceChartConfigCalculationFromJSON(json['seriesCalculation']),
    };
}
export function WebServiceChartConfigQueryValueConfigToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'colorTheme': WebServiceChartConfigColorThemeToJSON(value['colorTheme']),
        'showBackgroundChart': value['showBackgroundChart'],
        'calculation': WebServiceChartConfigCalculationToJSON(value['calculation']),
        'seriesCalculation': WebServiceChartConfigCalculationToJSON(value['seriesCalculation']),
    };
}
