// @sentio/api/dist/src/models/WebServicePanel.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { WebServiceChartFromJSON, WebServiceChartToJSON, } from './WebServiceChart.js';
/**
 * Check if a given object implements the WebServicePanel interface.
 */
export function instanceOfWebServicePanel(value) {
    return true;
}
export function WebServicePanelFromJSON(json) {
    return WebServicePanelFromJSONTyped(json, false);
}
export function WebServicePanelFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'id': json['id'] == null ? undefined : json['id'],
        'name': json['name'] == null ? undefined : json['name'],
        'dashboardId': json['dashboardId'] == null ? undefined : json['dashboardId'],
        'chart': json['chart'] == null ? undefined : WebServiceChartFromJSON(json['chart']),
    };
}
export function WebServicePanelToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'id': value['id'],
        'name': value['name'],
        'dashboardId': value['dashboardId'],
        'chart': WebServiceChartToJSON(value['chart']),
    };
}
