// @sentio/api/dist/src/models/WebServiceChartConfigMarker.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { WebServiceChartConfigMarkerTypeFromJSON, WebServiceChartConfigMarkerTypeToJSON, } from './WebServiceChartConfigMarkerType.js';
/**
 * Check if a given object implements the WebServiceChartConfigMarker interface.
 */
export function instanceOfWebServiceChartConfigMarker(value) {
    return true;
}
export function WebServiceChartConfigMarkerFromJSON(json) {
    return WebServiceChartConfigMarkerFromJSONTyped(json, false);
}
export function WebServiceChartConfigMarkerFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'type': json['type'] == null ? undefined : WebServiceChartConfigMarkerTypeFromJSON(json['type']),
        'value': json['value'] == null ? undefined : json['value'],
        'color': json['color'] == null ? undefined : json['color'],
        'label': json['label'] == null ? undefined : json['label'],
        'valueX': json['valueX'] == null ? undefined : json['valueX'],
    };
}
export function WebServiceChartConfigMarkerToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'type': WebServiceChartConfigMarkerTypeToJSON(value['type']),
        'value': value['value'],
        'color': value['color'],
        'label': value['label'],
        'valueX': value['valueX'],
    };
}
