// @sentio/api/dist/src/models/WebServiceImportDashboardRequest.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the WebServiceImportDashboardRequest interface.
 */
export function instanceOfWebServiceImportDashboardRequest(value) {
    if (!('dashboardId' in value))
        return false;
    if (!('dashboardJson' in value))
        return false;
    return true;
}
export function WebServiceImportDashboardRequestFromJSON(json) {
    return WebServiceImportDashboardRequestFromJSONTyped(json, false);
}
export function WebServiceImportDashboardRequestFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'dashboardId': json['dashboardId'],
        'dashboardJson': json['dashboardJson'],
        'overrideLayouts': json['overrideLayouts'] == null ? undefined : json['overrideLayouts'],
    };
}
export function WebServiceImportDashboardRequestToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'dashboardId': value['dashboardId'],
        'dashboardJson': value['dashboardJson'],
        'overrideLayouts': value['overrideLayouts'],
    };
}
