// @sentio/api/dist/src/models/WebServiceGetProjectResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonPermissionFromJSON, CommonPermissionToJSON, } from './CommonPermission.js';
import { CommonProjectFromJSON, CommonProjectToJSON, } from './CommonProject.js';
/**
 * Check if a given object implements the WebServiceGetProjectResponse interface.
 */
export function instanceOfWebServiceGetProjectResponse(value) {
    return true;
}
export function WebServiceGetProjectResponseFromJSON(json) {
    return WebServiceGetProjectResponseFromJSONTyped(json, false);
}
export function WebServiceGetProjectResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'project': json['project'] == null ? undefined : CommonProjectFromJSON(json['project']),
        'permissions': json['permissions'] == null ? undefined : (json['permissions'].map(CommonPermissionFromJSON)),
    };
}
export function WebServiceGetProjectResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'project': CommonProjectToJSON(value['project']),
        'permissions': value['permissions'] == null ? undefined : (value['permissions'].map(CommonPermissionToJSON)),
    };
}
