// @sentio/api/dist/src/models/WebServiceChartConfigBarGaugeConfig.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { WebServiceChartConfigCalculationFromJSON, WebServiceChartConfigCalculationToJSON, } from './WebServiceChartConfigCalculation.js';
import { WebServiceChartConfigDirectionFromJSON, WebServiceChartConfigDirectionToJSON, } from './WebServiceChartConfigDirection.js';
import { WebServiceChartConfigSortFromJSON, WebServiceChartConfigSortToJSON, } from './WebServiceChartConfigSort.js';
/**
 * Check if a given object implements the WebServiceChartConfigBarGaugeConfig interface.
 */
export function instanceOfWebServiceChartConfigBarGaugeConfig(value) {
    return true;
}
export function WebServiceChartConfigBarGaugeConfigFromJSON(json) {
    return WebServiceChartConfigBarGaugeConfigFromJSONTyped(json, false);
}
export function WebServiceChartConfigBarGaugeConfigFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'direction': json['direction'] == null ? undefined : WebServiceChartConfigDirectionFromJSON(json['direction']),
        'calculation': json['calculation'] == null ? undefined : WebServiceChartConfigCalculationFromJSON(json['calculation']),
        'sort': json['sort'] == null ? undefined : WebServiceChartConfigSortFromJSON(json['sort']),
    };
}
export function WebServiceChartConfigBarGaugeConfigToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'direction': WebServiceChartConfigDirectionToJSON(value['direction']),
        'calculation': WebServiceChartConfigCalculationToJSON(value['calculation']),
        'sort': WebServiceChartConfigSortToJSON(value['sort']),
    };
}
