// @sentio/api/dist/src/models/WebServiceChartConfigCalculation.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 */
export const WebServiceChartConfigCalculation = {
    Last: 'LAST',
    First: 'FIRST',
    Mean: 'MEAN',
    Total: 'TOTAL',
    All: 'ALL',
    Min: 'MIN',
    Max: 'MAX'
};
export function instanceOfWebServiceChartConfigCalculation(value) {
    return Object.values(WebServiceChartConfigCalculation).includes(value);
}
export function WebServiceChartConfigCalculationFromJSON(json) {
    return WebServiceChartConfigCalculationFromJSONTyped(json, false);
}
export function WebServiceChartConfigCalculationFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
export function WebServiceChartConfigCalculationToJSON(value) {
    return value;
}
