// @sentio/api/dist/src/models/WebServiceChartConfigCompareTime.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonDurationFromJSON, CommonDurationToJSON, } from './CommonDuration.js';
/**
 * Check if a given object implements the WebServiceChartConfigCompareTime interface.
 */
export function instanceOfWebServiceChartConfigCompareTime(value) {
    return true;
}
export function WebServiceChartConfigCompareTimeFromJSON(json) {
    return WebServiceChartConfigCompareTimeFromJSONTyped(json, false);
}
export function WebServiceChartConfigCompareTimeFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'ago': json['ago'] == null ? undefined : CommonDurationFromJSON(json['ago']),
    };
}
export function WebServiceChartConfigCompareTimeToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'ago': CommonDurationToJSON(value['ago']),
    };
}
