// @sentio/api/dist/src/models/WebServiceChart.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonFormulaFromJSON, CommonFormulaToJSON, } from './CommonFormula.js';
import { CommonQueryFromJSON, CommonQueryToJSON, } from './CommonQuery.js';
import { CommonRetentionQueryFromJSON, CommonRetentionQueryToJSON, } from './CommonRetentionQuery.js';
import { CommonSegmentationQueryFromJSON, CommonSegmentationQueryToJSON, } from './CommonSegmentationQuery.js';
import { InsightsServiceQueryRequestQueryFromJSON, InsightsServiceQueryRequestQueryToJSON, } from './InsightsServiceQueryRequestQuery.js';
import { WebServiceChartChartTypeFromJSON, WebServiceChartChartTypeToJSON, } from './WebServiceChartChartType.js';
import { WebServiceChartConfigFromJSON, WebServiceChartConfigToJSON, } from './WebServiceChartConfig.js';
import { WebServiceChartDataSourceTypeFromJSON, WebServiceChartDataSourceTypeToJSON, } from './WebServiceChartDataSourceType.js';
import { WebServiceEventLogsConfigFromJSON, WebServiceEventLogsConfigToJSON, } from './WebServiceEventLogsConfig.js';
import { WebServiceNoteFromJSON, WebServiceNoteToJSON, } from './WebServiceNote.js';
/**
 * Check if a given object implements the WebServiceChart interface.
 */
export function instanceOfWebServiceChart(value) {
    return true;
}
export function WebServiceChartFromJSON(json) {
    return WebServiceChartFromJSONTyped(json, false);
}
export function WebServiceChartFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'type': json['type'] == null ? undefined : WebServiceChartChartTypeFromJSON(json['type']),
        'queries': json['queries'] == null ? undefined : (json['queries'].map(CommonQueryFromJSON)),
        'formulas': json['formulas'] == null ? undefined : (json['formulas'].map(CommonFormulaFromJSON)),
        'config': json['config'] == null ? undefined : WebServiceChartConfigFromJSON(json['config']),
        'note': json['note'] == null ? undefined : WebServiceNoteFromJSON(json['note']),
        'datasourceType': json['datasourceType'] == null ? undefined : WebServiceChartDataSourceTypeFromJSON(json['datasourceType']),
        'segmentationQueries': json['segmentationQueries'] == null ? undefined : (json['segmentationQueries'].map(CommonSegmentationQueryFromJSON)),
        'insightsQueries': json['insightsQueries'] == null ? undefined : (json['insightsQueries'].map(InsightsServiceQueryRequestQueryFromJSON)),
        'eventLogsConfig': json['eventLogsConfig'] == null ? undefined : WebServiceEventLogsConfigFromJSON(json['eventLogsConfig']),
        'retentionQuery': json['retentionQuery'] == null ? undefined : CommonRetentionQueryFromJSON(json['retentionQuery']),
        'sqlQuery': json['sqlQuery'] == null ? undefined : json['sqlQuery'],
    };
}
export function WebServiceChartToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'type': WebServiceChartChartTypeToJSON(value['type']),
        'queries': value['queries'] == null ? undefined : (value['queries'].map(CommonQueryToJSON)),
        'formulas': value['formulas'] == null ? undefined : (value['formulas'].map(CommonFormulaToJSON)),
        'config': WebServiceChartConfigToJSON(value['config']),
        'note': WebServiceNoteToJSON(value['note']),
        'datasourceType': WebServiceChartDataSourceTypeToJSON(value['datasourceType']),
        'segmentationQueries': value['segmentationQueries'] == null ? undefined : (value['segmentationQueries'].map(CommonSegmentationQueryToJSON)),
        'insightsQueries': value['insightsQueries'] == null ? undefined : (value['insightsQueries'].map(InsightsServiceQueryRequestQueryToJSON)),
        'eventLogsConfig': WebServiceEventLogsConfigToJSON(value['eventLogsConfig']),
        'retentionQuery': CommonRetentionQueryToJSON(value['retentionQuery']),
        'sqlQuery': value['sqlQuery'],
    };
}
