// @sentio/api/dist/src/models/WebServiceDashboardExtraTemplateView.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the WebServiceDashboardExtraTemplateView interface.
 */
export function instanceOfWebServiceDashboardExtraTemplateView(value) {
    return true;
}
export function WebServiceDashboardExtraTemplateViewFromJSON(json) {
    return WebServiceDashboardExtraTemplateViewFromJSONTyped(json, false);
}
export function WebServiceDashboardExtraTemplateViewFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'values': json['values'] == null ? undefined : json['values'],
    };
}
export function WebServiceDashboardExtraTemplateViewToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'values': value['values'],
    };
}
