// @sentio/api/dist/src/models/WebServiceChartConfigLineConfig.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { WebServiceChartConfigLineConfigStyleFromJSON, WebServiceChartConfigLineConfigStyleToJSON, } from './WebServiceChartConfigLineConfigStyle.js';
/**
 * Check if a given object implements the WebServiceChartConfigLineConfig interface.
 */
export function instanceOfWebServiceChartConfigLineConfig(value) {
    return true;
}
export function WebServiceChartConfigLineConfigFromJSON(json) {
    return WebServiceChartConfigLineConfigFromJSONTyped(json, false);
}
export function WebServiceChartConfigLineConfigFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'style': json['style'] == null ? undefined : WebServiceChartConfigLineConfigStyleFromJSON(json['style']),
    };
}
export function WebServiceChartConfigLineConfigToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'style': WebServiceChartConfigLineConfigStyleToJSON(value['style']),
    };
}
