// @sentio/api/dist/src/models/WebServiceDashboardLayouts.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { WebServiceDashboardLayoutsLayoutFromJSON, WebServiceDashboardLayoutsLayoutToJSON, } from './WebServiceDashboardLayoutsLayout.js';
/**
 * Check if a given object implements the WebServiceDashboardLayouts interface.
 */
export function instanceOfWebServiceDashboardLayouts(value) {
    return true;
}
export function WebServiceDashboardLayoutsFromJSON(json) {
    return WebServiceDashboardLayoutsFromJSONTyped(json, false);
}
export function WebServiceDashboardLayoutsFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'layouts': json['layouts'] == null ? undefined : (json['layouts'].map(WebServiceDashboardLayoutsLayoutFromJSON)),
    };
}
export function WebServiceDashboardLayoutsToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'layouts': value['layouts'] == null ? undefined : (value['layouts'].map(WebServiceDashboardLayoutsLayoutToJSON)),
    };
}
