// @sentio/api/dist/src/models/WebServiceChartDataSourceType.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 */
export const WebServiceChartDataSourceType = {
    Metrics: 'METRICS',
    Notes: 'NOTES',
    Analytics: 'ANALYTICS',
    Insights: 'INSIGHTS',
    Events: 'EVENTS',
    Retention: 'RETENTION',
    Sql: 'SQL'
};
export function instanceOfWebServiceChartDataSourceType(value) {
    return Object.values(WebServiceChartDataSourceType).includes(value);
}
export function WebServiceChartDataSourceTypeFromJSON(json) {
    return WebServiceChartDataSourceTypeFromJSONTyped(json, false);
}
export function WebServiceChartDataSourceTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
export function WebServiceChartDataSourceTypeToJSON(value) {
    return value;
}
