// @sentio/api/dist/src/models/WebServiceDashboard.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { mapValues } from '../runtime.js';
import { WebServiceDashboardDashboardVisibilityFromJSON, WebServiceDashboardDashboardVisibilityToJSON, } from './WebServiceDashboardDashboardVisibility.js';
import { WebServiceDashboardExtraFromJSON, WebServiceDashboardExtraToJSON, } from './WebServiceDashboardExtra.js';
import { WebServiceDashboardResponsiveLayoutsFromJSON, WebServiceDashboardResponsiveLayoutsToJSON, } from './WebServiceDashboardResponsiveLayouts.js';
import { WebServiceDashboardSharingFromJSON, WebServiceDashboardSharingToJSON, } from './WebServiceDashboardSharing.js';
import { WebServicePanelFromJSON, WebServicePanelToJSON, } from './WebServicePanel.js';
/**
 * Check if a given object implements the WebServiceDashboard interface.
 */
export function instanceOfWebServiceDashboard(value) {
    return true;
}
export function WebServiceDashboardFromJSON(json) {
    return WebServiceDashboardFromJSONTyped(json, false);
}
export function WebServiceDashboardFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'id': json['id'] == null ? undefined : json['id'],
        'name': json['name'] == null ? undefined : json['name'],
        'projectId': json['projectId'] == null ? undefined : json['projectId'],
        'description': json['description'] == null ? undefined : json['description'],
        'createdAt': json['createdAt'] == null ? undefined : (new Date(json['createdAt'])),
        'updatedAt': json['updatedAt'] == null ? undefined : (new Date(json['updatedAt'])),
        'panels': json['panels'] == null ? undefined : (mapValues(json['panels'], WebServicePanelFromJSON)),
        'layouts': json['layouts'] == null ? undefined : WebServiceDashboardResponsiveLayoutsFromJSON(json['layouts']),
        'extra': json['extra'] == null ? undefined : WebServiceDashboardExtraFromJSON(json['extra']),
        'sharing': json['sharing'] == null ? undefined : WebServiceDashboardSharingFromJSON(json['sharing']),
        '_default': json['default'] == null ? undefined : json['default'],
        'visibility': json['visibility'] == null ? undefined : WebServiceDashboardDashboardVisibilityFromJSON(json['visibility']),
        'ownerId': json['ownerId'] == null ? undefined : json['ownerId'],
        'tags': json['tags'] == null ? undefined : json['tags'],
        'url': json['url'] == null ? undefined : json['url'],
    };
}
export function WebServiceDashboardToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'id': value['id'],
        'name': value['name'],
        'projectId': value['projectId'],
        'description': value['description'],
        'createdAt': value['createdAt'] == null ? undefined : ((value['createdAt']).toISOString()),
        'updatedAt': value['updatedAt'] == null ? undefined : ((value['updatedAt']).toISOString()),
        'panels': value['panels'] == null ? undefined : (mapValues(value['panels'], WebServicePanelToJSON)),
        'layouts': WebServiceDashboardResponsiveLayoutsToJSON(value['layouts']),
        'extra': WebServiceDashboardExtraToJSON(value['extra']),
        'sharing': WebServiceDashboardSharingToJSON(value['sharing']),
        'default': value['_default'],
        'visibility': WebServiceDashboardDashboardVisibilityToJSON(value['visibility']),
        'ownerId': value['ownerId'],
        'tags': value['tags'],
        'url': value['url'],
    };
}
