// @sentio/api/dist/src/models/WebServiceDashboardExtra.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { mapValues } from '../runtime.js';
import { WebServiceDashboardExtraTemplateVariableFromJSON, WebServiceDashboardExtraTemplateVariableToJSON, } from './WebServiceDashboardExtraTemplateVariable.js';
import { WebServiceDashboardExtraTemplateViewFromJSON, WebServiceDashboardExtraTemplateViewToJSON, } from './WebServiceDashboardExtraTemplateView.js';
/**
 * Check if a given object implements the WebServiceDashboardExtra interface.
 */
export function instanceOfWebServiceDashboardExtra(value) {
    return true;
}
export function WebServiceDashboardExtraFromJSON(json) {
    return WebServiceDashboardExtraFromJSONTyped(json, false);
}
export function WebServiceDashboardExtraFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'templateVariables': json['templateVariables'] == null ? undefined : (mapValues(json['templateVariables'], WebServiceDashboardExtraTemplateVariableFromJSON)),
        'templateViews': json['templateViews'] == null ? undefined : (json['templateViews'].map(WebServiceDashboardExtraTemplateViewFromJSON)),
    };
}
export function WebServiceDashboardExtraToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'templateVariables': value['templateVariables'] == null ? undefined : (mapValues(value['templateVariables'], WebServiceDashboardExtraTemplateVariableToJSON)),
        'templateViews': value['templateViews'] == null ? undefined : (value['templateViews'].map(WebServiceDashboardExtraTemplateViewToJSON)),
    };
}
