// @sentio/api/dist/src/models/WebServiceChartConfigTableConfig.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { mapValues } from '../runtime.js';
import { WebServiceChartConfigCalculationFromJSON, WebServiceChartConfigCalculationToJSON, } from './WebServiceChartConfigCalculation.js';
import { WebServiceChartConfigColumnSortFromJSON, WebServiceChartConfigColumnSortToJSON, } from './WebServiceChartConfigColumnSort.js';
import { WebServiceChartConfigValueConfigFromJSON, WebServiceChartConfigValueConfigToJSON, } from './WebServiceChartConfigValueConfig.js';
/**
 * Check if a given object implements the WebServiceChartConfigTableConfig interface.
 */
export function instanceOfWebServiceChartConfigTableConfig(value) {
    return true;
}
export function WebServiceChartConfigTableConfigFromJSON(json) {
    return WebServiceChartConfigTableConfigFromJSONTyped(json, false);
}
export function WebServiceChartConfigTableConfigFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'calculation': json['calculation'] == null ? undefined : WebServiceChartConfigCalculationFromJSON(json['calculation']),
        'showColumns': json['showColumns'] == null ? undefined : json['showColumns'],
        'sortColumns': json['sortColumns'] == null ? undefined : (json['sortColumns'].map(WebServiceChartConfigColumnSortFromJSON)),
        'columnOrders': json['columnOrders'] == null ? undefined : json['columnOrders'],
        'columnWidths': json['columnWidths'] == null ? undefined : json['columnWidths'],
        'showPlainData': json['showPlainData'] == null ? undefined : json['showPlainData'],
        'calculations': json['calculations'] == null ? undefined : (mapValues(json['calculations'], WebServiceChartConfigCalculationFromJSON)),
        'valueConfigs': json['valueConfigs'] == null ? undefined : (mapValues(json['valueConfigs'], WebServiceChartConfigValueConfigFromJSON)),
        'rowLimit': json['rowLimit'] == null ? undefined : json['rowLimit'],
    };
}
export function WebServiceChartConfigTableConfigToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'calculation': WebServiceChartConfigCalculationToJSON(value['calculation']),
        'showColumns': value['showColumns'],
        'sortColumns': value['sortColumns'] == null ? undefined : (value['sortColumns'].map(WebServiceChartConfigColumnSortToJSON)),
        'columnOrders': value['columnOrders'],
        'columnWidths': value['columnWidths'],
        'showPlainData': value['showPlainData'],
        'calculations': value['calculations'] == null ? undefined : (mapValues(value['calculations'], WebServiceChartConfigCalculationToJSON)),
        'valueConfigs': value['valueConfigs'] == null ? undefined : (mapValues(value['valueConfigs'], WebServiceChartConfigValueConfigToJSON)),
        'rowLimit': value['rowLimit'],
    };
}
