// @sentio/api/dist/src/models/WebServiceChartConfigMappingRule.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { WebServiceChartConfigColorThemeFromJSON, WebServiceChartConfigColorThemeToJSON, } from './WebServiceChartConfigColorTheme.js';
/**
 * Check if a given object implements the WebServiceChartConfigMappingRule interface.
 */
export function instanceOfWebServiceChartConfigMappingRule(value) {
    return true;
}
export function WebServiceChartConfigMappingRuleFromJSON(json) {
    return WebServiceChartConfigMappingRuleFromJSONTyped(json, false);
}
export function WebServiceChartConfigMappingRuleFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'comparison': json['comparison'] == null ? undefined : json['comparison'],
        'value': json['value'] == null ? undefined : json['value'],
        'text': json['text'] == null ? undefined : json['text'],
        'colorTheme': json['colorTheme'] == null ? undefined : WebServiceChartConfigColorThemeFromJSON(json['colorTheme']),
    };
}
export function WebServiceChartConfigMappingRuleToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'comparison': value['comparison'],
        'value': value['value'],
        'text': value['text'],
        'colorTheme': WebServiceChartConfigColorThemeToJSON(value['colorTheme']),
    };
}
