// @sentio/api/dist/src/models/WebServiceNote.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { WebServiceNoteAlignmentFromJSON, WebServiceNoteAlignmentToJSON, } from './WebServiceNoteAlignment.js';
import { WebServiceNoteFontSizeFromJSON, WebServiceNoteFontSizeToJSON, } from './WebServiceNoteFontSize.js';
import { WebServiceNoteVerticalAlignmentFromJSON, WebServiceNoteVerticalAlignmentToJSON, } from './WebServiceNoteVerticalAlignment.js';
/**
 * Check if a given object implements the WebServiceNote interface.
 */
export function instanceOfWebServiceNote(value) {
    return true;
}
export function WebServiceNoteFromJSON(json) {
    return WebServiceNoteFromJSONTyped(json, false);
}
export function WebServiceNoteFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'content': json['content'] == null ? undefined : json['content'],
        'fontSize': json['fontSize'] == null ? undefined : WebServiceNoteFontSizeFromJSON(json['fontSize']),
        'textAlign': json['textAlign'] == null ? undefined : WebServiceNoteAlignmentFromJSON(json['textAlign']),
        'verticalAlign': json['verticalAlign'] == null ? undefined : WebServiceNoteVerticalAlignmentFromJSON(json['verticalAlign']),
        'backgroundColor': json['backgroundColor'] == null ? undefined : json['backgroundColor'],
        'textColor': json['textColor'] == null ? undefined : json['textColor'],
    };
}
export function WebServiceNoteToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'content': value['content'],
        'fontSize': WebServiceNoteFontSizeToJSON(value['fontSize']),
        'textAlign': WebServiceNoteAlignmentToJSON(value['textAlign']),
        'verticalAlign': WebServiceNoteVerticalAlignmentToJSON(value['verticalAlign']),
        'backgroundColor': value['backgroundColor'],
        'textColor': value['textColor'],
    };
}
