// @sentio/api/dist/src/models/WebServiceChartConfigValueConfig.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { WebServiceChartConfigMappingRuleFromJSON, WebServiceChartConfigMappingRuleToJSON, } from './WebServiceChartConfigMappingRule.js';
import { WebServiceChartConfigValueConfigStyleFromJSON, WebServiceChartConfigValueConfigStyleToJSON, } from './WebServiceChartConfigValueConfigStyle.js';
import { WebServiceChartConfigValueFormatterFromJSON, WebServiceChartConfigValueFormatterToJSON, } from './WebServiceChartConfigValueFormatter.js';
/**
 * Check if a given object implements the WebServiceChartConfigValueConfig interface.
 */
export function instanceOfWebServiceChartConfigValueConfig(value) {
    return true;
}
export function WebServiceChartConfigValueConfigFromJSON(json) {
    return WebServiceChartConfigValueConfigFromJSONTyped(json, false);
}
export function WebServiceChartConfigValueConfigFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'valueFormatter': json['valueFormatter'] == null ? undefined : WebServiceChartConfigValueFormatterFromJSON(json['valueFormatter']),
        'showValueLabel': json['showValueLabel'] == null ? undefined : json['showValueLabel'],
        'maxSignificantDigits': json['maxSignificantDigits'] == null ? undefined : json['maxSignificantDigits'],
        'dateFormat': json['dateFormat'] == null ? undefined : json['dateFormat'],
        'mappingRules': json['mappingRules'] == null ? undefined : (json['mappingRules'].map(WebServiceChartConfigMappingRuleFromJSON)),
        'style': json['style'] == null ? undefined : WebServiceChartConfigValueConfigStyleFromJSON(json['style']),
        'maxFractionDigits': json['maxFractionDigits'] == null ? undefined : json['maxFractionDigits'],
        'precision': json['precision'] == null ? undefined : json['precision'],
        'currencySymbol': json['currencySymbol'] == null ? undefined : json['currencySymbol'],
    };
}
export function WebServiceChartConfigValueConfigToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'valueFormatter': WebServiceChartConfigValueFormatterToJSON(value['valueFormatter']),
        'showValueLabel': value['showValueLabel'],
        'maxSignificantDigits': value['maxSignificantDigits'],
        'dateFormat': value['dateFormat'],
        'mappingRules': value['mappingRules'] == null ? undefined : (value['mappingRules'].map(WebServiceChartConfigMappingRuleToJSON)),
        'style': WebServiceChartConfigValueConfigStyleToJSON(value['style']),
        'maxFractionDigits': value['maxFractionDigits'],
        'precision': value['precision'],
        'currencySymbol': value['currencySymbol'],
    };
}
