// @sentio/api/dist/src/models/WebServiceChartConfigTimeRangeOverride.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonTimeRangeFromJSON, CommonTimeRangeToJSON, } from './CommonTimeRange.js';
import { WebServiceChartConfigCompareTimeFromJSON, WebServiceChartConfigCompareTimeToJSON, } from './WebServiceChartConfigCompareTime.js';
/**
 * Check if a given object implements the WebServiceChartConfigTimeRangeOverride interface.
 */
export function instanceOfWebServiceChartConfigTimeRangeOverride(value) {
    return true;
}
export function WebServiceChartConfigTimeRangeOverrideFromJSON(json) {
    return WebServiceChartConfigTimeRangeOverrideFromJSONTyped(json, false);
}
export function WebServiceChartConfigTimeRangeOverrideFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'enabled': json['enabled'] == null ? undefined : json['enabled'],
        'timeRange': json['timeRange'] == null ? undefined : CommonTimeRangeFromJSON(json['timeRange']),
        'compareTime': json['compareTime'] == null ? undefined : WebServiceChartConfigCompareTimeFromJSON(json['compareTime']),
    };
}
export function WebServiceChartConfigTimeRangeOverrideToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'enabled': value['enabled'],
        'timeRange': CommonTimeRangeToJSON(value['timeRange']),
        'compareTime': WebServiceChartConfigCompareTimeToJSON(value['compareTime']),
    };
}
