// @sentio/api/dist/src/models/WebServiceImportDashboardResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { WebServiceDashboardFromJSON, WebServiceDashboardToJSON, } from './WebServiceDashboard.js';
/**
 * Check if a given object implements the WebServiceImportDashboardResponse interface.
 */
export function instanceOfWebServiceImportDashboardResponse(value) {
    return true;
}
export function WebServiceImportDashboardResponseFromJSON(json) {
    return WebServiceImportDashboardResponseFromJSONTyped(json, false);
}
export function WebServiceImportDashboardResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'dashboard': json['dashboard'] == null ? undefined : WebServiceDashboardFromJSON(json['dashboard']),
    };
}
export function WebServiceImportDashboardResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'dashboard': WebServiceDashboardToJSON(value['dashboard']),
    };
}
