// @sentio/api/dist/src/models/WebServiceChartConfigColorTheme.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the WebServiceChartConfigColorTheme interface.
 */
export function instanceOfWebServiceChartConfigColorTheme(value) {
    return true;
}
export function WebServiceChartConfigColorThemeFromJSON(json) {
    return WebServiceChartConfigColorThemeFromJSONTyped(json, false);
}
export function WebServiceChartConfigColorThemeFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'textColor': json['textColor'] == null ? undefined : json['textColor'],
        'backgroundColor': json['backgroundColor'] == null ? undefined : json['backgroundColor'],
        'themeType': json['themeType'] == null ? undefined : json['themeType'],
    };
}
export function WebServiceChartConfigColorThemeToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'textColor': value['textColor'],
        'backgroundColor': value['backgroundColor'],
        'themeType': value['themeType'],
    };
}
