// @sentio/api/dist/src/models/WebServiceChartConfigSort.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { WebServiceChartConfigSortByFromJSON, WebServiceChartConfigSortByToJSON, } from './WebServiceChartConfigSortBy.js';
/**
 * Check if a given object implements the WebServiceChartConfigSort interface.
 */
export function instanceOfWebServiceChartConfigSort(value) {
    return true;
}
export function WebServiceChartConfigSortFromJSON(json) {
    return WebServiceChartConfigSortFromJSONTyped(json, false);
}
export function WebServiceChartConfigSortFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'sortBy': json['sortBy'] == null ? undefined : WebServiceChartConfigSortByFromJSON(json['sortBy']),
        'orderDesc': json['orderDesc'] == null ? undefined : json['orderDesc'],
    };
}
export function WebServiceChartConfigSortToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'sortBy': WebServiceChartConfigSortByToJSON(value['sortBy']),
        'orderDesc': value['orderDesc'],
    };
}
