// @sentio/api/dist/src/models/WebServiceChartConfigValueConfigStyle.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 */
export const WebServiceChartConfigValueConfigStyle = {
    Standard: 'Standard',
    Compact: 'Compact',
    Scientific: 'Scientific',
    Percent: 'Percent',
    Currency: 'Currency',
    None: 'None'
};
export function instanceOfWebServiceChartConfigValueConfigStyle(value) {
    return Object.values(WebServiceChartConfigValueConfigStyle).includes(value);
}
export function WebServiceChartConfigValueConfigStyleFromJSON(json) {
    return WebServiceChartConfigValueConfigStyleFromJSONTyped(json, false);
}
export function WebServiceChartConfigValueConfigStyleFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
export function WebServiceChartConfigValueConfigStyleToJSON(value) {
    return value;
}
