// @sentio/api/dist/src/models/WebServiceDashboardExtraTemplateVariable.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the WebServiceDashboardExtraTemplateVariable interface.
 */
export function instanceOfWebServiceDashboardExtraTemplateVariable(value) {
    return true;
}
export function WebServiceDashboardExtraTemplateVariableFromJSON(json) {
    return WebServiceDashboardExtraTemplateVariableFromJSONTyped(json, false);
}
export function WebServiceDashboardExtraTemplateVariableFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'field': json['field'] == null ? undefined : json['field'],
        'defaultValue': json['defaultValue'] == null ? undefined : json['defaultValue'],
        'sourceName': json['sourceName'] == null ? undefined : json['sourceName'],
        'options': json['options'] == null ? undefined : json['options'],
    };
}
export function WebServiceDashboardExtraTemplateVariableToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'field': value['field'],
        'defaultValue': value['defaultValue'],
        'sourceName': value['sourceName'],
        'options': value['options'],
    };
}
