// @sentio/api/dist/src/models/WebServiceDashboardSharing.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the WebServiceDashboardSharing interface.
 */
export function instanceOfWebServiceDashboardSharing(value) {
    return true;
}
export function WebServiceDashboardSharingFromJSON(json) {
    return WebServiceDashboardSharingFromJSONTyped(json, false);
}
export function WebServiceDashboardSharingFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'id': json['id'] == null ? undefined : json['id'],
        'dashboardId': json['dashboardId'] == null ? undefined : json['dashboardId'],
        'isPublic': json['isPublic'] == null ? undefined : json['isPublic'],
        'viewers': json['viewers'] == null ? undefined : json['viewers'],
    };
}
export function WebServiceDashboardSharingToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'id': value['id'],
        'dashboardId': value['dashboardId'],
        'isPublic': value['isPublic'],
        'viewers': value['viewers'],
    };
}
