// @sentio/api/dist/src/models/WebServiceExportDashboardResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the WebServiceExportDashboardResponse interface.
 */
export function instanceOfWebServiceExportDashboardResponse(value) {
    return true;
}
export function WebServiceExportDashboardResponseFromJSON(json) {
    return WebServiceExportDashboardResponseFromJSONTyped(json, false);
}
export function WebServiceExportDashboardResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'dashboardJson': json['dashboardJson'] == null ? undefined : json['dashboardJson'],
    };
}
export function WebServiceExportDashboardResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'dashboardJson': value['dashboardJson'],
    };
}
