// @sentio/api/dist/src/models/WebServiceDashboardResponsiveLayouts.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { mapValues } from '../runtime.js';
import { WebServiceDashboardLayoutsFromJSON, WebServiceDashboardLayoutsToJSON, } from './WebServiceDashboardLayouts.js';
/**
 * Check if a given object implements the WebServiceDashboardResponsiveLayouts interface.
 */
export function instanceOfWebServiceDashboardResponsiveLayouts(value) {
    return true;
}
export function WebServiceDashboardResponsiveLayoutsFromJSON(json) {
    return WebServiceDashboardResponsiveLayoutsFromJSONTyped(json, false);
}
export function WebServiceDashboardResponsiveLayoutsFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'responsiveLayouts': json['responsiveLayouts'] == null ? undefined : (mapValues(json['responsiveLayouts'], WebServiceDashboardLayoutsFromJSON)),
    };
}
export function WebServiceDashboardResponsiveLayoutsToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'responsiveLayouts': value['responsiveLayouts'] == null ? undefined : (mapValues(value['responsiveLayouts'], WebServiceDashboardLayoutsToJSON)),
    };
}
