// @sentio/api/dist/src/models/WebServiceChartConfigValueFormatter.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 */
export const WebServiceChartConfigValueFormatter = {
    NumberFormatter: 'NumberFormatter',
    DateFormatter: 'DateFormatter',
    StringFormatter: 'StringFormatter'
};
export function instanceOfWebServiceChartConfigValueFormatter(value) {
    return Object.values(WebServiceChartConfigValueFormatter).includes(value);
}
export function WebServiceChartConfigValueFormatterFromJSON(json) {
    return WebServiceChartConfigValueFormatterFromJSONTyped(json, false);
}
export function WebServiceChartConfigValueFormatterFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
export function WebServiceChartConfigValueFormatterToJSON(value) {
    return value;
}
