// @sentio/api/dist/src/models/WebServiceChartConfigYAxisConfig.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Check if a given object implements the WebServiceChartConfigYAxisConfig interface.
 */
export function instanceOfWebServiceChartConfigYAxisConfig(value) {
    return true;
}
export function WebServiceChartConfigYAxisConfigFromJSON(json) {
    return WebServiceChartConfigYAxisConfigFromJSONTyped(json, false);
}
export function WebServiceChartConfigYAxisConfigFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'min': json['min'] == null ? undefined : json['min'],
        'max': json['max'] == null ? undefined : json['max'],
        'scale': json['scale'] == null ? undefined : json['scale'],
        'stacked': json['stacked'] == null ? undefined : json['stacked'],
    };
}
export function WebServiceChartConfigYAxisConfigToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'min': value['min'],
        'max': value['max'],
        'scale': value['scale'],
        'stacked': value['stacked'],
    };
}
