// @sentio/api/dist/src/models/WebServiceGetDashboardResponse.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { CommonPermissionFromJSON, CommonPermissionToJSON, } from './CommonPermission.js';
import { WebServiceDashboardFromJSON, WebServiceDashboardToJSON, } from './WebServiceDashboard.js';
/**
 * Check if a given object implements the WebServiceGetDashboardResponse interface.
 */
export function instanceOfWebServiceGetDashboardResponse(value) {
    return true;
}
export function WebServiceGetDashboardResponseFromJSON(json) {
    return WebServiceGetDashboardResponseFromJSONTyped(json, false);
}
export function WebServiceGetDashboardResponseFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'dashboards': json['dashboards'] == null ? undefined : (json['dashboards'].map(WebServiceDashboardFromJSON)),
        'permissions': json['permissions'] == null ? undefined : (json['permissions'].map(CommonPermissionFromJSON)),
    };
}
export function WebServiceGetDashboardResponseToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'dashboards': value['dashboards'] == null ? undefined : (value['dashboards'].map(WebServiceDashboardToJSON)),
        'permissions': value['permissions'] == null ? undefined : (value['permissions'].map(CommonPermissionToJSON)),
    };
}
