// @sentio/api/dist/src/models/WebServiceChartConfig.js
/* tslint:disable */
/* eslint-disable */
/**
 * Sentio API
 * Sentio Open API for query data
 *
 * The version of the OpenAPI document: version not set
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { WebServiceChartConfigBarGaugeConfigFromJSON, WebServiceChartConfigBarGaugeConfigToJSON, } from './WebServiceChartConfigBarGaugeConfig.js';
import { WebServiceChartConfigLineConfigFromJSON, WebServiceChartConfigLineConfigToJSON, } from './WebServiceChartConfigLineConfig.js';
import { WebServiceChartConfigMarkerFromJSON, WebServiceChartConfigMarkerToJSON, } from './WebServiceChartConfigMarker.js';
import { WebServiceChartConfigPieConfigFromJSON, WebServiceChartConfigPieConfigToJSON, } from './WebServiceChartConfigPieConfig.js';
import { WebServiceChartConfigQueryValueConfigFromJSON, WebServiceChartConfigQueryValueConfigToJSON, } from './WebServiceChartConfigQueryValueConfig.js';
import { WebServiceChartConfigTableConfigFromJSON, WebServiceChartConfigTableConfigToJSON, } from './WebServiceChartConfigTableConfig.js';
import { WebServiceChartConfigTimeRangeOverrideFromJSON, WebServiceChartConfigTimeRangeOverrideToJSON, } from './WebServiceChartConfigTimeRangeOverride.js';
import { WebServiceChartConfigValueConfigFromJSON, WebServiceChartConfigValueConfigToJSON, } from './WebServiceChartConfigValueConfig.js';
import { WebServiceChartConfigYAxisConfigFromJSON, WebServiceChartConfigYAxisConfigToJSON, } from './WebServiceChartConfigYAxisConfig.js';
/**
 * Check if a given object implements the WebServiceChartConfig interface.
 */
export function instanceOfWebServiceChartConfig(value) {
    return true;
}
export function WebServiceChartConfigFromJSON(json) {
    return WebServiceChartConfigFromJSONTyped(json, false);
}
export function WebServiceChartConfigFromJSONTyped(json, ignoreDiscriminator) {
    if (json == null) {
        return json;
    }
    return {
        'yAxis': json['yAxis'] == null ? undefined : WebServiceChartConfigYAxisConfigFromJSON(json['yAxis']),
        'barGauge': json['barGauge'] == null ? undefined : WebServiceChartConfigBarGaugeConfigFromJSON(json['barGauge']),
        'valueConfig': json['valueConfig'] == null ? undefined : WebServiceChartConfigValueConfigFromJSON(json['valueConfig']),
        'timeRangeOverride': json['timeRangeOverride'] == null ? undefined : WebServiceChartConfigTimeRangeOverrideFromJSON(json['timeRangeOverride']),
        'tableConfig': json['tableConfig'] == null ? undefined : WebServiceChartConfigTableConfigFromJSON(json['tableConfig']),
        'queryValueConfig': json['queryValueConfig'] == null ? undefined : WebServiceChartConfigQueryValueConfigFromJSON(json['queryValueConfig']),
        'pieConfig': json['pieConfig'] == null ? undefined : WebServiceChartConfigPieConfigFromJSON(json['pieConfig']),
        'markers': json['markers'] == null ? undefined : (json['markers'].map(WebServiceChartConfigMarkerFromJSON)),
        'lineConfig': json['lineConfig'] == null ? undefined : WebServiceChartConfigLineConfigFromJSON(json['lineConfig']),
    };
}
export function WebServiceChartConfigToJSON(value) {
    if (value == null) {
        return value;
    }
    return {
        'yAxis': WebServiceChartConfigYAxisConfigToJSON(value['yAxis']),
        'barGauge': WebServiceChartConfigBarGaugeConfigToJSON(value['barGauge']),
        'valueConfig': WebServiceChartConfigValueConfigToJSON(value['valueConfig']),
        'timeRangeOverride': WebServiceChartConfigTimeRangeOverrideToJSON(value['timeRangeOverride']),
        'tableConfig': WebServiceChartConfigTableConfigToJSON(value['tableConfig']),
        'queryValueConfig': WebServiceChartConfigQueryValueConfigToJSON(value['queryValueConfig']),
        'pieConfig': WebServiceChartConfigPieConfigToJSON(value['pieConfig']),
        'markers': value['markers'] == null ? undefined : (value['markers'].map(WebServiceChartConfigMarkerToJSON)),
        'lineConfig': WebServiceChartConfigLineConfigToJSON(value['lineConfig']),
    };
}
