// @sentio/sdk/src/eth/builtin/internal/eacaggregatorproxy-processor.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { BigNumberish, Overrides, Interface } from "ethers";
import {
  addContractByABI,
  getContractByABI,
  addProcessor,
  getProcessor,
  getProvider,
  transformEtherError,
  BindOptions,
  BaseProcessor,
  BaseProcessorTemplate,
  BoundContractView,
  ContractContext,
  ContractView,
  DummyProvider,
  EthChainId,
  TypedCallTrace,
  EthContext,
  EthFetchConfig,
  PreprocessResult,
} from "@sentio/sdk/eth";
import { EthCallParam, EthCallContext, PreparedData } from "@sentio/protos";

import { EACAggregatorProxy__factory } from "./index.js";
import {
  EACAggregatorProxy,
  AnswerUpdatedEvent,
  AnswerUpdatedEventFilter,
  NewRoundEvent,
  NewRoundEventFilter,
  OwnershipTransferRequestedEvent,
  OwnershipTransferRequestedEventFilter,
  OwnershipTransferredEvent,
  OwnershipTransferredEventFilter,
} from "./EACAggregatorProxy.js";

export interface AcceptOwnershipCallObject {}

export type AcceptOwnershipCallTrace = TypedCallTrace<
  [],
  AcceptOwnershipCallObject
>;

export interface AccessControllerCallObject {}

export type AccessControllerCallTrace = TypedCallTrace<
  [],
  AccessControllerCallObject
>;

export interface AggregatorCallObject {}

export type AggregatorCallTrace = TypedCallTrace<[], AggregatorCallObject>;

export interface ConfirmAggregatorCallObject {
  _aggregator: string;
}

export type ConfirmAggregatorCallTrace = TypedCallTrace<
  [string],
  ConfirmAggregatorCallObject
>;

export interface DecimalsCallObject {}

export type DecimalsCallTrace = TypedCallTrace<[], DecimalsCallObject>;

export interface DescriptionCallObject {}

export type DescriptionCallTrace = TypedCallTrace<[], DescriptionCallObject>;

export interface GetAnswerCallObject {
  _roundId: bigint;
}

export type GetAnswerCallTrace = TypedCallTrace<[bigint], GetAnswerCallObject>;

export interface GetRoundDataCallObject {
  _roundId: bigint;
}

export type GetRoundDataCallTrace = TypedCallTrace<
  [bigint],
  GetRoundDataCallObject
>;

export interface GetTimestampCallObject {
  _roundId: bigint;
}

export type GetTimestampCallTrace = TypedCallTrace<
  [bigint],
  GetTimestampCallObject
>;

export interface LatestAnswerCallObject {}

export type LatestAnswerCallTrace = TypedCallTrace<[], LatestAnswerCallObject>;

export interface LatestRoundCallObject {}

export type LatestRoundCallTrace = TypedCallTrace<[], LatestRoundCallObject>;

export interface LatestRoundDataCallObject {}

export type LatestRoundDataCallTrace = TypedCallTrace<
  [],
  LatestRoundDataCallObject
>;

export interface LatestTimestampCallObject {}

export type LatestTimestampCallTrace = TypedCallTrace<
  [],
  LatestTimestampCallObject
>;

export interface OwnerCallObject {}

export type OwnerCallTrace = TypedCallTrace<[], OwnerCallObject>;

export interface PhaseAggregatorsCallObject {}

export type PhaseAggregatorsCallTrace = TypedCallTrace<
  [bigint],
  PhaseAggregatorsCallObject
>;

export interface PhaseIdCallObject {}

export type PhaseIdCallTrace = TypedCallTrace<[], PhaseIdCallObject>;

export interface ProposeAggregatorCallObject {
  _aggregator: string;
}

export type ProposeAggregatorCallTrace = TypedCallTrace<
  [string],
  ProposeAggregatorCallObject
>;

export interface ProposedAggregatorCallObject {}

export type ProposedAggregatorCallTrace = TypedCallTrace<
  [],
  ProposedAggregatorCallObject
>;

export interface ProposedGetRoundDataCallObject {
  _roundId: bigint;
}

export type ProposedGetRoundDataCallTrace = TypedCallTrace<
  [bigint],
  ProposedGetRoundDataCallObject
>;

export interface ProposedLatestRoundDataCallObject {}

export type ProposedLatestRoundDataCallTrace = TypedCallTrace<
  [],
  ProposedLatestRoundDataCallObject
>;

export interface SetControllerCallObject {
  _accessController: string;
}

export type SetControllerCallTrace = TypedCallTrace<
  [string],
  SetControllerCallObject
>;

export interface TransferOwnershipCallObject {
  _to: string;
}

export type TransferOwnershipCallTrace = TypedCallTrace<
  [string],
  TransferOwnershipCallObject
>;

export interface VersionCallObject {}

export type VersionCallTrace = TypedCallTrace<[], VersionCallObject>;

const templateContract = EACAggregatorProxy__factory.connect(
  "0x0",
  DummyProvider
);

const iface = new Interface([
  "function acceptOwnership() returns ()",
  "function accessController() view returns (address)",
  "function aggregator() view returns (address)",
  "function confirmAggregator(address _aggregator) returns ()",
  "function decimals() view returns (uint8)",
  "function description() view returns (string)",
  "function getAnswer(uint256 _roundId) view returns (int256)",
  "function getRoundData(uint80 _roundId) view returns (uint80 roundId, int256 answer, uint256 startedAt, uint256 updatedAt, uint80 answeredInRound)",
  "function getTimestamp(uint256 _roundId) view returns (uint256)",
  "function latestAnswer() view returns (int256)",
  "function latestRound() view returns (uint256)",
  "function latestRoundData() view returns (uint80 roundId, int256 answer, uint256 startedAt, uint256 updatedAt, uint80 answeredInRound)",
  "function latestTimestamp() view returns (uint256)",
  "function owner() view returns (address)",
  "function phaseAggregators(uint16) view returns (address)",
  "function phaseId() view returns (uint16)",
  "function proposeAggregator(address _aggregator) returns ()",
  "function proposedAggregator() view returns (address)",
  "function proposedGetRoundData(uint80 _roundId) view returns (uint80 roundId, int256 answer, uint256 startedAt, uint256 updatedAt, uint80 answeredInRound)",
  "function proposedLatestRoundData() view returns (uint80 roundId, int256 answer, uint256 startedAt, uint256 updatedAt, uint80 answeredInRound)",
  "function setController(address _accessController) returns ()",
  "function transferOwnership(address _to) returns ()",
  "function version() view returns (uint256)",
]);

export class EACAggregatorProxyContractView extends ContractView<EACAggregatorProxy> {
  constructor(contract: EACAggregatorProxy) {
    super(contract);
    this.callStatic.contract = contract;
  }

  async accessController(
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<string> {
    try {
      return await this.contract.getFunction("accessController()")(
        overrides || {}
      );
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async aggregator(
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<string> {
    try {
      return await this.contract.getFunction("aggregator()")(overrides || {});
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async decimals(
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<bigint> {
    try {
      return await this.contract.getFunction("decimals()")(overrides || {});
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async description(
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<string> {
    try {
      return await this.contract.getFunction("description()")(overrides || {});
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async getAnswer(
    _roundId: BigNumberish,
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<bigint> {
    try {
      return await this.contract.getFunction("getAnswer(uint256)")(
        _roundId,
        overrides || {}
      );
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async getRoundData(
    _roundId: BigNumberish,
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<
    [bigint, bigint, bigint, bigint, bigint] & {
      roundId: bigint;
      answer: bigint;
      startedAt: bigint;
      updatedAt: bigint;
      answeredInRound: bigint;
    }
  > {
    try {
      return await this.contract.getFunction("getRoundData(uint80)")(
        _roundId,
        overrides || {}
      );
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async getTimestamp(
    _roundId: BigNumberish,
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<bigint> {
    try {
      return await this.contract.getFunction("getTimestamp(uint256)")(
        _roundId,
        overrides || {}
      );
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async latestAnswer(
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<bigint> {
    try {
      return await this.contract.getFunction("latestAnswer()")(overrides || {});
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async latestRound(
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<bigint> {
    try {
      return await this.contract.getFunction("latestRound()")(overrides || {});
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async latestRoundData(
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<
    [bigint, bigint, bigint, bigint, bigint] & {
      roundId: bigint;
      answer: bigint;
      startedAt: bigint;
      updatedAt: bigint;
      answeredInRound: bigint;
    }
  > {
    try {
      return await this.contract.getFunction("latestRoundData()")(
        overrides || {}
      );
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async latestTimestamp(
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<bigint> {
    try {
      return await this.contract.getFunction("latestTimestamp()")(
        overrides || {}
      );
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async owner(
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<string> {
    try {
      return await this.contract.getFunction("owner()")(overrides || {});
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async phaseAggregators(
    arg0: BigNumberish,
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<string> {
    try {
      return await this.contract.getFunction("phaseAggregators(uint16)")(
        arg0,
        overrides || {}
      );
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async phaseId(
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<bigint> {
    try {
      return await this.contract.getFunction("phaseId()")(overrides || {});
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async proposedAggregator(
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<string> {
    try {
      return await this.contract.getFunction("proposedAggregator()")(
        overrides || {}
      );
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async proposedGetRoundData(
    _roundId: BigNumberish,
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<
    [bigint, bigint, bigint, bigint, bigint] & {
      roundId: bigint;
      answer: bigint;
      startedAt: bigint;
      updatedAt: bigint;
      answeredInRound: bigint;
    }
  > {
    try {
      return await this.contract.getFunction("proposedGetRoundData(uint80)")(
        _roundId,
        overrides || {}
      );
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async proposedLatestRoundData(
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<
    [bigint, bigint, bigint, bigint, bigint] & {
      roundId: bigint;
      answer: bigint;
      startedAt: bigint;
      updatedAt: bigint;
      answeredInRound: bigint;
    }
  > {
    try {
      return await this.contract.getFunction("proposedLatestRoundData()")(
        overrides || {}
      );
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  async version(
    overrides?: Overrides,
    preparedData?: PreparedData,
    ethCallContext?: EthCallContext
  ): Promise<bigint> {
    try {
      return await this.contract.getFunction("version()")(overrides || {});
    } catch (e) {
      const stack = new Error().stack;
      throw transformEtherError(e, undefined, stack);
    }
  }

  callStatic = {
    contract: this.contract,

    async acceptOwnership(
      overrides?: Overrides,
      preparedData?: PreparedData,
      ethCallContext?: EthCallContext
    ): Promise<void> {
      try {
        return await this.contract
          .getFunction("acceptOwnership()")
          .staticCall(overrides || {});
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    async confirmAggregator(
      _aggregator: string,
      overrides?: Overrides,
      preparedData?: PreparedData,
      ethCallContext?: EthCallContext
    ): Promise<void> {
      try {
        return await this.contract
          .getFunction("confirmAggregator(address)")
          .staticCall(_aggregator, overrides || {});
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    async proposeAggregator(
      _aggregator: string,
      overrides?: Overrides,
      preparedData?: PreparedData,
      ethCallContext?: EthCallContext
    ): Promise<void> {
      try {
        return await this.contract
          .getFunction("proposeAggregator(address)")
          .staticCall(_aggregator, overrides || {});
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    async setController(
      _accessController: string,
      overrides?: Overrides,
      preparedData?: PreparedData,
      ethCallContext?: EthCallContext
    ): Promise<void> {
      try {
        return await this.contract
          .getFunction("setController(address)")
          .staticCall(_accessController, overrides || {});
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    async transferOwnership(
      _to: string,
      overrides?: Overrides,
      preparedData?: PreparedData,
      ethCallContext?: EthCallContext
    ): Promise<void> {
      try {
        return await this.contract
          .getFunction("transferOwnership(address)")
          .staticCall(_to, overrides || {});
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
  };

  encodeCall = {
    acceptOwnership(ethCallContext: EthCallContext): EthCallParam {
      try {
        const iface = new Interface(["function acceptOwnership()"]);
        const calldata = iface.encodeFunctionData("acceptOwnership", []);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    accessController(ethCallContext: EthCallContext): EthCallParam {
      try {
        const iface = new Interface(["function accessController()"]);
        const calldata = iface.encodeFunctionData("accessController", []);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    aggregator(ethCallContext: EthCallContext): EthCallParam {
      try {
        const iface = new Interface(["function aggregator()"]);
        const calldata = iface.encodeFunctionData("aggregator", []);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    confirmAggregator(
      _aggregator: string,
      ethCallContext: EthCallContext
    ): EthCallParam {
      try {
        const iface = new Interface(["function confirmAggregator(address)"]);
        const calldata = iface.encodeFunctionData("confirmAggregator", [
          _aggregator,
        ]);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    decimals(ethCallContext: EthCallContext): EthCallParam {
      try {
        const iface = new Interface(["function decimals()"]);
        const calldata = iface.encodeFunctionData("decimals", []);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    description(ethCallContext: EthCallContext): EthCallParam {
      try {
        const iface = new Interface(["function description()"]);
        const calldata = iface.encodeFunctionData("description", []);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    getAnswer(
      _roundId: BigNumberish,
      ethCallContext: EthCallContext
    ): EthCallParam {
      try {
        const iface = new Interface(["function getAnswer(uint256)"]);
        const calldata = iface.encodeFunctionData("getAnswer", [_roundId]);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    getRoundData(
      _roundId: BigNumberish,
      ethCallContext: EthCallContext
    ): EthCallParam {
      try {
        const iface = new Interface(["function getRoundData(uint80)"]);
        const calldata = iface.encodeFunctionData("getRoundData", [_roundId]);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    getTimestamp(
      _roundId: BigNumberish,
      ethCallContext: EthCallContext
    ): EthCallParam {
      try {
        const iface = new Interface(["function getTimestamp(uint256)"]);
        const calldata = iface.encodeFunctionData("getTimestamp", [_roundId]);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    latestAnswer(ethCallContext: EthCallContext): EthCallParam {
      try {
        const iface = new Interface(["function latestAnswer()"]);
        const calldata = iface.encodeFunctionData("latestAnswer", []);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    latestRound(ethCallContext: EthCallContext): EthCallParam {
      try {
        const iface = new Interface(["function latestRound()"]);
        const calldata = iface.encodeFunctionData("latestRound", []);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    latestRoundData(ethCallContext: EthCallContext): EthCallParam {
      try {
        const iface = new Interface(["function latestRoundData()"]);
        const calldata = iface.encodeFunctionData("latestRoundData", []);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    latestTimestamp(ethCallContext: EthCallContext): EthCallParam {
      try {
        const iface = new Interface(["function latestTimestamp()"]);
        const calldata = iface.encodeFunctionData("latestTimestamp", []);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    owner(ethCallContext: EthCallContext): EthCallParam {
      try {
        const iface = new Interface(["function owner()"]);
        const calldata = iface.encodeFunctionData("owner", []);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    phaseAggregators(
      arg0: BigNumberish,
      ethCallContext: EthCallContext
    ): EthCallParam {
      try {
        const iface = new Interface(["function phaseAggregators(uint16)"]);
        const calldata = iface.encodeFunctionData("phaseAggregators", [arg0]);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    phaseId(ethCallContext: EthCallContext): EthCallParam {
      try {
        const iface = new Interface(["function phaseId()"]);
        const calldata = iface.encodeFunctionData("phaseId", []);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    proposeAggregator(
      _aggregator: string,
      ethCallContext: EthCallContext
    ): EthCallParam {
      try {
        const iface = new Interface(["function proposeAggregator(address)"]);
        const calldata = iface.encodeFunctionData("proposeAggregator", [
          _aggregator,
        ]);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    proposedAggregator(ethCallContext: EthCallContext): EthCallParam {
      try {
        const iface = new Interface(["function proposedAggregator()"]);
        const calldata = iface.encodeFunctionData("proposedAggregator", []);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    proposedGetRoundData(
      _roundId: BigNumberish,
      ethCallContext: EthCallContext
    ): EthCallParam {
      try {
        const iface = new Interface(["function proposedGetRoundData(uint80)"]);
        const calldata = iface.encodeFunctionData("proposedGetRoundData", [
          _roundId,
        ]);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    proposedLatestRoundData(ethCallContext: EthCallContext): EthCallParam {
      try {
        const iface = new Interface(["function proposedLatestRoundData()"]);
        const calldata = iface.encodeFunctionData(
          "proposedLatestRoundData",
          []
        );
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    setController(
      _accessController: string,
      ethCallContext: EthCallContext
    ): EthCallParam {
      try {
        const iface = new Interface(["function setController(address)"]);
        const calldata = iface.encodeFunctionData("setController", [
          _accessController,
        ]);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    transferOwnership(
      _to: string,
      ethCallContext: EthCallContext
    ): EthCallParam {
      try {
        const iface = new Interface(["function transferOwnership(address)"]);
        const calldata = iface.encodeFunctionData("transferOwnership", [_to]);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
    version(ethCallContext: EthCallContext): EthCallParam {
      try {
        const iface = new Interface(["function version()"]);
        const calldata = iface.encodeFunctionData("version", []);
        return {
          context: ethCallContext,
          calldata,
        };
      } catch (e) {
        const stack = new Error().stack;
        throw transformEtherError(e, undefined, stack);
      }
    },
  };
}

export class EACAggregatorProxyBoundContractView extends BoundContractView<
  EACAggregatorProxy,
  EACAggregatorProxyContractView
> {
  async accessController(overrides?: Overrides): Promise<string> {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.accessController(
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async aggregator(overrides?: Overrides): Promise<string> {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.aggregator(
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async decimals(overrides?: Overrides): Promise<bigint> {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.decimals(
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async description(overrides?: Overrides): Promise<string> {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.description(
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async getAnswer(
    _roundId: BigNumberish,
    overrides?: Overrides
  ): Promise<bigint> {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.getAnswer(
      _roundId,
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async getRoundData(
    _roundId: BigNumberish,
    overrides?: Overrides
  ): Promise<
    [bigint, bigint, bigint, bigint, bigint] & {
      roundId: bigint;
      answer: bigint;
      startedAt: bigint;
      updatedAt: bigint;
      answeredInRound: bigint;
    }
  > {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.getRoundData(
      _roundId,
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async getTimestamp(
    _roundId: BigNumberish,
    overrides?: Overrides
  ): Promise<bigint> {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.getTimestamp(
      _roundId,
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async latestAnswer(overrides?: Overrides): Promise<bigint> {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.latestAnswer(
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async latestRound(overrides?: Overrides): Promise<bigint> {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.latestRound(
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async latestRoundData(
    overrides?: Overrides
  ): Promise<
    [bigint, bigint, bigint, bigint, bigint] & {
      roundId: bigint;
      answer: bigint;
      startedAt: bigint;
      updatedAt: bigint;
      answeredInRound: bigint;
    }
  > {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.latestRoundData(
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async latestTimestamp(overrides?: Overrides): Promise<bigint> {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.latestTimestamp(
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async owner(overrides?: Overrides): Promise<string> {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.owner(
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async phaseAggregators(
    arg0: BigNumberish,
    overrides?: Overrides
  ): Promise<string> {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.phaseAggregators(
      arg0,
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async phaseId(overrides?: Overrides): Promise<bigint> {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.phaseId(
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async proposedAggregator(overrides?: Overrides): Promise<string> {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.proposedAggregator(
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async proposedGetRoundData(
    _roundId: BigNumberish,
    overrides?: Overrides
  ): Promise<
    [bigint, bigint, bigint, bigint, bigint] & {
      roundId: bigint;
      answer: bigint;
      startedAt: bigint;
      updatedAt: bigint;
      answeredInRound: bigint;
    }
  > {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.proposedGetRoundData(
      _roundId,
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async proposedLatestRoundData(
    overrides?: Overrides
  ): Promise<
    [bigint, bigint, bigint, bigint, bigint] & {
      roundId: bigint;
      answer: bigint;
      startedAt: bigint;
      updatedAt: bigint;
      answeredInRound: bigint;
    }
  > {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.proposedLatestRoundData(
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  async version(overrides?: Overrides): Promise<bigint> {
    const ethCallContext = {
      chainId: this.context.chainId,
      blockTag: "0x" + this.context.blockNumber.toString(16),
      address: this.context.address,
    };
    return await this.view.version(
      {
        blockTag: this.context.blockNumber,
        ...overrides,
      },
      this.context.preparedData,
      ethCallContext
    );
  }

  callStatic = {
    view: this.view,
    context: this.context,

    async acceptOwnership(overrides?: Overrides): Promise<void> {
      const ethCallContext = {
        chainId: this.context.chainId,
        blockTag: "0x" + this.context.blockNumber.toString(16),
        address: this.context.address,
      };
      return await this.view.callStatic.acceptOwnership(
        {
          blockTag: this.context.blockNumber,
          ...overrides,
        },
        this.context.preparedData,
        ethCallContext
      );
    },
    async confirmAggregator(
      _aggregator: string,
      overrides?: Overrides
    ): Promise<void> {
      const ethCallContext = {
        chainId: this.context.chainId,
        blockTag: "0x" + this.context.blockNumber.toString(16),
        address: this.context.address,
      };
      return await this.view.callStatic.confirmAggregator(
        _aggregator,
        {
          blockTag: this.context.blockNumber,
          ...overrides,
        },
        this.context.preparedData,
        ethCallContext
      );
    },
    async proposeAggregator(
      _aggregator: string,
      overrides?: Overrides
    ): Promise<void> {
      const ethCallContext = {
        chainId: this.context.chainId,
        blockTag: "0x" + this.context.blockNumber.toString(16),
        address: this.context.address,
      };
      return await this.view.callStatic.proposeAggregator(
        _aggregator,
        {
          blockTag: this.context.blockNumber,
          ...overrides,
        },
        this.context.preparedData,
        ethCallContext
      );
    },
    async setController(
      _accessController: string,
      overrides?: Overrides
    ): Promise<void> {
      const ethCallContext = {
        chainId: this.context.chainId,
        blockTag: "0x" + this.context.blockNumber.toString(16),
        address: this.context.address,
      };
      return await this.view.callStatic.setController(
        _accessController,
        {
          blockTag: this.context.blockNumber,
          ...overrides,
        },
        this.context.preparedData,
        ethCallContext
      );
    },
    async transferOwnership(_to: string, overrides?: Overrides): Promise<void> {
      const ethCallContext = {
        chainId: this.context.chainId,
        blockTag: "0x" + this.context.blockNumber.toString(16),
        address: this.context.address,
      };
      return await this.view.callStatic.transferOwnership(
        _to,
        {
          blockTag: this.context.blockNumber,
          ...overrides,
        },
        this.context.preparedData,
        ethCallContext
      );
    },
  };

  encodeCall = {
    view: this.view,
    context: this.context,

    acceptOwnership(overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.acceptOwnership({
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    accessController(overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.accessController({
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    aggregator(overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.aggregator({
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    confirmAggregator(
      _aggregator: string,
      overrides?: Overrides
    ): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.confirmAggregator(_aggregator, {
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    decimals(overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.decimals({
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    description(overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.description({
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    getAnswer(_roundId: BigNumberish, overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.getAnswer(_roundId, {
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    getRoundData(_roundId: BigNumberish, overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.getRoundData(_roundId, {
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    getTimestamp(_roundId: BigNumberish, overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.getTimestamp(_roundId, {
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    latestAnswer(overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.latestAnswer({
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    latestRound(overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.latestRound({
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    latestRoundData(overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.latestRoundData({
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    latestTimestamp(overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.latestTimestamp({
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    owner(overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.owner({
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    phaseAggregators(arg0: BigNumberish, overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.phaseAggregators(arg0, {
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    phaseId(overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.phaseId({
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    proposeAggregator(
      _aggregator: string,
      overrides?: Overrides
    ): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.proposeAggregator(_aggregator, {
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    proposedAggregator(overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.proposedAggregator({
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    proposedGetRoundData(
      _roundId: BigNumberish,
      overrides?: Overrides
    ): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.proposedGetRoundData(_roundId, {
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    proposedLatestRoundData(overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.proposedLatestRoundData({
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    setController(
      _accessController: string,
      overrides?: Overrides
    ): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.setController(_accessController, {
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    transferOwnership(_to: string, overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.transferOwnership(_to, {
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
    version(overrides?: Overrides): EthCallParam {
      let blockTagWithOverride = "0x" + this.context.blockNumber.toString(16);
      if (overrides?.blockTag) {
        blockTagWithOverride =
          typeof overrides.blockTag == "string"
            ? overrides.blockTag
            : "0x" + overrides.blockTag.toString(16);
      }

      return this.view.encodeCall.version({
        chainId: this.context.chainId.toString(),
        address: this.context.address,
        blockTag: blockTagWithOverride,
      });
    },
  };
}

export type EACAggregatorProxyContext = ContractContext<
  EACAggregatorProxy,
  EACAggregatorProxyBoundContractView
>;

export class EACAggregatorProxyProcessor extends BaseProcessor<
  EACAggregatorProxy,
  EACAggregatorProxyBoundContractView
> {
  onEventAnswerUpdated(
    handler: (
      event: AnswerUpdatedEvent,
      ctx: EACAggregatorProxyContext
    ) => void,
    filter?: AnswerUpdatedEventFilter | AnswerUpdatedEventFilter[],
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      event: AnswerUpdatedEvent,
      ctx: EACAggregatorProxyContext,
      preprocessStore: { [k: string]: any }
    ) => Promise<PreprocessResult>
  ): this {
    if (!filter) {
      filter = templateContract.filters[
        "AnswerUpdated(int256,uint256,uint256)"
      ](null, null, null);
    }
    return super.onEthEvent(handler, filter!, fetchConfig, preprocessHandler);
  }

  onEventNewRound(
    handler: (event: NewRoundEvent, ctx: EACAggregatorProxyContext) => void,
    filter?: NewRoundEventFilter | NewRoundEventFilter[],
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      event: NewRoundEvent,
      ctx: EACAggregatorProxyContext,
      preprocessStore: { [k: string]: any }
    ) => Promise<PreprocessResult>
  ): this {
    if (!filter) {
      filter = templateContract.filters["NewRound(uint256,address,uint256)"](
        null,
        null,
        null
      );
    }
    return super.onEthEvent(handler, filter!, fetchConfig, preprocessHandler);
  }

  onEventOwnershipTransferRequested(
    handler: (
      event: OwnershipTransferRequestedEvent,
      ctx: EACAggregatorProxyContext
    ) => void,
    filter?:
      | OwnershipTransferRequestedEventFilter
      | OwnershipTransferRequestedEventFilter[],
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      event: OwnershipTransferRequestedEvent,
      ctx: EACAggregatorProxyContext,
      preprocessStore: { [k: string]: any }
    ) => Promise<PreprocessResult>
  ): this {
    if (!filter) {
      filter = templateContract.filters[
        "OwnershipTransferRequested(address,address)"
      ](null, null);
    }
    return super.onEthEvent(handler, filter!, fetchConfig, preprocessHandler);
  }

  onEventOwnershipTransferred(
    handler: (
      event: OwnershipTransferredEvent,
      ctx: EACAggregatorProxyContext
    ) => void,
    filter?:
      | OwnershipTransferredEventFilter
      | OwnershipTransferredEventFilter[],
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      event: OwnershipTransferredEvent,
      ctx: EACAggregatorProxyContext,
      preprocessStore: { [k: string]: any }
    ) => Promise<PreprocessResult>
  ): this {
    if (!filter) {
      filter = templateContract.filters[
        "OwnershipTransferred(address,address)"
      ](null, null);
    }
    return super.onEthEvent(handler, filter!, fetchConfig, preprocessHandler);
  }

  onCallAcceptOwnership(
    handler: (
      call: AcceptOwnershipCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: AcceptOwnershipCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0x79ba5097",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallAccessController(
    handler: (
      call: AccessControllerCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: AccessControllerCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0xbc43cbaf",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallAggregator(
    handler: (
      call: AggregatorCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: AggregatorCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0x245a7bfc",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallConfirmAggregator(
    handler: (
      call: ConfirmAggregatorCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: ConfirmAggregatorCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0xa928c096",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallDecimals(
    handler: (call: DecimalsCallTrace, ctx: EACAggregatorProxyContext) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: DecimalsCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0x313ce567",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallDescription(
    handler: (
      call: DescriptionCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: DescriptionCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0x7284e416",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallGetAnswer(
    handler: (call: GetAnswerCallTrace, ctx: EACAggregatorProxyContext) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: GetAnswerCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0xb5ab58dc",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallGetRoundData(
    handler: (
      call: GetRoundDataCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: GetRoundDataCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0x9a6fc8f5",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallGetTimestamp(
    handler: (
      call: GetTimestampCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: GetTimestampCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0xb633620c",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallLatestAnswer(
    handler: (
      call: LatestAnswerCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: LatestAnswerCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0x50d25bcd",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallLatestRound(
    handler: (
      call: LatestRoundCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: LatestRoundCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0x668a0f02",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallLatestRoundData(
    handler: (
      call: LatestRoundDataCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: LatestRoundDataCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0xfeaf968c",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallLatestTimestamp(
    handler: (
      call: LatestTimestampCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: LatestTimestampCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0x8205bf6a",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallOwner(
    handler: (call: OwnerCallTrace, ctx: EACAggregatorProxyContext) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: OwnerCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0x8da5cb5b",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallPhaseAggregators(
    handler: (
      call: PhaseAggregatorsCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: PhaseAggregatorsCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0xc1597304",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallPhaseId(
    handler: (call: PhaseIdCallTrace, ctx: EACAggregatorProxyContext) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: PhaseIdCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0x58303b10",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallProposeAggregator(
    handler: (
      call: ProposeAggregatorCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: ProposeAggregatorCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0xf8a2abd3",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallProposedAggregator(
    handler: (
      call: ProposedAggregatorCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: ProposedAggregatorCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0xe8c4be30",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallProposedGetRoundData(
    handler: (
      call: ProposedGetRoundDataCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: ProposedGetRoundDataCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0x6001ac53",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallProposedLatestRoundData(
    handler: (
      call: ProposedLatestRoundDataCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: ProposedLatestRoundDataCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0x8f6b4d91",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallSetController(
    handler: (
      call: SetControllerCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: SetControllerCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0x92eefe9b",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallTransferOwnership(
    handler: (
      call: TransferOwnershipCallTrace,
      ctx: EACAggregatorProxyContext
    ) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: TransferOwnershipCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0xf2fde38b",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  onCallVersion(
    handler: (call: VersionCallTrace, ctx: EACAggregatorProxyContext) => void,
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      call: VersionCallTrace,
      ctx: EACAggregatorProxyContext
    ) => Promise<PreprocessResult>
  ): this {
    return super.onEthTrace(
      "0x54fd4d50",
      handler as any,
      fetchConfig,
      preprocessHandler
    );
  }

  public static filters = {
    AnswerUpdated(
      current?: BigNumberish | null,
      roundId?: BigNumberish | null,
      updatedAt?: null
    ): AnswerUpdatedEventFilter {
      return templateContract.filters["AnswerUpdated(int256,uint256,uint256)"](
        current,
        roundId,
        updatedAt
      );
    },
    NewRound(
      roundId?: BigNumberish | null,
      startedBy?: string | null,
      startedAt?: null
    ): NewRoundEventFilter {
      return templateContract.filters["NewRound(uint256,address,uint256)"](
        roundId,
        startedBy,
        startedAt
      );
    },
    OwnershipTransferRequested(
      from?: string | null,
      to?: string | null
    ): OwnershipTransferRequestedEventFilter {
      return templateContract.filters[
        "OwnershipTransferRequested(address,address)"
      ](from, to);
    },
    OwnershipTransferred(
      from?: string | null,
      to?: string | null
    ): OwnershipTransferredEventFilter {
      return templateContract.filters["OwnershipTransferred(address,address)"](
        from,
        to
      );
    },
  };

  protected CreateBoundContractView(): EACAggregatorProxyBoundContractView {
    const view = getEACAggregatorProxyContract(
      this.config.network,
      this.config.address
    );
    return new EACAggregatorProxyBoundContractView(this.config.address, view);
  }

  public static bind(options: BindOptions): EACAggregatorProxyProcessor {
    if (!options.name) {
      options.name = "EACAggregatorProxy";
    }
    let processor = getProcessor(options) as EACAggregatorProxyProcessor;
    if (!processor) {
      processor = new EACAggregatorProxyProcessor(options);
      addProcessor(options, processor);
    }
    return processor;
  }
}

export class EACAggregatorProxyProcessorTemplate extends BaseProcessorTemplate<
  EACAggregatorProxy,
  EACAggregatorProxyBoundContractView
> {
  bindInternal(options: BindOptions) {
    if (!options.name) {
      options.name = "EACAggregatorProxy";
    }
    let processor = getProcessor(options) as EACAggregatorProxyProcessor;
    if (!processor) {
      processor = new EACAggregatorProxyProcessor(options);
      addProcessor(options, processor);
    }
    return processor;
  }

  onEventAnswerUpdated(
    handler: (
      event: AnswerUpdatedEvent,
      ctx: EACAggregatorProxyContext
    ) => void,
    filter?: AnswerUpdatedEventFilter | AnswerUpdatedEventFilter[],
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      event: AnswerUpdatedEvent,
      ctx: EACAggregatorProxyContext,
      preprocessStore: { [k: string]: any }
    ) => Promise<PreprocessResult>
  ): this {
    if (!filter) {
      filter = templateContract.filters[
        "AnswerUpdated(int256,uint256,uint256)"
      ](null, null, null);
    }
    return super.onEthEvent(handler, filter!, fetchConfig, preprocessHandler);
  }

  onEventNewRound(
    handler: (event: NewRoundEvent, ctx: EACAggregatorProxyContext) => void,
    filter?: NewRoundEventFilter | NewRoundEventFilter[],
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      event: NewRoundEvent,
      ctx: EACAggregatorProxyContext,
      preprocessStore: { [k: string]: any }
    ) => Promise<PreprocessResult>
  ): this {
    if (!filter) {
      filter = templateContract.filters["NewRound(uint256,address,uint256)"](
        null,
        null,
        null
      );
    }
    return super.onEthEvent(handler, filter!, fetchConfig, preprocessHandler);
  }

  onEventOwnershipTransferRequested(
    handler: (
      event: OwnershipTransferRequestedEvent,
      ctx: EACAggregatorProxyContext
    ) => void,
    filter?:
      | OwnershipTransferRequestedEventFilter
      | OwnershipTransferRequestedEventFilter[],
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      event: OwnershipTransferRequestedEvent,
      ctx: EACAggregatorProxyContext,
      preprocessStore: { [k: string]: any }
    ) => Promise<PreprocessResult>
  ): this {
    if (!filter) {
      filter = templateContract.filters[
        "OwnershipTransferRequested(address,address)"
      ](null, null);
    }
    return super.onEthEvent(handler, filter!, fetchConfig, preprocessHandler);
  }

  onEventOwnershipTransferred(
    handler: (
      event: OwnershipTransferredEvent,
      ctx: EACAggregatorProxyContext
    ) => void,
    filter?:
      | OwnershipTransferredEventFilter
      | OwnershipTransferredEventFilter[],
    fetchConfig?: Partial<EthFetchConfig>,
    preprocessHandler?: (
      event: OwnershipTransferredEvent,
      ctx: EACAggregatorProxyContext,
      preprocessStore: { [k: string]: any }
    ) => Promise<PreprocessResult>
  ): this {
    if (!filter) {
      filter = templateContract.filters[
        "OwnershipTransferred(address,address)"
      ](null, null);
    }
    return super.onEthEvent(handler, filter!, fetchConfig, preprocessHandler);
  }
}

export function getEACAggregatorProxyContract(
  chainId: EthChainId,
  address: string
): EACAggregatorProxyContractView {
  let contract = getContractByABI(
    "EACAggregatorProxy",
    address,
    chainId
  ) as EACAggregatorProxyContractView;
  if (!contract) {
    const rawContract = EACAggregatorProxy__factory.connect(
      address,
      getProvider(chainId)
    );
    contract = new EACAggregatorProxyContractView(rawContract);
    addContractByABI("EACAggregatorProxy", address, chainId, contract);
  }
  return contract;
}

export function getEACAggregatorProxyContractOnContext(
  context: EthContext,
  address: string
): EACAggregatorProxyBoundContractView {
  const view = getEACAggregatorProxyContract(context.getChainId(), address);
  const boundView = new EACAggregatorProxyBoundContractView(address, view);
  boundView.context = context;
  if (boundView.callStatic) {
    boundView.callStatic.context = context;
  }
  return boundView;
}
