// @sentio/sdk/src/eth/builtin/internal/weth9-test-utils.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { LogParams } from "ethers/providers";
import { EthChainId } from "@sentio/sdk/eth";
import {
  ApprovalEventObject,
  TransferEventObject,
  DepositEventObject,
  WithdrawalEventObject,
} from "./WETH9.js";
import { getWETH9Contract } from "./weth9-processor.js";
const mockField = {
  blockHash:
    "0x0000000000000000000000000000000000000000000000000000000000000000",
  blockNumber: 0,
  logIndex: 0,
  removed: false,
  transactionHash:
    "0x0000000000000000000000000000000000000000000000000000000000000000",
  transactionIndex: 0,
};

export function mockApprovalLog(
  contractAddress: string,
  event: ApprovalEventObject
): LogParams {
  const contract = getWETH9Contract(EthChainId.ETHEREUM, contractAddress);
  const encodedLog = contract.rawContract.interface.encodeEventLog(
    "Approval(address,address,uint256)",
    [event.src, event.guy, event.wad]
  );
  return {
    ...mockField,
    index: 0,
    address: contractAddress,
    data: encodedLog.data,
    topics: encodedLog.topics,
  };
}

export function mockTransferLog(
  contractAddress: string,
  event: TransferEventObject
): LogParams {
  const contract = getWETH9Contract(EthChainId.ETHEREUM, contractAddress);
  const encodedLog = contract.rawContract.interface.encodeEventLog(
    "Transfer(address,address,uint256)",
    [event.src, event.dst, event.wad]
  );
  return {
    ...mockField,
    index: 0,
    address: contractAddress,
    data: encodedLog.data,
    topics: encodedLog.topics,
  };
}

export function mockDepositLog(
  contractAddress: string,
  event: DepositEventObject
): LogParams {
  const contract = getWETH9Contract(EthChainId.ETHEREUM, contractAddress);
  const encodedLog = contract.rawContract.interface.encodeEventLog(
    "Deposit(address,uint256)",
    [event.dst, event.wad]
  );
  return {
    ...mockField,
    index: 0,
    address: contractAddress,
    data: encodedLog.data,
    topics: encodedLog.topics,
  };
}

export function mockWithdrawalLog(
  contractAddress: string,
  event: WithdrawalEventObject
): LogParams {
  const contract = getWETH9Contract(EthChainId.ETHEREUM, contractAddress);
  const encodedLog = contract.rawContract.interface.encodeEventLog(
    "Withdrawal(address,uint256)",
    [event.src, event.wad]
  );
  return {
    ...mockField,
    index: 0,
    address: contractAddress,
    data: encodedLog.data,
    topics: encodedLog.topics,
  };
}
