// @sentio/sdk/lib/btc/types.js
import { BaseContext, normalizeLabels } from '../core/index.js';
export class BTCContext extends BaseContext {
    chainId;
    name;
    tx;
    address;
    constructor(chainId, name, tx, address) {
        super({});
        this.chainId = chainId;
        this.name = name;
        this.tx = tx;
        this.address = address;
    }
    getMetaDataInternal(name, labels) {
        return {
            address: this.address,
            contractName: this.name,
            blockNumber: BigInt(this.tx.blockheight ?? 0),
            transactionIndex: 0,
            transactionHash: this.tx.txid,
            chainId: this.getChainId(),
            name: name,
            logIndex: 0,
            labels: normalizeLabels(labels)
        };
    }
    getChainId() {
        return this.chainId;
    }
}
export class BTCBlockContext extends BaseContext {
    chainId;
    name;
    block;
    address;
    constructor(chainId, name, block, address) {
        super({});
        this.chainId = chainId;
        this.name = name;
        this.block = block;
        this.address = address;
    }
    getMetaDataInternal(name, labels) {
        return {
            address: this.address ?? '',
            contractName: this.name,
            blockNumber: BigInt(this.block.height ?? 0),
            transactionIndex: 0,
            transactionHash: '',
            chainId: this.getChainId(),
            name: name,
            logIndex: 0,
            labels: normalizeLabels(labels)
        };
    }
    getChainId() {
        return this.chainId;
    }
}
